{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bca4648f-79f2-458c-afe5-77d7cf3359af",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pathlib\n",
    "import itertools\n",
    "import string\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# for plotting\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('seaborn-whitegrid')\n",
    "\n",
    "''' This part of the lab module is used for making transformations\n",
    "\n",
    "\n",
    "HELPER FUNCTIONS: \n",
    "-----------------\n",
    "ng_to_nmol\n",
    "ODtime\n",
    "time_to_inculate\n",
    "\n",
    "\n",
    "'''\n",
    "# Helper functions\n",
    "def ng_to_nmol(ng, bp):\n",
    "\n",
    "    ''' To do a transformation it is important to have the right ratio of plasmid to insert.\n",
    "        In other words this is done by calculating the nanomolar ratios and this tool can do that\n",
    "\n",
    "        parm: ng        eg. nanogram\n",
    "        param: bp       eg. basepairs\n",
    "\n",
    "        nmol = ng/(bp*650)\n",
    "    '''\n",
    "    if ng > 0 and bp > 0: \n",
    "        return ng/(bp*650)\n",
    "    else: \n",
    "        return 'non-valid_input'\n",
    "\n",
    "def ODtime(initialOD: float, time: float, td: float = 0.5):\n",
    "    \"\"\"\n",
    "    - initialOD in OD\n",
    "    - time in hours\n",
    "    - doupling time : td in h^-1\n",
    "    \"\"\"\n",
    "    if initialOD >= 0 and time >= 0: \n",
    "        return round(initialOD*2**(time*td),3)\n",
    "    else:\n",
    "        return 'non-valid_input'\n",
    "\n",
    "\n",
    "def time_to_inculate(initialOD = 0.0025, td=0.4, verbose=False, transformation_time:int = 12):\n",
    "\n",
    "    '''\n",
    "    Input:\n",
    "\n",
    "    param: initialOD in OD\n",
    "    param: td is doubling time \n",
    "    param: transformations time is given as the time you want to transform\n",
    "    param: verbose : Provides extra information \n",
    "\n",
    "\n",
    "    Returns:\n",
    "    A plot of cell growth at different td\n",
    "    Note: This is used to calculate when the cells should be used for transformation\n",
    "\n",
    "    # OD 1 = 1 * 10^7 cells / ml\n",
    "    For a succesfull S.cerevisiae transformation between 1 to 2 Ã— 10^7 cells/ml should be used\n",
    "    Normal doupling time is between 3-6 hours\n",
    "\n",
    "\n",
    "    '''\n",
    "    if verbose:\n",
    "        print(\"GOAL: to get enough cells in exponential phase for transformation\")\n",
    "    print(\"Assumed that: \")\n",
    "    print(\"- transformation time \" +str(transformation_time)  +\" (reached OD=1 the day after)\")\n",
    "\n",
    "\n",
    "    times = list(range(0, 37))\n",
    "    ods_025_3     = [ODtime(initialOD,time, td=0.3) for time in times]\n",
    "    ods_025_input = [ODtime(initialOD,time, td=td) for time in times]\n",
    "    ods_025_5     = [ODtime(initialOD,time, td=0.5) for time in times]\n",
    "    fig = plt.figure()\n",
    "    ax = plt.axes()\n",
    "\n",
    "    #ax.set_xlim(lims)\n",
    "    ax.set_ylim([0, 2.2])\n",
    "    ax.plot(times, [2] * len(times), 'r-', label='end of exponential phase')\n",
    "    ax.plot(times, [1] * len(times), 'k-', label='target')\n",
    "    ax.plot(times, ods_025_3    , label='iOD='+str(initialOD)+', td=0.3');\n",
    "    ax.plot(times, ods_025_input, label='iOD='+str(initialOD)+', td=' + str(td));\n",
    "    ax.plot(times, ods_025_5    , label='iOD='+str(initialOD)+', td=0.5');\n",
    "\n",
    "    plt.xlabel('time, h^-1')\n",
    "    plt.ylabel('OD')\n",
    "    plt.legend()\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "    def inoculation_time(times,ods):\n",
    "\n",
    "        def find_closest(A, target):\n",
    "            #A must be sorted\n",
    "            idx = A.searchsorted(target)\n",
    "            idx = np.clip(idx, 1, len(A)-1)\n",
    "            left = A[idx-1]\n",
    "            right = A[idx]\n",
    "            idx -= target - left < right - target\n",
    "            return idx\n",
    "        \n",
    "        \n",
    "        # In how many hours will the cells have reached OD 1?\n",
    "        hours_to_OD1 = times[find_closest(np.array(ods), 1)]\n",
    "        print(\"Hours to OD = 1: \\t\" + str(hours_to_OD1)+ \" hours\")\n",
    "        \n",
    "        ### When do u need to innoculate?\n",
    "        when_to_inoculate = 11+24 - hours_to_OD1\n",
    "        print(\"Time of inoculation: \\t\" + str(when_to_inoculate) + \"(the day before)\")\n",
    "        \n",
    "        # If i innoculate now? \n",
    "        from datetime import datetime, timedelta, time\n",
    "        print('\\nIf you innoculate now, the cells will have reached OD= 1 by:  ', datetime.now()  + timedelta(hours=hours_to_OD1))\n",
    "        \n",
    "    inoculation_time(times,ods_025_input)\n",
    "\n",
    "    \n",
    "    if verbose:\n",
    "        print()\n",
    "        print(\"How to hit initialOD = 0.0025 (e.g. from colony)? Guess. Inoculate 9/10 + 1/10 'normal' colony per ~10 ml\")\n",
    "        print(\"How much volume? ~2 ml per transformation\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8e4fee22-d1b7-4c79-a53a-90644fc43906",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Assumed that: \n",
      "- transformation time 12 (reached OD=1 the day after)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYEAAAEBCAYAAACe6Rn8AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAABOlklEQVR4nO3dd3RU1drH8W96L0ACCSShhE7EGKoiKIKCCiIiTQUVLor3igKCIMKVK7wCKha6WFBRpFoQK4iC0oRAgIQJJQ1CSCN1ZjIts98/BiKBJLSZOZPM/qzlCpl2fnMS58nZZ59nuwghBJIkSZJTclU6gCRJkqQcWQQkSZKcmCwCkiRJTkwWAUmSJCcmi4AkSZITk0VAkiTJibkrHeB6xMfHKx1BkiSpVurUqVPVd4ha5MCBAzf83GPHjlkxiW3Vlqwy541bdmiZiPk0RuhN+kq3O1TWojNCvBYoxP6Pr7jLoXLWwNo54zMKRNNpW8TPiees+rq23p81fXbK4SBJUkC2NpsG3g3wdPNUOkr1CtIsX+s1VzaHA0nKKgGgQ+NAhZNYjywCkqSAbE02YX5hSseoWUGq5Wv9FsrmcCDHsooJ8vGgSbCP0lGsRhYBSVJAtiabcL9wpWPUrDANXD0gKELpJA7jWFYJHRoH4uLionQUq5FFQJLsTAhRS44E0iA4ClzdlE7iEEzlZpKzS2kfXneGgkAWAUmyuxJDCVqTthYUgVQ5FHSJlDwNepOZDk1kEZAk6SZka7IBaOTXSOEkNRACCtOhvjwpfFFSVjEAHRoHKZzEumQRkCQ7y9HmABDm68BHAtoC0JfImUGXOJZVgpe7Ky1C/JSOYlWyCEiSnV08EnDoE8NyZtAVkrJKaBsWgLtb3frYrFvvRpJqgWxNNu4u7oT4hCgdpXqFF64RkMNBgOVkflJWMe3r2FAQyCIgSXZ3TnOOUN9Q3Bx51k1BGuACwU2VTuIQMgvLKNGZ6tRFYhfJIiBJdlY7poemQmAT8PBWOolDqItXCl8ki4Ak2VmtKAKFaXIo6BLHzpXg6gJtw2QRkCTpJpiFmRxtjuMXgYI0qNdM6RQO41hWMS1C/fHxdOAhvBski4Ak2VGBrgCj2ejY00P1paDJlTODLpF0oV1EXWST9QSMRiMzZszg7NmzGAwGnnvuOfr06VNx//bt21m6dCnu7u4MGTKEYcOGodPpmDp1KufPn8fPz48FCxZQv359W8STJMXkaC5cI+DIRwKF6ZavcjgIgAKNgXPFujpbBGxyJLB582aCg4NZs2YNH330EXPmzKm4z2g0Mm/ePD755BNWr17NunXryM/P56uvvqJ169asWbOGhx9+mGXLltkimiQp6pzmHODgRUC2kK7kWMVJ4bo3PRRsVAT69+/Piy++CFjm17q5/TOOlpKSQlRUFEFBQXh6etKpUyf2799PfHw8PXv2BKBXr17s2bPHFtEkSVG160IxWQTgn3YRda1x3EU2GQ7y87NcVq1Wq3nhhReYOHFixX1qtZqAgIBKj1Wr1ZVu9/Pzo7S01HqBPv+cqCVLwNfXeq9pQ1Faba3IKnNev+wuGrzaQnD/h4Er2xE7RNbWpyHEHfoPqvYhDpHzGlgjZ1LLB2kcEEG9B++zUqorXVPOMWNg9Girb9tmawyfO3eO//znPzz22GMMHDiw4nZ/f380Gk3F9xqNhoCAgEq3azQaAgOrrroqleq6swRlZRFgNqPRaq/7uUow15KsMuf1y/Qy0lDtglZbVuX9jpDV27MMNO7oasjhCDmvhTVyJvqE0qYky6bv91pyFmdlUXwDn39XY5MikJ+fz5gxY/jvf//L7bffXum+6OhoMjIyKCoqwtfXlwMHDjB27FiysrLYsWMHHTt2ZOfOndUuityuXbvrD9SuHapBg27suQpQqVS1IqvMef3yf3yCxm7e+P39UZX3O0TWd2+BqO74vf9htQ9xiJzX4GZzag0m0l77hYEP3YHfxy9ZMVll15LTD2h8g68fHx9f7X02KQIrVqygpKSEZcuWVZzgHTp0KGVlZQwfPpzp06czduxYhBAMGTKERo0aMXLkSKZNm8bIkSPx8PBg4cKFtogmSYo6pzlH9/DuSseonskAJZnyfMAFydmlCAHt6+jMILBREZg5cyYzZ86s9v577rmHe+65p9JtPj4+LFq0yBZxJMkhmMwm8svyHfukcNFpEGY5M+iCutwu4iJ5sZgk2UmeNg+zMDv49FDZQvpSdXFh+cvJIiBJdpKttUwPdewikGL5KoeDAEg8W/cWlr+cLAKSZCcXrxFw6JYROUngGwJ+oUonUZxGb+LYuRI6Na2ndBSbkkVAkuykVlwtnJMIjTpAHf7L91odOl1EuVnQuVndbl8ji4Ak2Um2JpsAjwD8Pf2VjlK1chPkqiDsFqWTOIT96QW4ukBcVLDSUWxKFgFJspNsTTaN/BopHaN6Balg0lmOBCQOZBTQNiyQAG8PpaPYlCwCkmQnDr+YTM5Ry9dGMcrmcADGcjOHThfRpVndPh8AsghIkt04/GIyOUng6g6hbZROojjVuRK0hnK6NK/b5wNAFgFJsgudSUeBrsCxLxTLToSQ1uDupXQSxf2dVgBA56ayCEiSZAU52lqwmExOkhwKuuBAeiGR9X0IC/JWOorNySIgSXbg8NcIaAssPYPkSWGEEBzIKKCLExwFgCwCkmQXFUXAUY8EcpIsX8PkkUD6eS35akOdvz7gIlkEJMkOLhYBh50ierEINJLXCOxPt5wPcIaZQSCLgCTZRbY2m/re9fFyc9CTrjlHLe0i/BsqnURxB9ILqOfrQcuGDnpRn5XJIiBJdpClznLsmUE5SZahINkuggPphXRqWr9ON427lCwCkmQHqcWpNA9y0M6cF9tFyJlB5JXqSc3XOM1QEMgiIEk2pzFqyNZk0yLIQXv0F6RcaBchi0B8xoXrA5zkpDDYcKH5w4cP8/bbb7N69eqK2/Ly8pg8eXLF9yqVipdeeokRI0bQq1cvmjVrBkBsbCwvvWS79TwlyZ7SitMAaBHsoEUgJ9HyVU4PZX96IV7urtzSJEjpKHZjkyLw4YcfsnnzZnx8Kq/GExoaWlEUDh06xLvvvsuwYcM4ffo0HTp0YMWKFbaII0mKSi22rNblsEcC2YmyXcQFB9ILiI0MxtPdeQZJbPJOo6KiWLx4cbX3CyGYM2cOs2fPxs3NjaSkJHJychg1ahTjxo0jNTXVFrEkSREpRSm4u7oTGRCpdJSq5SRBSBunbxehNZhIzCqhixMNBYGNjgT69etHZmZmtfdv376dVq1a0aKF5S+j0NBQnnnmGe6//34OHDjA1KlT2bRpU5XPValUN5RJp9Pd8HPtrbZklTmvzZHMI4R7hXPy+MmrPlaJrC0zD6FteBtZ17FdpffptbqenAnnyig3Cxq5qe3+3pTcnzY7J1CTzZs3M3r06IrvY2JicHNzA6Bz587k5uYihKhyila7du1uaJsqleqGn2tvtSWrzHlt8pLzaNuw7TVlsHtWbQGU5RLUugdB17FdpffptbqenL+cPYGLCwy6syOBdl5DwNb7Mz4+vtr7FBn4SkxMJC4uruL7JUuW8NlnnwGQnJxMeHi408zRleo2fbmeM6VnHPd8QMWVwvKk8IH0QtqGBdq9ACjNLkcC33//PVqtluHDh1NQUIC/v3+lD/lnnnmGqVOnsmPHDtzc3Jg3b549YkmSzWWUZGAWZqKDo5WOUrWKmUHO3S7CVG7m4OlCHu0UoXQUu7NZEYiIiGD9+vUADBw4sOL2+vXr891331V6bFBQECtXrrRVFElSjMPPDMpJBL9QCHDQnkZ2ojpXallExslOCoO8WEySbCqtKA0XXGga2FTpKFXLTpRDQfzTNK6zE10pfJEsApJkQynFKUQERODt7oCLk8h2ERUOZBQQUc+H8CCfqz+4jpFFQJJsKLU41XGHggpSoFzv9EVACMH+9EKnHAoCWQQkyWZMZhMZxRmOWwSyj1q+OvlCMhnnteSV6p1yKAhkEZAkmzmrPovBbHDgnkFJlnYRIc7dLuLi+YCu8khAkiRrSi2qBTODQtqAu6fSSRR1IL2QYF8PokOdYxGZy8kiIEk2cnF6qMOuI5Cd6PRDQQD7Mwro3LQerq7OeYGqLAKSZCOpxak09G1IgGeA0lGupC2A0iynnx56Xq0nNU/jVOsHXE4WAUmykdQiB54ZVHGlsHMfCRzIKAScZ1H5qsgiIEk2IIRw7OmhF3sGhTl3u4j9aQV4ursS40SLyFxOFgFJsoEcbQ5ak9Zxi0D2hXYR/g2VTqKo/RmFxEYG4+XupnQUxcgiIEk2UDEzyGGnhx51+qEgrcFE0tlipx4KAlkEJMkmHLpxXLkJcpOd/qTwwYwiTGbh1CeFQRYBSbKJlOIUgryCqO/tgB8w509Z2kU4+fmAP47n4unuSrfmDvgzsiNZBCTJBlKLUokOinbMxZHkzCAAfj+eS/cWDfD1VGSBRYchi4Ak2UBacZrjXiSWkwiuHhDSWukkijl9XktKnobebUKVjqI4WQQkycoKdAUU6gsd83wAWGYGhTp3u4g/TuQC0LuNc8+OAhsWgcOHDzNq1Kgrbv/000958MEHGTVqFKNGjSI1NRWdTseECRN47LHHGDduHAUFBbaKJUk2d3FmkOMuKZnk9CeFf0/OpXmIH81C/JSOojibDIZ9+OGHbN68GR+fKxdoSExMZMGCBcTE/DMeuWrVKlq3bs2ECRP44YcfWLZsGTNnzrRFNEmyOYeeGVTRLsJ5zwfojOXsTjnPY92ilI7iEGxyJBAVFcXixYurvC8pKYmVK1cycuRIPvjgAwDi4+Pp2bMnAL169WLPnj22iCVJdpFanIqPuw9hfmFKR7lSxUlh5z0S2JN6Hr3JLIeCLrDJkUC/fv3IzMys8r4HH3yQxx57DH9/f55//nl+//131Go1AQGWJlt+fn6UlpZW+9oqleqGMul0uht+rr3VlqwyZ9WOnj1KY6/GJCcnX/dzbZ213vHfCANOlHpRfhPbqc0/+6/35ePl7kKQIReVKl+hZJUpuT/tOjdKCMGTTz5Z8YF/1113cezYMfz9/dFoNABoNBoCAwOrfY127drd0LZVKtUNP9feaktWmbNqOYk5dA3rekPbtHnW43ng15DWt915Uy9TW3/2QggOf/8HPVuFcmuM4xwN2Xp/xsfHV3ufXWcHqdVqBgwYgEajQQjBvn37iImJIS4ujh07dgCwc+dOOnXqZM9YkmQ1aoOaHG2OA7eLSHTqoaDUfA2nC7TcLYeCKtjlSOD7779Hq9UyfPhwJk2axOjRo/H09OT222/nrrvuomvXrkybNo2RI0fi4eHBwoUL7RFLkqwurTgNcNCTwgaNZWbQHS8onUQxvydbpobeLa8PqGCzIhAREcH69esBGDhwYMXtDz/8MA8//HClx/r4+LBo0SJbRZEku3HomUGZ+8FsgqZ3KJ1EMX8cz6N1I38i6vkqHcVhyIvFJMmKUopT8HD1ICIgQukoV8rYAy6uENlV6SSK0OhN7Es7L2cFXUYWAUmyorSiNJoGNsXd1QH70WTsslwf4O2cC6jsOpWPsVzI8wGXkUVAkqzIYVcTMxkg8wA07aF0EsX8fjwPfy93Ojv5+gGXk0VAkqxEX64nU53pmDODziWAqQya3q50EkUIIfjjeC53tgzBw01+7F1K7g1JspL04nTMwuyYRwIZuy1fo5zzpPDxnFLOFevo3VbOCrqcLAKSZCUOPT00Yzc0aAX+zvkh+HtyHoA8H1AFWQQkyUpSilNwdXGlWVAzpaNUZi6H03udemro78dzaR8eSKNAb6WjOBxZBCTJSlKLUmni3wQvNy+lo1SWewz0xU5bBIrLjMRnFMqhoGrIIiBJVpJabFlS0uFkXOjK66RF4K+T+ZSbhbw+oBqyCEiSFZjMJtJL0mke7IBLSmbsgsAICHbO/vm/H88lyMeD2MhgpaM4JFkEJMkKMkszMZlNjndSWAg4vcdpjwLMQvDH8Tx6tQ7FXU4NrZLcK5JkBRd7BjnccFBBKqhznPb6gJQCA/lqvVxQvgayCEiSFVwsAs2DHGw46OL1AU56pfD+TC0uLtCrtSwC1ZFFQJKsILUolYa+DfH39Fc6SmUZu8G3AYS0VjqJIvaf1dIxIpgQfwebseVAZBGQJCtw2JlBp3dD1O3g4qJ0Ersr0Bg4nieHgq5GFgFJuklmYbY0jnO0nkElWVCY7rQnhXeeyEOAnBp6FbIISNJNytZkU2Yqc7yZQRXnA5yzCPyWnEuQtyu3NHHO1tnXymZNzw8fPszbb7/N6tWrK92+ZcsWPvvsM9zc3GjdujWzZ8/G1dWVwYMH4+9vGU+NiIhg3rx5toomSVZ1JP8IAO0btFc4yWUydoOnPzS6RekkdqfRm9h2LIe7m/vh6up8Q2HXwyZF4MMPP2Tz5s34+PhUul2n0/Hee+/x/fff4+Pjw+TJk/n999+58847EUJcUTAkqTY4nHsYbzdv2tRvo3SUyk7vgchu4OaAC9zY2NZjOZQZy+nd3MFO1DsgmwwHRUVFsXjx4itu9/T0ZO3atRXFwWQy4eXlRXJyMmVlZYwZM4bRo0eTkJBgi1iSZBMJuQnEhMTg4eqhdJR/aAssPYOc9PqAbxPO0iTYhw6NZMO4q7HJnwj9+vUjMzPzittdXV0JCQkBYPXq1Wi1Wnr06MGJEycYO3YsQ4cOJT09nXHjxvHzzz/j7n5lPJVKdUOZdDrdDT/X3mpLVpnTspCMqkDFQ2EPWWUb1srqf3YnkUA6EZTZ4L078s++qKycnSfyeLRDMAa93mFzXkrJ/Wn340Sz2cxbb71FWloaixcvxsXFhebNm9O0adOKfwcHB5OXl0d4ePgVz2/Xrt0NbVelUt3wc+2ttmSVOWF/9n7KRTl92/elXcTNb8NqWU9/AW6eNLt9MHhY/69hR/7Zf7orDbOAsX07Yi7MdNicl7L1/oyPj6/2PrvPDvrvf/+LXq9n2bJlFcNCGzduZP78+QDk5OSgVqsJDZVzeyXHl5CbAMCtobcqG+Ryp/dAk842KQCO7tuELNqGBdAmLEDpKLWCXY4Evv/+e7RaLTExMWzcuJHOnTvz5JNPAjB69GgeffRRXnnlFUaOHImLiwtvvPFGlUNBkuRoEvISaBHUgiAvB5qGqFdDVgLcOVHpJHaXnq8h4UwRr9zfVukotYbNPmkjIiJYv349AAMHDqy4PTk5ucrHL1y40FZRJMkmzMLM4bzD9Inqo3SUyjL3gyh3yusDvkvIwsUFHoptrHSUWkNeLCZJNyi9OJ1ifTGxobFKR6ksYze4uEJEV6WT2JUQgm8TztKteX3Cg3yu/gQJkEVAkm5YQl4CALENYxXNcYXTeyCsI3gHKp3Ero5kFpOWr+Hh2CZKR6lVZBGQpBt0KPcQwV7BNAtspnSUf5j0luEgJxwK+jbhLJ5urtx/y5WzCqXqySIgSTcoITeB2NBYXBypQ2dWAph0TlcETOVmvj98jnvaNiTIx4Eu2qsFrnpi2GAwEB8fT2FhIWFhYcTGxuLqKmuH5NwKdYWkl6QzqOUgpaNUlrHL8jXKua4U3pVynny1nodvkyeEr1eNn+YqlYpBgwaxadMmDh8+zKpVqxgwYACnTp2yVz5JckiH8w4DON5J4dN7IKQN+IUoncSuvjt0lgBvd+6WbaOvW41HAm+//TZLly6lRYt/WuSePHmSBQsW8OGHH9o8nCQ5qoTcBNxd3IkJiVE6yj/M5XB6L8Q8onQSuyozlPNLUjYDb22Mt4eb0nFqnRqPBHQ6XaUCANCqVSuMRqNNQ0mSozuUe4h2Ddrh7e5AV+TmJIG+xOnWE96qykFjKGeQnBV0Q2osAm5uVVdVs9lskzCSVBsYy40knU9yvFYRp7ZZvjpZEfju0FnCg7zp1ry+0lFqpRqHg3Jycli3bl2l24QQ5Obm2jSUJDmy5IJk9OV6bmt4m9JRKkveAo1vgyDn+Yu4QGNgx4k8xvZsLhePuUE1FoGBAweSl5fH2bNnOXv2LI0bNyYiIoIBAwbYK58kOZxDuYcAB7tIrCQLzsbDPbOUTmJXPxzJwmQW8gKxm1DjcNDTTz9NYmIiqamphIaGcurUKY4cOcJTTz1lp3iS5HgS8hJo4t+Ehr4ONBMl+QfL13YPKZvDzr5NyKJNowDahTvX1dHWVOORwMKFC+nfvz8PP/xwxW0bNmzgzTff5PXXX7d1NklyOEIIEnIT6BLWRekolam+h5DWENpa6SR2c/q8lviMQl7u72DLetYyNR4JJCcnVyoAAEOHDuX48eO2zCRJDitLk0VeWZ5jDQVpCyD9L2jrXMO03yWcBZCzgm5SjUWgup7+1c0akqS67uIiMg51UvjEz5bW0e2cpwhc7BjatXl9mgTLjqE3o8YiEBwczNGjRyvddvToUYKCHGgBDUmyo0O5h/B196VlcEulo/xDtQUCm0DjOKWT2M2hM0Wk5MmOodZQ4zmBl19+meeee45u3boRGRlJZmYme/bsYfny5Vd94cOHD/P222+zevXqSrdv376dpUuX4u7uzpAhQxg2bBg6nY6pU6dy/vx5/Pz8WLBgAfXryzm/kuM5nHeYjqEdcXd1kJXvDBpI+Q3iRoMjNbKzsVW70gnwdmeQXDzmptV4JBAREcHGjRvp0qULRqORjh07sn79eiIjI2t80Q8//JCZM2ei1+sr3W40Gpk3bx6ffPIJq1evZt26deTn5/PVV1/RunVr1qxZw8MPP8yyZctu/p1JkpVpjBpOFJ5wrPMBp36zdA11ovMB54rL+PHoOUZ0icTPy0GKcS121T3o5eVFv379rutFo6KiWLx4MS+//HKl21NSUoiKiqoYTurUqRP79+8nPj6ef/3rXwD06tXL6kXg888/Z8mSJfj6+lr1dW1Fq9XWiqzOltMQbsDc38yXb37JhqwNVkh2pevNOqNtOt0auPHI2P9SLux3JKDkz74wsifljbvy46LpbH2rpMbH1qXf0TFjxjB69Girb9smZbRfv35kZmZecbtarSYgIKDiez8/P9RqdaXb/fz8KC0trfa1VSrVdefJysrCbDaj1Wqv+7lKqC1ZnS2nsZ4RBBhPGzEZTFZIdqXryermIujeoJidOf6Uaspskqc6Sv3shas7JQ074pWXjLEwm6t1MatLv6NZWVk39Pl3NXY9lvL390ej0VR8r9FoCAgIqHS7RqMhMLD6Cz/atWt33dtt164dgwYNuqHnKkGlUtWKrM6W89mtz5JXlsfXf31thVRVu66sKdth9WAGTFnBgLYP2ixTVZT62a/Zd5oZ3xxl9ayxdG0+9aqPd7bf0erEx8dXe59dV4eJjo4mIyODoqIiDAYDBw4c4LbbbiMuLo4dO3YAsHPnTjp16mTPWJJ0VeXmco7kHXGs9QNUW8DDF6LvUTqJXQghWLUrjZgmgXRpVk/pOHWGXY4Evv/+e7RaLcOHD2f69OmMHTsWIQRDhgyhUaNGjBw5kmnTpjFy5Eg8PDxYuHChPWJJ0jVLKU5BbVQ7zvUBZrOlVUTLvuDhHPPk/zqVz8lcNQuH3upYS3rWcjYrAhEREaxfvx6wNKK76J577uGeeyr/5eLj48OiRYtsFUWSbtrFi8Qc5kjg7AFQZ0O7gVd/bB3xyV9phPh7MeBWuZC8NcnFgiXpGiTkJtDAuwERARFKR7FQfQ+u7tDqPqWT2EVKnprfj+cxqntTvNxlxwJrkkVAkq7BodxDxDaMdYxhCCEsawc07wU+wUqnsYvPdqfj6ebK492jlI5S58giIElXkV+WT6Y603GGgnJVUJDqNBeIFWuNbDiQyUOxjQnx91I6Tp0ji4AkXcXh3MOAAy0ik7wFcAE7TwtVyroDpykzlvN0j2ZKR6mTZBGQpKs4mHsQT1dP2jdor3QUC9X3ENEFAsKUTmJzpnIzn+3OoHuL+nRoLBtX2oIsApJUAyEEf5z5g06NOuHp5ql0HCjMgOwjTtM2euuxHM4WlfF0j+ZKR6mzZBGQpBqcLDrJ6dLT9G3aV+koFheXkXSS8wGf7Eojsr4Pfds1UjpKnSWLgCTVYFvGNlxw4Z4oB7kqN3kLNGwPDaKVTmJzRzOL2Z9eyFN3NMfN1QFmZdVRsghIUg22Zmzltoa3EeITonQUUOdBxm6nOQpYtSsNfy93hnV2kGsz6ihZBCSpGunF6ZwqOsW9Te9VOorF8R8B4RRXCeeW6Pj+SBaPdoogwNtD6Th1miwCklSNbae3AdAnqo/CSS5QbYbgKAi7RekkNvf5ngxMZsFTdzRTOkqdJ4uAJFVjW8Y2YhrEEO7vAL1qCjMsq4jdMrTOLyOZW6rjk11pPBATTrMQP6Xj1HmyCEhSFc6pz5F0PslxZgUd+MTy4d95jNJJbG7RbycxmMxM6ddG6ShOQRYBSarCxaEghygCRh0c/BzaPABBdfskaWqemq/+PsPIrlE0l0cBdiGLgCRVYVvGNlrVa0XTwKZKR4Gkr6GsALo+o3QSm3v71+N4u7vyQp9WSkdxGrIISNJl8svyOZR7iHujHGRW0N8fQkgbS9fQOuzQ6UJ+PJrNuF4tCA2QjeLsRRYBSbrM9tPbEQj6NHWAWUGZ8ZB1ELr8q06fEBZCMP+nZEL8PflXzxZKx3EqNllZzGw2M3v2bI4fP46npydz586laVPLYbVKpeKNN96oeGxCQgJLly6lY8eO9OvXj9atWwPQt29fnnzySVvEk6QabcvYRtPAprQKdoAhif0fgqc/3DpC6SQ29fvxXPalFTBnUAf8veyy6q10gU329rZt2zAYDKxbt46EhATmz5/P8uXLAWjXrh2rV68G4KeffqJhw4b06tWL3bt3M2DAAGbNmmWLSJJ0TYr1xezP3s+THZ5UfgEZTT4kfg1xo8A7UNksNlRuFiz46TjNGvgyoqtcNMbebDIcFB8fT8+ePQGIjY0lMTHxisdotVoWL17Mq6++CkBiYiJJSUk88cQTvPDCC+Tm5toimiTV6I8zf2ASJseYFXTwcyjXW4aC6rCvD2ZyPKeUqf3a4uEmR6jtzSZHAmq1Gn9//4rv3dzcMJlMuLv/s7mNGzfSv39/6tevD0CLFi2IiYnhjjvuYPPmzcydO7fKxedVKtUNZdLpdDf8XHurLVnrYs5vTnxDiGcIrrmuqPLs/94qsprLid7zAcaGnTh9HjjvWPvZWj97vcnMgh/P0LqBF83dC1Gpim4+3CXq4u+o1QkbeOONN8QPP/xQ8X3Pnj2veMyjjz4qsrKyKr4vLS0VJpNJCCGEVqsVffr0ueI5Bw4cuOFMx44du+Hn2lttyVrXcqoNahH3eZyYv2++jRNVryKr6gchXgsUIulbxbLU5GLOO+6445qfM2nSJPHII4+IU6dOVdy24o9Toum0LWL3qXyrZxTi5n5HV69eLYQQYseOHWLt2rXVPm7RokVizZo11/Sa1e0vW/+/VNNnp02OveLi4ti5cydgOfF78WTvRaWlpRgMBsLD/7kcf+bMmfzyyy8A7Nmzhw4dOtgimiRV68/MPzGYDY4xFPT3SghoDG3qzhKSu3fvZtOmTURHW9pgF2kNLP39FL3bhHJ7dAOF013p4nnMXr16MXz4cIXT2I5NhoPuvfdedu3axYgRIxBC8MYbb7Bq1SqioqLo06cPaWlpNGnSpNJzXnrpJWbMmMFXX32Fj48Pc+fOtUU0SarW1oytNPBuoOiC8kHffQf/egrOxkNwU/jeCgVpzBgYPbrau41GI6+99hoZGRmYzWYmTpxIt27dGDhwIF27duX48eO4uLiwbNkyfH19mTVrFkePHqV169YYDIYrXm/Xrl289957eHl5ERwczBtvvME777yDWq3mueeeq/hwXfZHCqWFeehSvmLU9sV4eXkxZ84cCgsLmTp1Khs2bOCnn35i586dTJ06lRdffJHQ0FBycnLo1asXkyZNIjMzkxkzZlBeXo6LiwszZ86kbdu23HfffcTFxXHs2DEiIiJYvHgxZrP5mt/nF198QXFxMbNnz6Zjx46kpqYyZcoUFi5cSGJiIkVFRbRt25Z58+ZVuU8XL15Mamoq58+fp6SkhJkzZ9K5c2cMBgMvvfQSWVlZBAcHs2jRIs6fP8/cuXPx9PQkLy+PiRMn0rdvX95991327duHyWTivvvu45lnnuH48eMVn40X921AQMDN/X7Y9BjEyuRwkGOpSznLjGWiyxddxP92/88Oiap3dt48IW6JFKKZuxB33iHEXXfd/H+ffVbjNr/88kvx5ptvCiGEKCgoEA888IAQQojevXuL+Ph4IYQQkydPFlu2bBE//fSTmDx5sjh27Jg4e/as6NChQ6XXMpvNonfv3iI7O1sIIcSnn34q5s+3DK9dOhSSWagVrV79Udz1yJPijz/+EEIIsXv3bjF58mQhhBCrVq0SEyZMEIMHDxalpaXizJkzolu3bqKwsFCYTCYxbNgwkZiYKCZMmCC2bt0qhLD8nAcPHiyEEKJt27YiKytLHDt2TAwfPlwcOnTout7npXk3bdok3nrrLVFaWipWrlwphBCivLxc9O/fX2RnZ1c5HLRo0SIxffp0IYQQJ06cEAMHDhRCCNG+fXtx5swZIYQQTzzxhDh8+LDYtWuXWLdunRBCiPj4ePHUU09V5Dpz5ozQ6/Xiq6++EkIIMXToUHHy5EkhhBDr168X77zzTo0/24tq+uyUE3IlCdidtZsyU5niQ0ElD9xLY7EUWj8JQz6yyzZPnDhBfHw8R44cAcBkMlFQUABA+/btAQgPD0ev15Obm0vHjh0BaNy4caUhXYDCwkL8/f1p1MiyHGSXLl145513rtjmO7+eAMBTncMHH3zARx99hBCiYvLIiBEjWLp0Kf/+97/x9/ev+Ms7ODgYgI4dO5KWlkZKSgpdunQBLNPPs7OzAahXrx7h4eEUFRVVZL+e91kVLy8vCgoKmDx5Mr6+vmi1WoxGY7X7tXv37gC0atWK/Px8AIKCgoiIsPR/CgkJoaysjNDQUD755BP279+Pi4sLJpMJgLfeeouFCxeSn59fMdsyJSWF//3vf4DlCK5Zs2bVbv9aySIgSVguEAv0DKRLWBdFcwRl/AL6Eugyzm7bbNGiBWFhYYwfPx6dTsfy5csrPmwvv1aiZcuW/PDDD3Tt2pWcnBxycnIq3V+vXj3UajW5ubk0bNiQv//++4oPqviMQr4+lMm4ni3IKopmzJgxxMXFkZKSwv79+wF48803GTt2LF9//TV9+/bFxcWFlJQUysrK8PT05MiRIwwZMoTo6GgOHDhAnz59UKlUhISEVJn7et8nWK5ivtTOnTs5d+4c7733HgUFBWzduvWKx1wqKSmJQYMGceLEiYqiWNV23n//fe6++26eeOIJNm3axDfffIPBYODnn3+uKKAPPPAADz74IM2bN2fBggU0btyY+Ph48vLyqt3+tZJFQHJ6xnIjf2T+Qe/I3ni4KriKlRDUO7XRsmhMZFe7bXbEiBHMnDmTJ554ArVazWOPPYara9VzRvr06cOuXbuYOnUq0dHR1KtXr9L9Li4uzJ07lwkTJuDi4kJQUFClcXOdsZypGw/TOMiHF/q0ovCWacyePRu9Xo9Op+PVV19l27ZtpKenM2vWLGJjY5kyZQoLFizAw8ODF198kfz8fPr370/btm15+eWXmTVrFp988gkmk4n/+7//s8r7BIiOjmbKlCnccccdgOXoY9myZTz++OO4uLgQGRlZ4/VMKpWKJ598krKyMubMmVPt4/r37897773HTz/9RFhYGIWFhXh6ehIUFMSwYcPw9vamR48eNG7cmNmzZzNt2jRMJhMuLi41vt9rdk0DSg5CnhNwLHUl51+Zf4mYT2PE76d/t0+g6qT9ZZkWGl/zGL4juNGf/Rs/HBNNp20Rf57Iu67nnTlzRgwdOvS6t6fU7+j1TBsVog5OEZWk2mRrxlZ83X25vfHtygbZ/yHlnoEQ86iyOWzk0OlCPvwzlZFdI7mzVYjScaQL5HCQ5NRMZhO/n/mdXhG98HJTsH1xSRaovqeo5VAaePoql8NGLMNARwgL9GbGA+2u+/kRERGsX7/eBslsY8KECUpHuGbySEByalsztlKgK+DBFgpflPXHfMCFwlZDlc1hI+//dpJTuWrmDelIgLeC512kK8giIDktIQSrElfRPKg5vSIUXLAlVwWHVkPXcRj9GyuXw0YOnynigx0pDOscwV2tQ5WOI11GFgHJaf2d/TeqAhVPtn8SVxcF/1fY+l/wDIBeU5XLYCN6UzlTNhymYYA3rz7YXuk4UhVkEZCc1qqkVTTwbsCA6AHKhUjdASd/hV4vgW995XLYyOLfTnEyV828R24hyEcOAzkiWQQkp3S84Di7zu7i8XaPK3dC2GyGX2dCUBR0fVaZDIBer2fDhg1Wf9341FxWfPolj3aKoHfbhlZ/fck6ZBGQnNJnSZ/h4+7DsDbDlAtxdANkH4E+/wUPb8Vi5OXlWb0IGExmpq7+C8/T+5glh4EcmpwiKjmdbE02P6X9xIi2IwjyClImhLEMfnsdwmMhZkjFzd999x3PPfecVTc1ZswYRtfQRXTFihWcOnWKJUuWkJiYiF6vr9TNcsCAATRr1gwPDw9mzZrFlClTKCoqokOHDuzdu5etW7fy999/8+677+Lm5kZkZCT1ez3B6V3f4qfJZfUnH/D8889b9T1J1iOLgOR0vjj2BQLBqPajlAuxbwWUZMLgFVBD6wJ7GD9+PCdOnCAuLo4uXbrQrVs3Dh48yOLFi+nbty9arZZ///vftG/fnjfeeIM+ffoQFxdHQUEBu3btQgjBrFmzWLNmDQ0aNODVOQv44PO19H3kCQq2fyQLgIOTRUByKqWGUjae3Ei/Zv1orNR0TM15+PMdaN0fmvesdNegQYOYPn26IrFCQ0NZvnw5GzdurNTNEqB58+aApYvl4MGDAejcuTMABQUF5ObmMnHiRIzlZo5m5OMb3pYX+rZi9nb7vw/p+shzApJT2XBiAxqjhqdjnlYuxI4FYNDAva8rl+ESrq6umM1m3n//fQYNGsRbb71Ft27dKnXIvNhorXXr1hw6dAiwrBoIls6hYWFhvL94CeKu5zG26cvLowcS7OuF2Wy2+/uRro88EpCchrHcyJfHvqR7eHfa1m+rTIjzKXDgY4gbDaFtlMlwmQYNGmA0Gjl58iRvvvkmK1eurOhmeblx48bx8ssvs2nTJpo1a4a7uzuurq68+uqrDBg2mrxSHW0jQ7m/xxj8/f0xGo289dZbTJ1a966BqCtsUgTMZjOzZ8/m+PHjeHp6MnfuXJo2bVpx/9y5czl48CB+fn4ALFu2DKPRyJQpU9DpdDRs2JB58+bh4+Nji3iSk/oh7Qdyy3KZ06P6tr42t202uHnB3a8ol+EyXl5efPfdd9Xev337P2M6R48e5YUXXsDDw4PCwsKKfvZp7pFkxT3D+LuimX7/PwW2pteVHINNisC2bdswGAysW7eOhIQE5s+fX7GuKFgWW/joo4+oX/+fi2Pmzp3LgAEDeOSRR1i5ciXr1q3jqaeeskU8yQkJIfgs6TPa1GujXLfQ0/tAtRnungEBjZTJcJMiIiKYMWMGer0eb29vXn31VXan5DN7cxJ92jZkaj/HOLqRrp1NzgnEx8dXLIcWGxtLYmJixX1ms5mMjAz++9//MmLECDZu3HjFc3r16sXu3bttEU1yUn+e/ZNTRad4ssOTVa7uZHNCWC4M8w+DO2rvbJno6GjWrVvHvHnzWLt2LYGNW/DvLw/SIsSP90bE4uaqwL6VbopNjgTUajX+/v4V37u5uWEymXB3d0er1fLEE0/w9NNPU15ezujRo4mJiUGtVhMQEACAn58fpaWlVb62SqW6oUw6ne6Gn2tvtSVrbcq5TLWMBp4NaKZrpkjmgDPbicj8m6wuMyhOOV3t42rTPj1wOInJP57FZCpneo96ZKadUjrWFWrT/lQqp02KgL+/PxqNpuJ7s9lcsYC0j48Po0ePrhjv7969O8nJyRXP8fb2RqPREBgYWOVrt2t3/b3IwVI8bvS59lZbstaWnN/v/56k0iSmdJ7CLR1usX8ATT78sAgadqDx/VNo7OpW7UNryz5NTDrG0gMazpaaWD2mK3e0dMxFYmrL/rR1zvj4+Grvs8lwUFxcHDt37gQs08hat25dcV96ejojR46kvLwco9HIwYMH6dChA3FxcezYsQOwLOjcqVMnW0STnNDm7M0EeATwaGsFVuwSAr79N5QVXLgwrPoCUJusTihkmyqX1wa2d9gCIF0bmxwJ3HvvvezatYsRI0YghOCNN95g1apVREVF0adPHwYNGsSwYcPw8PBg0KBBtGrViueee45p06axfv166tWrx8KFC20RTXIyZ0rPsK9gH0/HPI2fh5/9A+xdDid/gfvfgvCO9t++DXyXcJZ1R4t4rFsUo7o3vfoTJIdmkyLg6urK669XvhAmOjq64t//+te/+Ne//lXp/pCQED7++GNbxJGc2NKEpbi5uPF4u8ftv/GsQ5a1Ato8CF3H2X/712jnzp2cO3eO4cOHo9FoePfdd1GpVLi4uODv78+0adNo3rw5+/bt4z8TXqDEowHBHi6cSvPnJ68neeCBB6p97atNFwdYv349a9euxd3dneeee47evXtTUFBQ5ZTxTz/9lB9++AGAu+66i+effx4hBL169aJZs2aAZTLKSy+9VGWeoqIi/vzzTwYOHHjFPvjxxx+ZP39+te+lqpyXOnDgAAsWLMDFxYUuXbrUnmsjbLrEvZUdOHDghp977NgxKyaxrdqS1dFz7jyzU8R8GiNm/zrb/hvXlQjxfqwQC9sJoTl/zU9Tep9OmjRJfP755xXfq1Qqcf/994uSkhKxfP1PotU9w8WARX+KfYeOCrVaLQYPHlxj5l9++UVMmzZNCCHEoUOHxPjx4yvdn5ubKwYMGCD0er0oKSmp+PecOXPEpk2bhBBCfPDBB2LVqlXi9OnTYvDgwcJkMgmz2SyGDx8uVCqVSE9PF88++2yV27882969e8XEiROveNyOHTsqclalupyXGjx4sDh9+rQQQognnnhCJCUlVft6V8tpbTV9dsorhqU6SWPUMGfvHFoEtWBw48H2D/DDFChMhye3XNdiMdtSSpm9c49VowzrHMmQThHV3v/111+TmprKmDFjOHHiBO+8807FfW3btqV379689+lGPj9SSpCPO1+M7UZWxin8/PwYPnw4P//8MxqNhvfff7/S6z711FM1ThcHOHLkCLfddhuenp54enoSFRVFcnIy8fHxPPusZY2FXr168c477/D444/z0Ucf4eZmOa9iMpnw8vIiKSmJnJwcRo0ahbe3N6+88gotWrSo8r2uWLGC5ORk1q1bR+fOnZkxYwY+Pj74+PgQFBSERqNh/PjxlZ7TrVs32rVrV2XOjh3/GeJbv3497u7uaDQa1Go1vr6+V/vROARZBKQ6adHBRWRrsvn8/s/xOG/nFa0Or4Uja+Gu6dCsh323fRMyMzOJjIy84na1exBf/XaIpm070tanAUG+HmRduK9BgwYkJSXRuXNnVq9efcVzt2/fXu10caDS1HCwTA9Xq9VVThn38PCgfv36CCF48803ad++Pc2bNyc/P59nnnmG+++/nwMHDjB16lQ2bdpU5XscP348a9euZfjw4Tz77LO88MIL9OjRg5UrV5Kamoqfn1+V7+O7776rMuel3N3dSUhIYPLkyURHRxMWFlbNnnYssghIdU5CbgJfJX/FiLYjiG0Yi+q8Hedfn0+BLZMh6o4bWjO4b3QAEwYoM6WxYcOGZGVlVbrt16RsvvotnvCo5sx4oB3ff5NU6f6srCzCwsI4cOBAlUcCNU0Xhyunk2s0GgICAqqdMq7X65kxYwZ+fn689tprAMTExFQcHXTu3Jnc3FyEEFe9KDA9Pb3iL/m4uDhSU1NrPBKoKuflYmNj2b59O++++y4rV67khRdeqDGDI5BFQKpTDOUGZu+eTSO/RrwY96J9N27Sw8anwd0ThnwIbrXrf6+wsDCioqL48ssvefzxx/k5MZvnl36HT56KNZ+8ccXFYGq1mg0bNvD+++/TokWLKv+CNplM/P777zzwwANXTBcH6NixI++99x56vR6DwUBKSgqtW7eumDL+yCOPVEwZF0Lw73//m27duvHMM89UvMaSJUsIDg5m3LhxJCcnEx4eXm0BuNgxFSyTVQ4dOkSvXr0qhqmqOxLIy8urMudFQggef/xxli9fTlBQEH5+fhgMhmvc88qqXb+lknQVHx39iJTiFJb1WWb/KaHbZsO5wzBiDQRVPwbvyBYsWMCbb77JfQMeJv18GUGBgXyyaiVNGjYgM+0Ue/fuZdSoUZSVleHt7c2ECROqHX+HqqeLA5WmjI8aNYrHHnsMIQSTJk3Cy8uryinj27Zt4++//8ZgMPDnn38CMHnyZJ555hmmTp3Kjh07cHNzY968eQD83//9H7GxsZUuwoqKiuLEiRN8+umnTJ8+nWnTpvHxxx9Tv359vLyqX2s6NDS0ypx79uwhPj6e559/njFjxjBu3Dg8PT0JDQ1l7ty51viR2J5NT0lbmZwd5FgcLefJgpMi9vNY8fKOlyvdbpecx38W4rVAIX6YclMv4wj79IcjWaLFKz+IwUv/EiVlhiof4wg5r+bzzz8Xv/32m9IxromcHSRJN6ncXM5ru1/D38OfaV2n2XfjBanw7XPQ6Ba4V8E21Vaw/sAZXvn6KLGRwXz6dBcCvO18Ut2K+vTpQ3FxsdIxHJ4sAlKdsPb4Wo7kH+GNO9+gvve1T8m8aUVn4LOHLO0hhq4CD2/7bduKjOVm/u8HFZ/uTqdHywZ8MKoz/l61++OhcePGsghcg9r9U5YkIEudxfsH3+fOJncyoMUA+2245Bx8NhB0JfDkZghpZb9tW9F5tZ7/rDnI3tQCxt7ZnFfub4u7m1x51lnIIiDVakIIXt9jaVEyq/ss+60VoMmHzweBJg9GfQuNY+2zXStLyirmmc/jyVPrWTj01hovKpPqJlkEpFptS+oWdmXtYnrX6TT2b2yfjZYVwucPQ1EGPLEJIrvYZ7tWtvlwFi9vPEw9X082jr+djhHBSkeSFCCLgFRrpRanMv/v+XQM7ciINiPss1FdCXwxBPKPw8ivoNmd9tmuFZWbBW/+kswHO1Lp0qweyx7vRGhA9dMjpbpNFgGpVsrWZPPs1mfxcPVgfs/5uNmjT79BA2uGW64FGLYaWva1/TatrFhrZMLaQ+w8kcfj3aJ4bWAH9u7+i+3X2EV04sSJtGzZEo1Gg6enJ6NHj5ZdRC+zYsUKjh8/zrvvvlvtazkSWQSkWqdYX8z4reMpNZSyqt8qIgOu7HdjdUYdrH0MzuyFIR9B2+o/+BzVrlP5TNt0hJwSHW8MvoXHukUBlgZtF82aNYvbbruNmTNnApCcnMx//vMf1q1bB1hWArxYJKKiohg1ahTNmzevdlWsbdu2YTAYWLduHQkJCcyfP5/ly5dX3J+Xl8fq1avZtGkTer2exx57jB49erBs2TIGDBjAI488wsqVK1m3bh19+vRh8+bNbNiwAVdXV0aOHEnfvn3x8fGhQ4cOrFix4qr74Pjx42zfvv2KInA11eX09PSs9LgdO3bwxx9/EB4efl2vryRZBKRapcxUxvO/Pc/p0tOs6LuCdg3s0GfHZID1oyH1D3h4OcQMsdmmgtJ+hL1TrPqaZTEjef1MR776+wzNGviy9pnb6dS0XsX919JF9NdffyUiovJJY9lFtHIX0YyMDNatW8cLL7zAhg0bruln4whsUgSudgh4s4d0knMymo1M2TGFw3mHWXj3QrqGd7X9RssKYeNYSPkNHnwHYh+z/TatqFBrYNHPyazT1uPZXi2YdG9rvD2qHjqrrotoZGQkWVlZVxQBkF1EL9JoNLz++ussWLCAlJSUKrftqGxSBGo6BDxz5sxNH9JJzkcIwezds9mZuZNZ3Wdxb9N7bb/R7ERY9zgUn4WBi6DTkzbfZHHzB2j8wM3/8VOoMfD6lmN8c/wsrRv58/VTtxIbGVzjc6rqIgqWv3AvXRnwUrKLqMWuXbvIy8tj0qRJlJSUkJuby8qVKys1unNUNikCNR0ChoWF3fQhneR83o1/l80pm/l37L8Z1maY7Td4dCNsngBegfD0jxBph6MOK/nx6Dn++10iRVojL/RpxX96R+PlfvUT55d3EQVISkpi+/btPPfcc6hUlVtyyy6i/7yX++67j/vuuw+Affv2sXbt2lpRAMBGRUCtVld7CGiNQzrJuXyW9BmrklYxos0Ixnccf/Un3IxyE2x7DfYsgcjuMOwzCKgdi4NkFZUxZ8sxfkrMJqZJIJ+P6Ub7xoHX9RoXu4gOHToUNzc3AgMDWbZsWcVf4rKLaPVdRGsrFyGEsPaLzps3j1tvvbVi6livXr3YuXNnxf2XH9K5ublRVlaGm5tbxdn2nj17snPnzkoVPT4+/oaXbNPpdHh7146+LrUlqz1y7sjfwdLUpXSv352J0RNxdbn+dgbXmtNNX0ST3TPxyz1AQctHyYl9Edzs20DtRvZpqb6cdUeL2KwqAeCJ2HoM6RCEm6vtrp6uDb+jW7ZsISYmpuI8oyOz9f7UarV06tSpyvtsciQQFxdX7SHgzR7SVTcV7WpUKtUNP9feaktWW+fckrqFFWkr6BbWjaV9l+Lp5nn1J1XhmnJmJcC6caDOhUHLqH/b49ixDV2F69mnZYZyVu1OY/kfp1HrTTxyWwST7m1FRD3br21bG35Hg4KCKC4udvicYPv9GR8fX+19NikCVR0CXjz8M5vN13VIJzkfQ7mBt/a/xdrja4lrGMd7vd+74QJwVULA4a9gyyTwDYExP0OTONtsy0pM5WY2xmfy7rYT5JTouadtQ17u34a2Ydc39FPXyS6i18YmRcDV1ZXXX3+90m2Xzi44evRolc9buXKlLeJItcg59Tle2vESR/OP8mT7J3mx04t4uNpoSCb/JPz0MqRsh2Y9Yein4Bdim21ZgRCCX5JyeOuXZFLyNNwWFcyiEbfRrUUDpaNJtZi8WExyGLvP7mban9Mwmo28c/c7tpsGqlfDzrdgz1Lw8IH+C6DLvxx2TWCDycwPR7P45K90jp4tJjrUjw9GdeK+9o3s1zVVqrMc87decipmYeaDIx+wPGE50cHRvHv3uzQLamb9DQkBSd/ArzOh5CzEPg59Z4N/Q+tvywrySvV8uS+DL/edJq9UT3SoHwuG3MKQuAjZ71+yGlkEJEUV6YqY/td0dp3dxcAWA5nZfSa+HjY4sZmbDD9NhbSdENYRHl0FUd2svx0rSDxbzCe70thy+ByGcjO924TyVI/m9GwZgqsNZ/xIzkkWAUkxR/OO8tKOl8gvy2dW91kMbT3U+sMbZUU0PPQ+nNoAnv7w4ELo9DTYo+voddCbytl2LJdl27JIyk3Fz9ONkV0jefKOZrQI9b/6C9yEnTt3ck52Eb3pLqJbt25lwYIFFc3jJkyYQNeujn+RoSwCkt1la7JZcmgJm1M2E+4Xzur7V9MhpIN1N1KcCXuXQ/xn1DeoIW409HkN/BznJGq5WbAv7TybE7L48eg5SnQmwvzdmTWgPUM7RxBop0XeZRfRq7uWLqKJiYlMnTqVfv36XddrK00WAcluSg2lfJL4CauPrcYszIxuP5pxHccR5BVkvY1kH4XdiyFxk+UcQMwjpIUPpMUdg6y3jZsghOBIZjGbD2fx/eEsckv1+Hm60a9DGA/FNuZY9ib+Uq/lrz+st83BrQbzUPRD1d4vu4hap4toUlISKpWKzz77jI4dOzJlypRKfZIcleMnlGo9o9nIhuMbWHF4BYX6Qh5o/gATbptARICV1rMVAlJ/h12LLF89/aHrs9D9OQiORH9Zzxt7E0JwPKeUH49msznhLOnntXi6uXJ3m1AGxTbhnrYN8fG0fLAl5yo35i+7iN54F1GAHj160LdvXyIiInjttddYu3YtTzzxRJU5HIksApLNCCHYdnob7x98n4ySDLqEdeGlTi9Zb+hHVwLJW2DPMsg5Cv6NLEM+nZ8Gn3pXf74NFWuN/HUqnx0nctlxIo+cEj0uLnB7iwY8d3c0/TuEE+R75XDPXSF3Mb6njfsjVUN2Eb3xLqIAQ4YMqcjZp08ffvnllxq37yhkEZCszlhuZEfmDj5L+oyEvASig6JZ2mcpPZv0vPkTvwYNnPgZEr+Gk1uhXA8hbeChJdBxGLgrs1ZuuVlw9GwxO47nseNELglnijALCPR2p2erUHq1DuHuNg1pFOi4/XZkF9Eb7yIqhOChhx5i7dq1hIWFsWfPHjp0sPJ5LhuRRUCympSiFL4++TVbUrdQoCugkW8jXrv9NR5u+TDurjfxq2Ysg5O/Wj74T/wCpjLwD7P8xd/hEYjoAq72nTevN5WTeLaEA+kF7E8v5EBGAUVaIy4u0DEimOd7t+SuNqHcGhFcq+b0yy6iVbuWLqJz587l+eefx9vbm+joaIYNs0PLcyuwSRdRW4mPj6+2E97V1IaGVxfVlqwqlYrI6Eh+Tv+Zb059w5G8I7i7uHN35N0MbjWYHo173PgC8CXnIP1Py4f+8Z/AqAG/UGg/CDoMhqjbr3mapzX2Z3GZkYMZhexPL+BAeiEJmUUYTJa/KFuE+NGpaT3ubBVCz1ah1Pe78T5Hteln7+g5V69eTZMmTbjnnnuUjnJV9mggZ9cuolLdpi/XczDnIGtS17Dv4D7KTGVEB0UzpfMUBrQYQAOfG5iGqTlv+dBP22n57/xJy+0+9eGWRyHmEWh6p81bOwghyCvVk3SuBNW5Eo5llXDsXAlp+RqEAHdXFzo0CWJ096Z0blafTk3rERqgzBCUVLM+ffrIBnLXQBYB6arMwsyJwhPsydrD3nN7ic+JR1+ux9vVmwejH2Rwq8F0DOl47eP9Qljm8Z9LgIzdlg/9nAtTBz39oekdlqUcm/eCRjE2u7BLrTeRlqchJU+NKtvyga86V0K+2lDxmIh6PrQPD+Th2CZ0blaP2MhgfD3l/za1gewiem3kb7NUpWxNNnuy9rDn3B72ndtHga4AgOigaIa2HsrtjW8noCiA22Juq/mFhIDiM5Z+/ecSLnw9DNp8y/3u3hDZDe6ZBc3vgsaxVl3IxVRuJqvESHZyLil5alLzNaTmqUnN05Bbqq94nKebK60a+dO7TUPaNw6kfXggbcMDCfKx76IykmRvsgg4OSEEudpcjp0/xrGCYxw7fwzVeRV5ZXkANPBuwO2Nb+f28NvpHt6dRn6NKp6rKlVd+kJQes7Snvn8STifAnnJFz7wz1se4+IGDdtB6/6WD/vwWAi7BTxufMaM2SzIU+s5U6Als7CMMwVazhRqOVNQxplCLeeKdZSbBXAGgCAfD1qE+tGzVSgtQv2IDvWjRag/zRr44elee07gSpK1yCLgREoNpZwpPcPp0tOcLDxZ8YF/Xmf5kHZ1caV5YHO6h3enQ0gHuoR1oVVwq3+GecqNUJhhGcopziTkxD5IKv7nQ99wycUzHr7QoCW0eQDCb4XGt0GjDpbWzddACIHWUM55tYHsEh3ZJTpyinVX/Du3RI+h3FzpuaEBXkTW8yEuqh6R9X3w0BfTo2MrWoT4Ud/PU7ZflqRLyCJQhxjKDZwvO09uWS6ZpZmcLj3NmRLLh/6Z0jMVQzpg+cBvEdSCHk160L5eWzr4NaG1RxC+erXlL/eSPDizuuIDn+JMy1/6/DOZLAQXCI6EBq0ss3UatISQVpavAY3B1bXiw7xUZ6K0wEiJrpBSndHyvc5EodZAgcby33mNgQKNngK15d96k/mK9+jt4UpYoDeNAr3p1LQeYYHeNKnnQ2Q9XyLr+9Ak2Lfi6tuLVCoV7ZopsVikJDk+mxSBq3UOvJ6ugc5MZ9JRrC+m2FBMsb6YEn0JxYZiCnQFnC87z/my8+Tr8skvs/xXaiit9HwXXGjkEUCURyC9PRoQ5d6IKDNEGoxE6dT4nD4NqoOgK6py+8LNE5N/Ewz+jSlrdAeaZmGUeIZR7NmIAreGHM0Hz3phlg/0EhOluUZKdCZKdWmU6k5QqjOh1psuDMdUz9fTjfp+njTw8yTU34s2jQKp7+dBfT8vGvh7EhboTViQ5YM/0Ntd/iUvSVZkkyJQU+fA6+ka+NRTT9ki3k0TQmASJozlRoxmI8ZyIyaTDqNRi86oRm/QoDNq0Bs16IxatAYNOpMWnbEMnUmHxqhFaypDa9JSZtKhLddTZtZTVq6nzGyg2FSG+m8jBq78S/giHwENyqFBeTktysvpZjQQWm4ipLyckPJyIkwmmphMeF3y+atz8UbjGoDGNYAUlwCKCKeANuR7BJBvDiC3PICccj9yTP6cF4EUEIDQuEJOVQksl9C7uaYS4O1u+c/LgwBvd5oE+xDoHXDhdo/Lvlr+HXjha7CvB94ejtXWWZKciU2KQE2dA6+na6C1ioAqNZ7/2/EMuj/LMQNmF8ughuXfwvL1iu8FwqXy7eWAyQVMVvhL1Mdsxtcs8BX/fK1nFjQWgkCzmaByM77l4Gt2xdvshle5G57lHniY3fEo98Jk9kGLF2rhgxZvNHhRILw5izd6V280boFoXQPQuAegdQtE7xYI7l54uLvg4eaKp5sr3h5ueHtc+Oruhp+HK+093Ljtwu1+nu74errh53XZV093fL3cOJuWQuwt7eVf5pJUi9mkCKjV6mo7B15P18CqxMfH31CmSV3q3iL2wVZ7pYulrhqGC/+pQYflvwvzfTh48KDVUtjSjf7eKKG2ZJU5rUupnDYpAjV1DrzeroGXutGWEZIkSVLVbDIxOi4ujp07dwJc0TmwY8eOxMfHo9frKS0tvaJrIFDRNVCSJEmyLZs0kLs4O+jEiRMVnQN37txZ0TVw/fr1rFu3DiEEzz77LP369SM/P59p06ah0Wgqugb6+tpgwXFJkiSpQq3qInojrmWha0cxePDginMpERERFS1xHcnhw4d5++23Wb16NRkZGUyfPh0XFxdatWrFa6+9hqudWzpX59Kcx44d49lnn61YiHzkyJE1Lo5uD0ajkRkzZnD27FkMBgPPPfccLVu2dMj9WVXW8PBwh9un5eXlzJw5k7S0NFxcXPjf//6Hl5eXw+3TqnKaTCbl9qeo43755Rcxbdo0IYQQhw4dEuPHj1c4UdV0Op0YNGiQ0jFqtHLlSjFgwAAxdOhQIYQQzz77rNi7d68QQohZs2aJX3/9Vcl4FS7PuX79evHxxx8rnKqyjRs3irlz5wohhCgsLBR33XWXw+7PqrI64j7dunWrmD59uhBCiL1794rx48c75D6tKqeS+1P5PzNs7GoLXTuK5ORkysrKGDNmDKNHjyYhIUHpSFeIiopi8eLFFd8nJSXRtWtXwDKtd/fu3UpFq+TynImJifzxxx88/vjjzJgx44q1YZXQv39/XnzxRcBy3Ymbm5vD7s+qsjriPu3bty9z5swBLMteBgYGOuQ+rSqnkvuzzheB6qarOhpvb2/Gjh3Lxx9/zP/+9z+mTJnicDn79etXaX1Ycck6rjVN67W3y3N27NiRl19+mS+//JLIyEiWLl2qYDoLPz8//P39UavVvPDCC0ycONFh92dVWR1xnwK4u7szbdo05syZw8CBAx12n16eU8n9WeeLwNUWunYUzZs356GHHsLFxYXmzZsTHBxMXl6e0rFqdOnYanXTeh3BvffeS0xMTMW/jx07pnAii3PnzjF69GgGDRrEwIEDHXp/Xp7VUfcpWJbI/OWXX5g1axZ6/T/twh1tn16a884771Rsf9b5IlDTdFVHsnHjRubPnw9ATk4OarWa0NBQhVPVrH379uzbtw+wTOvt3LmzwomqNnbsWI4cOQLgMAuA5+fnM2bMGKZOncqjjz4KOO7+rCqrI+7Tb7/9lg8++AAAHx8fXFxciImJcbh9WlXO559/XrH96TSzgy6drhodHa10rCsYDAZeeeUVsrKycHFxYcqUKcTFxSkd6wqZmZlMnjyZ9evXk5aWxqxZszAajbRo0YK5c+fi5uYYfYAuzZmUlMScOXPw8PAgJCSEOXPmVBoiVMLcuXP56aefKi3S/uqrrzJ37lyH259VZZ04cSJvvfWWQ+1TrVbLK6+8Qn5+PiaTiXHjxhEdHe1wv6NV5QwPD1fsd7TOFwFJkiSpenV+OEiSJEmqniwCkiRJTkwWAUmSJCcmi4AkSZITk0VAkiTJickiIDkdvV7Phg0bAPj666/57bffrL6Ne+65p9KFShcVFxezZs2aGp+bkZHBwIEDrZ5Jkqoii4DkdPLy8iqKwCOPPEKfPn3stu1PP/2UNWvWkJSUVOX93377LZMmTaKgoMBumSTn5nj9EyTJxlasWMGpU6dYsmQJQghCQkJo0aIFK1euxMPDg+zsbEaMGMHevXtJTk5m9OjRPPbYY/z999+8++67uLm5ERkZyeuvv46Hh0e125k9ezaZmZkALFmyBFdXV/bs2cP777/PO++8U2V/mKCgIL744gvuvfdem71/SbqUPBKQnM748eNp2bIlzz//fKXbs7OzWbx4MbNnz2b58uW8+eabfPjhhxULIM2aNYslS5bwxRdf0KhRI7755psatzNkyBBWr15NkyZN2LVrF59//jmPPPII0dHRuLi4VNkfpnfv3nIxJcmu5JGAJF3QqlUrPDw8CAgIICoqCk9PT4KCgtDr9RQUFJCbm8vEiRMB0Ol03HHHHTW+3sWGYCEhIeh0Ov7zn/9U3LdkyRIAnn32WbRaLa1bt2bWrFm2eWOSVANZBCSn4+rqitlsvuL2iy2Hq1KvXj3CwsJYtmwZAQEB/Pbbb1f9i72m17voYiMxSVKKHA6SnE6DBg0wGo289dZb1/wcV1dXXn31VZ555hlGjBjBmjVraN26NXl5eUyaNMmGaSXJtmQDOUm6CSaTibfffpvp06crHUWSbogsApJ0E4xGI0VFRQ6/9oMkVUcWAUmSJCcmzwlIkiQ5MVkEJEmSnJgsApIkSU5MFgFJkiQnJouAJEmSE5NFQJIkyYn9P2XrZIPjgh2UAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hours to OD = 1: \t22 hours\n",
      "Time of inoculation: \t13(the day before)\n",
      "\n",
      "If you innoculate now, the cells will have reached OD= 1 by:   2022-05-31 13:42:56.825817\n"
     ]
    }
   ],
   "source": [
    "time_to_inculate(initialOD = 0.0025, td=0.4, transformation_time = 12, verbose= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a7a36fcb-a13f-421f-9a6c-c3208911e44a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "# standard libraries\n",
    "import os\n",
    "import pathlib\n",
    "import itertools\n",
    "import string\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pydna\n",
    "import math\n",
    "import pathlib\n",
    "import textwrap as _textwrap\n",
    "from pydna._pretty import pretty_str as _pretty_str\n",
    "import csv\n",
    "\n",
    "\n",
    "''' This part of the lab module is used for PCRs\n",
    "\n",
    "HELPER FUNCTIONS: \n",
    "-----------------\n",
    "pcr_volumes (formerly known as volumes)\n",
    "det_proc_speed\n",
    "det_elon_time\n",
    "PCR_program\n",
    "det_no_of_thermal_cyclers\n",
    "pcr_locations1\n",
    "transf_locations1\n",
    "\n",
    "'''\n",
    "\n",
    "def grouper(iterable, max_diff):\n",
    "    prev = None\n",
    "    group = []\n",
    "    for item in iterable:\n",
    "        if not prev or item - prev <= max_diff:\n",
    "            group.append(item)\n",
    "        else:\n",
    "            yield group\n",
    "            group = [item]\n",
    "        prev = item\n",
    "    if group:\n",
    "        yield group\n",
    "\n",
    "\n",
    "def pcr_volumes(vol_p_reac=10, no_of_reactions = 1, standard_reagents = [], standard_volumes = []):\n",
    "\n",
    "    ''' Can make a reaction scheme for PCR master mixes such as:\n",
    "\n",
    "        3-reactions:\n",
    "        -----------------------\n",
    "    \t               vol_p_reac\tvol_p_3_reac\n",
    "        Template       \t 1.0\t        3.0\n",
    "        Primer 1\t     2.5\t        7.5\n",
    "        Primer 2\t     2.5\t        7.5\n",
    "        H20\t             19.0\t        57.0\n",
    "        MM\t             25.0\t        75.0\n",
    "        Total\t         50.0         \t150.0\n",
    "        -----------------------\n",
    "\n",
    "        If this is used as input:\n",
    "        ------------------------\n",
    "        (vol_p_reac = 10,\n",
    "        no_of_reactions = 6,\n",
    "        standard_reagents = [\"DNA\",\"Buffer, Cutsmart\",\"H20\",\"Enz, USER\"],\n",
    "        standard_volumes = [1,1,7,1])\n",
    "\n",
    "        The following reaction scheme will be made:\n",
    "        -------------------------\n",
    "                        vol_p_reac\tvol_p_x_reac\n",
    "        DNA\t                1.0\t        6.0\n",
    "        Buffer, Cutsmart\t1.0\t        6.0\n",
    "        H20\t                7.0\t        42.0\n",
    "        Enz, USER\t        1.0\t        6.0\n",
    "        Total\t            10.0      \t60.0\n",
    "        -------------------------\n",
    "\n",
    "\n",
    "        '''\n",
    "    standard_total_volume = float(sum(standard_volumes))\n",
    "    volumes_p_x = [float(val)/standard_total_volume*float(vol_p_reac) for val in standard_volumes]\n",
    "    volumes_p_x_p_y_reactions = [val*no_of_reactions for val in volumes_p_x]\n",
    "    volumes_p_x_p_y_reactions\n",
    "\n",
    "    volumes_p_x_plus_total = volumes_p_x + [sum(volumes_p_x)]\n",
    "    volumes_p_x_p_y_reactions_plus_total = volumes_p_x_p_y_reactions + [sum(volumes_p_x_p_y_reactions)]\n",
    "    reagents_plus_total = standard_reagents + ['Total']\n",
    "\n",
    "    volumes_df  = pd.DataFrame(data =  {'vol_p_reac': volumes_p_x_plus_total,\n",
    "                                        'vol_p_'+str(no_of_reactions)+'_reac': volumes_p_x_p_y_reactions_plus_total},\n",
    "                               index  = reagents_plus_total)\n",
    "    return(volumes_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a33a3183-802f-4eed-907b-ef48ebdf45dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vol_p_reac</th>\n",
       "      <th>vol_p_6_reac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DNA</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Buffer, Cutsmart</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H20</th>\n",
       "      <td>7.0</td>\n",
       "      <td>42.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Enz, USER</th>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>10.0</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  vol_p_reac  vol_p_6_reac\n",
       "DNA                      1.0           6.0\n",
       "Buffer, Cutsmart         1.0           6.0\n",
       "H20                      7.0          42.0\n",
       "Enz, USER                1.0           6.0\n",
       "Total                   10.0          60.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "vol_p_reac = 10.0\n",
    "\n",
    "no_of_reactions = 6\n",
    "standard_reagents = [\"DNA\",\"Buffer, Cutsmart\",\"H20\",\"Enz, USER\"]\n",
    "standard_volumes = [1,1,7,1]\n",
    "\n",
    "pcr_reaction_df = pcr_volumes(vol_p_reac,no_of_reactions ,standard_reagents, standard_volumes)\n",
    "\n",
    "pcr_reaction_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ed8b8cc9-6306-4cc8-8b8a-35fc8e8b774c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'vol_p_reac': {'DNA': 1.0,\n",
       "  'Buffer, Cutsmart': 1.0,\n",
       "  'H20': 7.0,\n",
       "  'Enz, USER': 1.0,\n",
       "  'Total': 10.0},\n",
       " 'vol_p_6_reac': {'DNA': 6.0,\n",
       "  'Buffer, Cutsmart': 6.0,\n",
       "  'H20': 42.0,\n",
       "  'Enz, USER': 6.0,\n",
       "  'Total': 60.0}}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcr_reaction_df.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "db36d217-c837-4558-8d86-45330599580b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dict ={'vol_p_reac': {'DNA': 1.0,\n",
    "  'Buffer, Cutsmart': 1.0,\n",
    "  'H20': 7.0,\n",
    "  'Enz, USER': 1.0,\n",
    "  'Total': 10.0},\n",
    " 'vol_p_6_reac': {'DNA': 6.0,\n",
    "  'Buffer, Cutsmart': 6.0,\n",
    "  'H20': 42.0,\n",
    "  'Enz, USER': 6.0,\n",
    "  'Total': 60.0}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "408ab3eb-09a7-46fb-951e-cedc95a89439",
   "metadata": {},
   "outputs": [],
   "source": [
    "pcr_reaction_df_dict = pd.DataFrame(data =df_dict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "449d4436-faad-4d4c-a921-b0ab43a0dd1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pd.testing.assert_index_equal(pcr_reaction_df_dict,pcr_reaction_df,check_names=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c9f01d37-13cc-4814-81e8-dcdbdaeb6ad5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def det_proc_speed(amplicon):\n",
    "    '''\n",
    "    This function determines process speed dependent on the which polymerase is used\n",
    "    '''\n",
    "    \n",
    "    if \"proc_speed\" in amplicon.forward_primer.annotations:\n",
    "        print(\"proc_speed already set\")\n",
    "        return amplicon\n",
    "\n",
    "    # proc_speed units are seconds/kb\n",
    "    if amplicon.annotations['polymerase'] == \"OneTaq Hot Start\":\n",
    "        proc_speed = 60\n",
    "    elif amplicon.annotations['polymerase'] == \"Q5 Hot Start\":\n",
    "        proc_speed = 30\n",
    "    elif amplicon.annotations['polymerase'] == \"Phusion\":\n",
    "        proc_speed = 30\n",
    "\n",
    "    amplicon.annotations['proc_speed'] = proc_speed\n",
    "\n",
    "    return amplicon\n",
    "\n",
    "def det_elon_time(amplicon):\n",
    "\n",
    "    ''' This function determines elongation time for an amplicon\n",
    "    and add the elongation time to the amplicon annotations'''\n",
    "\n",
    "    if \"elongation_time\" in amplicon.forward_primer.annotations:\n",
    "        print(\"elongation_time already set\")\n",
    "        return amplicon\n",
    "\n",
    "    # elongation_time units are seconds\n",
    "    elongation_time = amplicon.annotations['proc_speed'] * len(amplicon)/1000\n",
    "    amplicon.annotations['elongation_time'] = math.ceil(elongation_time)\n",
    "\n",
    "    return amplicon\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ad7eaa03-d35a-4538-86d4-bfb788f69de1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dseqrecord\n",
      "circular: False\n",
      "size: 2123\n",
      "ID: 2123bp_uEC9clUhipmw1j2qRizKWJzHCHc\n",
      "Name: 2123bp_PCR_prod\n",
      "Description: pcr_product_description_<unknown description>\n",
      "Number of features: 2\n",
      "/molecule_type=DNA\n",
      "/polymerase=OneTaq Hot Start\n",
      "/proc_speed=60\n",
      "/elongation_time=128\n",
      "Dseq(-2123)\n",
      "taca..gcac\n",
      "atgt..cgtg\n"
     ]
    }
   ],
   "source": [
    "from pydna.dseqrecord import Dseqrecord\n",
    "from pydna.readers import read\n",
    "from pydna.amplify import pcr\n",
    "from pydna.primer import Primer\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "    \n",
    "# initialize\n",
    "middle = 'a'*2000\n",
    "template = Dseqrecord(\"tacactcaccgtctatcattatcagcgacgaagcgagcgcgaccgcgagcgcgagcgca\"+middle+\"caggagcgagacacggcgacgcgagcgagcgagcgatactatcgactgtatcatctgatagcac\")\n",
    "p1 = Primer(\"tacactcaccgtctatcattatc\")\n",
    "p2 = Primer(\"cgactgtatcatctgatagcac\").reverse_complement()\n",
    "amplicon = pcr(p1, p2, template)\n",
    "amplicon.annotations['polymerase'] = \"OneTaq Hot Start\"\n",
    "det_proc_speed(amplicon)\n",
    "\n",
    "\n",
    "print(    det_elon_time(amplicon))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fd3c51d8-ae29-485e-9a10-74994a2edca4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def well_keys_96(row = True):\n",
    "    '''\n",
    "    If true it generates keys for a 96 well plate by row.\n",
    "    else it does it by column\n",
    "    '''\n",
    "    key_list = 'ABCDEFGH'\n",
    "\n",
    "    if row == True:\n",
    "        wells_keys = []\n",
    "        for i in range(0,len(key_list)):\n",
    "            for j in range(1,13):\n",
    "                wells_keys.append(key_list[i]+str(j))\n",
    "\n",
    "    else:\n",
    "        wells_keys = []\n",
    "        for i in range(1,13):\n",
    "            for j in range(0,len(key_list)):\n",
    "                wells_keys.append(str(key_list[j])+str(i))\n",
    "\n",
    "    return wells_keys\n",
    "\n",
    "well_keys = well_keys_96()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "79bd9ff3-a4e5-46bd-97b2-6f8e214730fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A1', 'B1', 'C1', 'D1', 'E1', 'F1', 'G1', 'H1', 'A2', 'B2', 'C2', 'D2', 'E2', 'F2', 'G2', 'H2', 'A3', 'B3', 'C3', 'D3', 'E3', 'F3', 'G3', 'H3', 'A4', 'B4', 'C4', 'D4', 'E4', 'F4', 'G4', 'H4', 'A5', 'B5', 'C5', 'D5', 'E5', 'F5', 'G5', 'H5', 'A6', 'B6', 'C6', 'D6', 'E6', 'F6', 'G6', 'H6', 'A7', 'B7', 'C7', 'D7', 'E7', 'F7', 'G7', 'H7', 'A8', 'B8', 'C8', 'D8', 'E8', 'F8', 'G8', 'H8', 'A9', 'B9', 'C9', 'D9', 'E9', 'F9', 'G9', 'H9', 'A10', 'B10', 'C10', 'D10', 'E10', 'F10', 'G10', 'H10', 'A11', 'B11', 'C11', 'D11', 'E11', 'F11', 'G11', 'H11', 'A12', 'B12', 'C12', 'D12', 'E12', 'F12', 'G12', 'H12']\n"
     ]
    }
   ],
   "source": [
    "print(well_keys_96(row = False ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8bd53382-c047-499c-a9bb-ef7f4ab1ea60",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_virtual_plates_fromDF(f_primers:list, r_primers:list, templates:list, Dataframe_with_PCR_contents):\n",
    "    '''\n",
    "    THis function can make virtual plates from lists of primers and templates.\n",
    "    The Pandas DataFrame is used to calculate how much pcr and h2o is needed for the reactions.\n",
    "    '''\n",
    "\n",
    "    #initializing\n",
    "    source_plate1 = lab.Plate96(name=\"1\") # Forward primers\n",
    "    source_plate2 = lab.Plate96(name=\"2\") # Reverse primers\n",
    "    source_plate3 = lab.Plate96(name=\"3\") # Templates\n",
    "    source_plate4 = lab.Plate2x4(name=\"4\") # PCR mix and H20\n",
    "    well_keys = well_keys_96()\n",
    "    volume_needed = len(Dataframe_with_PCR_contents)+1\n",
    "\n",
    "    # Adding each primer/template to the virtual plates here a volume of 5 is arbitrarily chosen\n",
    "    for i in range(0,len(f_primers)):\n",
    "        source_plate1[well_keys[i]].add_content({f_primers[i]:1},volume= 5.0)\n",
    "    for j in range(0,len(r_primers)):\n",
    "        source_plate2[well_keys[j]].add_content({r_primers[j]:1},volume= 5.0)\n",
    "    for k in range(0,len(templates)):\n",
    "        source_plate3[well_keys[k]].add_content({templates[k]:1},volume= 5.0 )\n",
    "\n",
    "    # Adding MM and H2O to the final plate\n",
    "    source_plate4[\"A1\"].add_content({'PCR_2x_mix':10},volume = volume_needed*10)\n",
    "    source_plate4[\"A2\"].add_content({'H2O_MQ':10},volume = volume_needed*7)\n",
    "\n",
    "    return source_plate1, source_plate2, source_plate3, source_plate4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b4a38e81-8b09-48b5-a401-49220fa5531a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import synbiopython.lab_automation as lab\n",
    "# Read in test data\n",
    "test_data = pd.read_excel('../tests/files_for_testing/Random_PCR_list.xlsx')\n",
    "\n",
    "# Add test primers and templates\n",
    "fwd_primers = ['F0069','F0078', 'F0083','F0088','F0093','F0098','F00103','F00108']\n",
    "rev_primers = ['R00174','R00175','R00176','R00177']\n",
    "templates = ['pRPL15B']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0866674e-235d-42d1-ba84-c1dd82d6673c",
   "metadata": {},
   "outputs": [],
   "source": [
    "f, r, t, pcr_mix = make_virtual_plates_fromDF(fwd_primers, rev_primers, templates, test_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d3e069ed-579f-4d45-9514-878a5ab6fefa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "synbiopython.lab_automation.containers.builtin_containers.Plate96"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7d086e73-f807-4d2c-8e8e-99aa70d0662b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A1'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f.index_to_wellname(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "70ff53b1-1a03-4cc4-8706-022d9edff085",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "96"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f.num_wells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9524af07-628b-4ea7-bf53-9a031af4876c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{1: ['A1', 'B1', 'C1', 'D1', 'E1', 'F1', 'G1', 'H1'], 2: ['A2', 'B2', 'C2', 'D2', 'E2', 'F2', 'G2', 'H2'], 3: ['A3', 'B3', 'C3', 'D3', 'E3', 'F3', 'G3', 'H3'], 4: ['A4', 'B4', 'C4', 'D4', 'E4', 'F4', 'G4', 'H4'], 5: ['A5', 'B5', 'C5', 'D5', 'E5', 'F5', 'G5', 'H5'], 6: ['A6', 'B6', 'C6', 'D6', 'E6', 'F6', 'G6', 'H6'], 7: ['A7', 'B7', 'C7', 'D7', 'E7', 'F7', 'G7', 'H7'], 8: ['A8', 'B8', 'C8', 'D8', 'E8', 'F8', 'G8', 'H8'], 9: ['A9', 'B9', 'C9', 'D9', 'E9', 'F9', 'G9', 'H9'], 10: ['A10', 'B10', 'C10', 'D10', 'E10', 'F10', 'G10', 'H10'], 11: ['A11', 'B11', 'C11', 'D11', 'E11', 'F11', 'G11', 'H11'], 12: ['A12', 'B12', 'C12', 'D12', 'E12', 'F12', 'G12', 'H12']}\n"
     ]
    }
   ],
   "source": [
    "print(f.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e8f0961b-f4a2-4986-850e-8af88fe72d12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "aaf4157a-55fd-46f8-813d-99bf75001272",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method Plate.list_well_data_fields of Plate96(1)>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f.list_well_data_fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c0390d51-2668-4a1c-9ff1-283d59e5fb96",
   "metadata": {},
   "outputs": [],
   "source": [
    "def picklist_from_plates(F_primer_plate, R_primer_plate, Templates_plate, MM_H20_plate, PCR_dataframe):\n",
    "    '''This function can generate picklist from virtual plates and pandas dataframe with PCR components'''\n",
    "\n",
    "    FORWARD_PRIMERS = F_primer_plate.to_pandas_dataframe()\n",
    "    REVERSE_PRIMERS = R_primer_plate.to_pandas_dataframe()\n",
    "    TEMPLATES = Templates_plate.to_pandas_dataframe()\n",
    "    MM_H2O = MM_H20_plate.to_pandas_dataframe()\n",
    "\n",
    "    destination_plate = lab.Plate96(name=\"5\")\n",
    "    picklist = lab.PickList()\n",
    "    counter = 0\n",
    "\n",
    "    # FIRST WE CHECK WICH COMPONENTS WE NEED FOR THE PCRs\n",
    "    for index,row in PCR_dataframe.iterrows():\n",
    "        templates = row['Template']\n",
    "        f_primers = row['forward_primer']\n",
    "        r_primers = row['reverse_primer']\n",
    "\n",
    "        # FORWARD PRIMERS\n",
    "        # we see if we can find the primer we need in the primer plate and if we find it we will add it to the picklist\n",
    "        for index1, row1 in FORWARD_PRIMERS.iterrows():\n",
    "            for k, v in  row1['content']['quantities'].items():\n",
    "                if k == f_primers:\n",
    "                    transfer_forward_primers = lab.Transfer(F_primer_plate.wells[index1], destination_plate.wells[well_keys[counter]], 1 )\n",
    "                    picklist.add_transfer(transfer = transfer_forward_primers)\n",
    "\n",
    "        # REVERSE PRIMERS\n",
    "        # we see if we can find the primer we need in the primer plate\n",
    "        for index2, row2 in REVERSE_PRIMERS.iterrows():\n",
    "            for k, v in row2['content']['quantities'].items():\n",
    "                if k == r_primers:\n",
    "                    transfer_reverse_primers = lab.Transfer(R_primer_plate.wells[index2], destination_plate.wells[well_keys[counter]], 1 )\n",
    "                    picklist.add_transfer(transfer = transfer_reverse_primers)\n",
    "\n",
    "        # TEMPLATE\n",
    "        # we see if we can find the primer we need in the plate\n",
    "        for index3, row3 in TEMPLATES.iterrows():\n",
    "            for k, v in row3['content']['quantities'].items():\n",
    "                if k == templates:\n",
    "                    transfer_template = lab.Transfer(Templates_plate.wells[index3], destination_plate.wells[well_keys[counter]], 1 )\n",
    "                    picklist.add_transfer(transfer = transfer_template)\n",
    "\n",
    "        # Master mix\n",
    "        transfer_MM = lab.Transfer(MM_H20_plate.wells['A1'], destination_plate.wells[well_keys[counter]], 10 )\n",
    "        picklist.add_transfer(transfer = transfer_MM)\n",
    "        #H20\n",
    "        transfer_H20 = lab.Transfer(MM_H20_plate.wells['A2'], destination_plate.wells[well_keys[counter]], 7 )\n",
    "        picklist.add_transfer(transfer = transfer_H20)\n",
    "\n",
    "        # TO MAKE SURE WE ADD TO THE CORRECT WELL\n",
    "        counter += 1\n",
    "\n",
    "    return picklist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "53c6b0cc-b5d6-472d-940d-928520f6736d",
   "metadata": {},
   "outputs": [],
   "source": [
    "picklisten = picklist_from_plates(f, r, t, pcr_mix, test_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ba0df261-ec63-4c00-8bb4-6fdbdee8309f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Transfer 1.00E+00L from 1 A1 into 5 A1'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "picklisten.to_plain_string()[:38]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b485da24-b971-426f-8a86-9b02abfa11f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1:A1, 5:A1, 1 \\n2:A1, 5:A1, 1 \\n3:A1, 5:A1, 1 \\n4:A1, 5:A1, 10 \\n4:A2, 5:A1, 7 \\n1:A1, 5:A2, 1 \\n2:A2, 5:A2, 1 \\n3:A1, 5:A2, 1 \\n4:A1, 5:A2, 10 \\n4:A2, 5:A2, 7 \\n1:A1, 5:A3, 1 \\n2:A3, 5:A3, 1 \\n3:A1, 5:A3, 1 \\n4:A1, 5:A3, 10 \\n4:A2, 5:A3, 7 \\n1:A1, 5:A4, 1 \\n2:A4, 5:A4, 1 \\n3:A1, 5:A4, 1 \\n4:A1, 5:A4, 10 \\n4:A2, 5:A4, 7 \\n1:A2, 5:A5, 1 \\n2:A1, 5:A5, 1 \\n3:A1, 5:A5, 1 \\n4:A1, 5:A5, 10 \\n4:A2, 5:A5, 7 \\n1:A2, 5:A6, 1 \\n2:A2, 5:A6, 1 \\n3:A1, 5:A6, 1 \\n4:A1, 5:A6, 10 \\n4:A2, 5:A6, 7 \\n1:A2, 5:A7, 1 \\n2:A3, 5:A7, 1 \\n3:A1, 5:A7, 1 \\n4:A1, 5:A7, 10 \\n4:A2, 5:A7, 7 \\n1:A2, 5:A8, 1 \\n2:A4, 5:A8, 1 \\n3:A1, 5:A8, 1 \\n4:A1, 5:A8, 10 \\n4:A2, 5:A8, 7 \\n1:A3, 5:A9, 1 \\n2:A1, 5:A9, 1 \\n3:A1, 5:A9, 1 \\n4:A1, 5:A9, 10 \\n4:A2, 5:A9, 7 \\n1:A3, 5:A10, 1 \\n2:A2, 5:A10, 1 \\n3:A1, 5:A10, 1 \\n4:A1, 5:A10, 10 \\n4:A2, 5:A10, 7 \\n1:A3, 5:A11, 1 \\n2:A3, 5:A11, 1 \\n3:A1, 5:A11, 1 \\n4:A1, 5:A11, 10 \\n4:A2, 5:A11, 7 \\n1:A3, 5:A12, 1 \\n2:A4, 5:A12, 1 \\n3:A1, 5:A12, 1 \\n4:A1, 5:A12, 10 \\n4:A2, 5:A12, 7 \\n1:A4, 5:B1, 1 \\n2:A1, 5:B1, 1 \\n3:A1, 5:B1, 1 \\n4:A1, 5:B1, 10 \\n4:A2, 5:B1, 7 \\n1:A4, 5:B2, 1 \\n2:A2, 5:B2, 1 \\n3:A1, 5:B2, 1 \\n4:A1, 5:B2, 10 \\n4:A2, 5:B2, 7 \\n1:A4, 5:B3, 1 \\n2:A3, 5:B3, 1 \\n3:A1, 5:B3, 1 \\n4:A1, 5:B3, 10 \\n4:A2, 5:B3, 7 \\n1:A4, 5:B4, 1 \\n2:A4, 5:B4, 1 \\n3:A1, 5:B4, 1 \\n4:A1, 5:B4, 10 \\n4:A2, 5:B4, 7 \\n1:A5, 5:B5, 1 \\n2:A1, 5:B5, 1 \\n3:A1, 5:B5, 1 \\n4:A1, 5:B5, 10 \\n4:A2, 5:B5, 7 \\n1:A5, 5:B6, 1 \\n2:A2, 5:B6, 1 \\n3:A1, 5:B6, 1 \\n4:A1, 5:B6, 10 \\n4:A2, 5:B6, 7 \\n1:A5, 5:B7, 1 \\n2:A3, 5:B7, 1 \\n3:A1, 5:B7, 1 \\n4:A1, 5:B7, 10 \\n4:A2, 5:B7, 7 \\n1:A5, 5:B8, 1 \\n2:A4, 5:B8, 1 \\n3:A1, 5:B8, 1 \\n4:A1, 5:B8, 10 \\n4:A2, 5:B8, 7 \\n1:A6, 5:B9, 1 \\n2:A1, 5:B9, 1 \\n3:A1, 5:B9, 1 \\n4:A1, 5:B9, 10 \\n4:A2, 5:B9, 7 \\n1:A6, 5:B10, 1 \\n2:A2, 5:B10, 1 \\n3:A1, 5:B10, 1 \\n4:A1, 5:B10, 10 \\n4:A2, 5:B10, 7 \\n1:A6, 5:B11, 1 \\n2:A3, 5:B11, 1 \\n3:A1, 5:B11, 1 \\n4:A1, 5:B11, 10 \\n4:A2, 5:B11, 7 \\n1:A6, 5:B12, 1 \\n2:A4, 5:B12, 1 \\n3:A1, 5:B12, 1 \\n4:A1, 5:B12, 10 \\n4:A2, 5:B12, 7 \\n1:A7, 5:C1, 1 \\n2:A1, 5:C1, 1 \\n3:A1, 5:C1, 1 \\n4:A1, 5:C1, 10 \\n4:A2, 5:C1, 7 \\n1:A7, 5:C2, 1 \\n2:A2, 5:C2, 1 \\n3:A1, 5:C2, 1 \\n4:A1, 5:C2, 10 \\n4:A2, 5:C2, 7 \\n1:A7, 5:C3, 1 \\n2:A3, 5:C3, 1 \\n3:A1, 5:C3, 1 \\n4:A1, 5:C3, 10 \\n4:A2, 5:C3, 7 \\n1:A7, 5:C4, 1 \\n2:A4, 5:C4, 1 \\n3:A1, 5:C4, 1 \\n4:A1, 5:C4, 10 \\n4:A2, 5:C4, 7 \\n1:A8, 5:C5, 1 \\n2:A1, 5:C5, 1 \\n3:A1, 5:C5, 1 \\n4:A1, 5:C5, 10 \\n4:A2, 5:C5, 7 \\n1:A8, 5:C6, 1 \\n2:A2, 5:C6, 1 \\n3:A1, 5:C6, 1 \\n4:A1, 5:C6, 10 \\n4:A2, 5:C6, 7 \\n1:A8, 5:C7, 1 \\n2:A3, 5:C7, 1 \\n3:A1, 5:C7, 1 \\n4:A1, 5:C7, 10 \\n4:A2, 5:C7, 7 \\n1:A8, 5:C8, 1 \\n2:A4, 5:C8, 1 \\n3:A1, 5:C8, 1 \\n4:A1, 5:C8, 10 \\n4:A2, 5:C8, 7 '"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "picklisten.to_flowbot_instructions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5ca912db-7e57-47a1-8590-aaae909b08f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "640"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "picklisten.total_transferred_volume()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "acb6a421-ba2c-4a02-afaf-78d15ee7c59f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class RobotAssembly:\n",
    "    \"\"\"Class to generate instructions for robot on demand. It can be initialized in two ways:\n",
    "    1. Use a design assembly object\n",
    "    2. Pandas dataframe with PCRs and Pandas dataframe with Forward and Reverse primers, and Template\n",
    "\n",
    "    Input:\n",
    "    :param DesignAssembly_object\n",
    "    :param Pandas_dataframe_PCR\n",
    "    :param F_primers: list\n",
    "    :param R_primers: list\n",
    "    :param:Templates:list\n",
    "\n",
    "    Robot_output:\n",
    "    :param  self.source_plateF_primers, self.source_plateR_primers, self.source_plateTemplates, self.source_platePCRmix = sourceplate objets\n",
    "    :param  self.picklist\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, Pandas_dataframe_PCR, F_primers:list, R_primers:list, Templates:list):\n",
    "\n",
    "        ###  1.INITIALIZING ##\n",
    "        ######################## Assembly from plates ############################\n",
    "        if Pandas_dataframe_PCR is not None and F_primers is not None and R_primers is not None and Templates is not None:\n",
    "            # Initializing instances\n",
    "            self.pandas_PCR = Pandas_dataframe_PCR\n",
    "            self.forward_primers = F_primers\n",
    "            self.reverse_primers = R_primers\n",
    "            self.templates = Templates\n",
    "\n",
    "            #Virtual plates\n",
    "            self.source_plateF_primers, self.source_plateR_primers, self.source_plateTemplates, self.source_platePCRmix = make_virtual_plates_fromDF( self.forward_primers, self.reverse_primers, self.templates, self.pandas_PCR)\n",
    "\n",
    "            # Generating Picklist from virtual plates\n",
    "            self.picklist = picklist_from_plates(self.source_plateF_primers, self.source_plateR_primers, self.source_plateTemplates, self.source_platePCRmix, self.pandas_PCR)\n",
    "\n",
    "        else:\n",
    "            print('Remember to put in all the neccesarry components')\n",
    "    ##########################################################\n",
    "    ###################  CLASS METHODS  ######################\n",
    "\n",
    "    def PlatesToExcelFile(self):\n",
    "        '''Returns an excel file of the plate setup that needs to be made before the flowbot can operate'''\n",
    "        df1 = self.source_plateF_primers.to_pandas_dataframe()\n",
    "        df2 = self.source_plateR_primers.to_pandas_dataframe()\n",
    "        df3 = self.source_plateTemplates.to_pandas_dataframe()\n",
    "        df4 = self.source_platePCRmix.to_pandas_dataframe()\n",
    "\n",
    "        with pd.ExcelWriter('Plate_instructions.xlsx') as writer:\n",
    "            return df1.to_excel(writer, sheet_name='Forward_primer_wells'), df2.to_excel(writer, sheet_name='Reverse_primer_wells'), df3.to_excel(writer, sheet_name='Template_wells'),df4.to_excel(writer, sheet_name='PCR_MIX_H20')\n",
    "\n",
    "\n",
    "    def print_well_df_to_string(self):\n",
    "        ''' Prints Pandas dataframe in string format'''\n",
    "\n",
    "        df1 = self.source_plateF_primers.to_pandas_dataframe()\n",
    "        df2 = self.source_plateR_primers.to_pandas_dataframe()\n",
    "        df3 = self.source_plateTemplates.to_pandas_dataframe()\n",
    "        df4 = self.source_platePCRmix.to_pandas_dataframe()\n",
    "\n",
    "        return print(' ###Forward primers: ', df1.to_csv(sep=' ', index=False, header=True),'\\n', '###Reverse primers: ', df2.to_csv(sep=' ', index=False, header=True),'\\n','###Templates: ',df3.to_csv(sep=' ', index=False, header=True),  df4.to_csv(sep=' ', index=False, header=True))\n",
    "\n",
    "    def FlowbotInstructionsToCSV(self):\n",
    "        '''Function that prints flowbot instructions to csv format'''\n",
    "\n",
    "        with open('Flowbot_instructions.csv','w') as outfile:\n",
    "\n",
    "            return print('source, target, volume','\\n',*self.picklist.to_flowbot_instructions(), sep='',end='\\n', file = outfile)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3a942a47-d568-445e-8d06-a4026b2dba0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_Robotassemly = RobotAssembly(test_data,fwd_primers ,rev_primers, templates)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "89448136-d7ac-4096-8c6b-4cc8ed7059ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pRPL15B']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_Robotassemly.templates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "96890adf-9c48-4563-bc82-9b56a431379a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "template = 'TCTTTGAAAAGATAATGTATGATTATGCTTTCACTCATATTTATACAGAAACTTGATGTTTTCTTTCGAGTATATACAAGGTGATTACATGTACGTTTGAAGTACAACTCTAGATTTTGTAGTGCCCTCTTGGGCTAGCGGTAAAGGTGCGCATTTTTTCACACCCTACAATGTTCTGTTCAAAAGATTTTGGTCAAACGCTGTAGAAGTGAAAGTTGGTGCGCATGTTTCGGCGTTCGAAACTTCTCCGCAGTGAAAGATAAATGATCGCCGTAGTAACGTCGCTGTCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGGTGCTTTTTTTGTTTTTTATGTCTTCGAGTCATGTAATTAGTTAAGTGCAGGT'\n",
    "primerF = 'CGTGCGAUTCTTTGAAAAGATAATGTATGA'\n",
    "primerR = 'ACCTGCACUTAACTAATTACATGACTCGA'\n",
    "\n",
    "from pydna.amplify import pcr\n",
    "gRNA1_pcr_prod = pcr(primerF,primerR, template)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d095e680-d884-47d2-908b-bfd6d8f2bb99",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b83cc7da-0c7f-4e7d-9364-1a072d4c8f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "def USER_enzyme(amplicon):\n",
    "    \"\"\"\n",
    "    Take pydna.amplicon.Amplicon\n",
    "    Return Dseqrecord trimmed according to pcr primers\n",
    "    \"\"\"\n",
    "    fw_U_idx = amplicon.forward_primer.seq.find(\"U\")\n",
    "    #fw_U_idx\n",
    "\n",
    "    rv_U_idx = amplicon.reverse_primer.seq.find(\"U\")\n",
    "    #rv_U_idx\n",
    "\n",
    "    digested_watson = amplicon.seq.watson[fw_U_idx:]\n",
    "    #digested_watson\n",
    "\n",
    "    digested_crick = amplicon.seq.crick[rv_U_idx+1:]\n",
    "    #digested_crick\n",
    "\n",
    "    digested_pcr = pydna.dseqrecord.Dseqrecord(pydna.dseq.Dseq(watson=digested_watson, crick = digested_crick),\n",
    "                                features = amplicon.features, annotations = amplicon.annotations)\n",
    "    #digested_pcr.seq\n",
    "    return digested_pcr\n",
    "\n",
    "DIGESTED = USER_enzyme(gRNA1_pcr_prod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8e50093a-be9b-4935-afbc-2a673727256a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "UTCTTTGAAAAGATAATGTATGATTATGCTTTCACTCATATTTATACAGAAACTTGATGTTTTCTTTCGAGTATATACAAGGTGATTACATGTACGTTTGAAGTACAACTCTAGATTTTGTAGTGCCCTCTTGGGCTAGCGGTAAAGGTGCGCATTTTTTCACACCCTACAATGTTCTGTTCAAAAGATTTTGGTCAAACGCTGTAGAAGTGAAAGTTGGTGCGCATGTTTCGGCGTTCGAAACTTCTCCGCAGTGAAAGATAAATGATCGCCGTAGTAACGTCGCTGTCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGGTGCTTTTTTTGTTTTTTATGTCTTCGAGTCATGTAATTAGTTAAGTGCAGGT"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "Digested_USER = USER_enzyme(gRNA1_pcr_prod)\n",
    "Digested_USER.seq.watson"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "914f468b-1e32-4eaf-945f-90a117c08e65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "418"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Digested_USER.seq.watson)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f6ddfd6d-e4e6-42f1-b80f-b801bd846d51",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_gRNAs(template, name):\n",
    "    gRNAs = []\n",
    "    for feature in template.features:\n",
    "        if name in feature.qualifiers.get('name',\"\"):\n",
    "            \n",
    "            gRNA = template[feature.location.start:feature.location.end]\n",
    "            gRNA.name = feature.qualifiers.get('name',\"\")\n",
    "            gRNAs.append(gRNA)\n",
    "    \n",
    "    return(gRNAs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d5840ab9-6fb7-4537-b9bf-239b6eb90c78",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqRecord import SeqRecord\n",
    "from pydna.amplify import pcr\n",
    "\n",
    "template = 'TCTTTGAAAAGATAATGTATGATTATGCTTTCACTCATATTTATACAGAAACTTGATGTTTTCTTTCGAGTATATACAAGGTGATTACATGTACGTTTGAAGTACAACTCTAGATTTTGTAGTGCCCTCTTGGGCTAGCGGTAAAGGTGCGCATTTTTTCACACCCTACAATGTTCTGTTCAAAAGATTTTGGTCAAACGCTGTAGAAGTGAAAGTTGGTGCGCATGTTTCGGCGTTCGAAACTTCTCCGCAGTGAAAGATAAATGATCGCCGTAGTAACGTCGCTGTCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGGTGCTTTTTTTGTTTTTTATGTCTTCGAGTCATGTAATTAGTTAAGTGCAGGT'\n",
    "primerF = 'CGTGCGAUTCTTTGAAAAGATAATGTATGA'\n",
    "primerR = 'ACCTGCACUTAACTAATTACATGACTCGA'\n",
    "\n",
    "rec_vector = pcr(primerF,primerR, template)\n",
    "\n",
    "# Adding a feature\n",
    "rec_vector.add_feature(40,60, name ='gRNA' , label = ['gRNA'])\n",
    "gRNA = extract_gRNAs(rec_vector, 'gRNA')\n",
    "\n",
    "assert gRNA[0].seq.watson == 'ACTCATATTTATACAGAAAC'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e9bdd1fb-23fa-4b7e-9eef-8b3320e715a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OrderedDict([('label', ['name']), ('ApEinfo_fwdcolor', ['#baffa3']), ('ApEinfo_revcolor', ['#ffbaba'])])\n",
      "OrderedDict([('label', ['name']), ('ApEinfo_fwdcolor', ['#baffa3']), ('ApEinfo_revcolor', ['#ffbaba'])])\n",
      "{'name': 'gRNA', 'label': ['gRNA']}\n",
      "{'name': 'gRNA', 'label': ['gRNA']}\n"
     ]
    }
   ],
   "source": [
    "rec_vector.add_feature(40,60, name ='gRNA' , label = ['gRNA'])\n",
    "for feature in rec_vector.features:\n",
    "    print(feature.qualifiers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "18eea0fa-48d1-4bbf-b5a6-5a6a19dc62b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['name']\n",
      "OrderedDict([('label', ['name']), ('ApEinfo_fwdcolor', ['#baffa3']), ('ApEinfo_revcolor', ['#ffbaba'])])\n",
      "['name']\n",
      "OrderedDict([('label', ['name']), ('ApEinfo_fwdcolor', ['#baffa3']), ('ApEinfo_revcolor', ['#ffbaba'])])\n",
      "['gRNA']\n",
      "{'name': 'gRNA', 'label': ['gRNA']}\n",
      "['gRNA']\n",
      "{'name': 'gRNA', 'label': ['gRNA']}\n"
     ]
    }
   ],
   "source": [
    "for feature in rec_vector.features:\n",
    "    print(feature.qualifiers['label'])\n",
    "    print(feature.qualifiers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "85babe1b-3af0-4054-afea-3b01eb2e5313",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('label', [name]),\n",
       "             ('ApEinfo_fwdcolor', ['#baffa3']),\n",
       "             ('ApEinfo_revcolor', ['#ffbaba'])])"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rec_vector.features[1].qualifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "847950a6-6c05-4e3f-be2a-7c8337cc7e7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Amplicon(20), Amplicon(20)]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gRNA = extract_gRNAs(rec_vector, 'gRNA')\n",
    "gRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9c8787b2-07a7-4f6e-98bf-ce9743184ef5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ACTCATATTTATACAGAAAC"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gRNA[0].seq.watson\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "df02d3b6-1e97-4958-a302-13527989be1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "template = 'TCTTTGAAAAGATAATGTATGATTATGCTTTCACTCATATTTATACAGAAACTTGATGTTTTCTTTCGAGTATATACAAGGTGATTACATGTACGTTTGAAGTACAACTCTAGATTTTGTAGTGCCCTCTTGGGCTAGCGGTAAAGGTGCGCATTTTTTCACACCCTACAATGTTCTGTTCAAAAGATTTTGGTCAAACGCTGTAGAAGTGAAAGTTGGTGCGCATGTTTCGGCGTTCGAAACTTCTCCGCAGTGAAAGATAAATGATCGCCGTAGTAACGTCGCTGTCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGGTGCTTTTTTTGTTTTTTATGTCTTCGAGTCATGTAATTAGTTAAGTGCAGGT'\n",
    "primerF = 'CGTGCGAUTCTTTGAAAAGATAATGTATGA'\n",
    "primerR = 'ACCTGCACUTAACTAATTACATGACTCGA'\n",
    "from pydna.amplify import pcr\n",
    "rec_vector = pcr(primerF,primerR, template)\n",
    "rec_vector.add_feature(-40,60, label =['gRNA'] , type = 'gRNA')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "dc77123e-4386-4011-a4d4-409ef0fafea9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_features_with_negative_loc(record):\n",
    "    \n",
    "    for i in range(len(record.features)):\n",
    "\n",
    "        if record.features[i].location.start < 0 or record.features[i].location.end < 0:\n",
    "            del record.features[i]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "95a32f07-0c48-499b-9946-e58a212fefe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "remove_features_with_negative_loc(rec_vector)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "0b5fbd01-c783-4028-b74b-0734954362c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(rec_vector.features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "7471f4e5-4239-49f0-b8f5-21de87787b2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def casembler(bg_strain, site_names=None, gRNAs=None, parts=None, assembly_limits=None, assembly_names=None, verbose=False, to_benchling=False):\n",
    "    \n",
    "\n",
    "    \"\"\"\n",
    "    Simulate in vivo assembly and integration\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    bg_strain      \n",
    "    site_names      list of names                      e.g. [X-3, XI-3]    \n",
    "    gRNAs           list of 20 bp seqrecords           e.g. [ATF1_gRNA, CroCPR_gRNA]\n",
    "    parts           list of list of parts              e.g. [[ATF1_repair_template],[CPR_repair_template]]\n",
    "    assembly_limits list of numbers of bp              e.g. [200,400]\n",
    "    assembly_names  list of names of DNA post assembly e.g. [\"X_3_tADH1_P2_pPGK1\", \"XI_3_UP_DW\"]\n",
    "    verbose         write DNA                          e.g. False\n",
    "    to_benchling    upload DNA                         e.g. False\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    \n",
    "    One dseqrecord\n",
    "\n",
    "    Examples\n",
    "    --------\n",
    "    \"\"\"\n",
    "    \n",
    "\n",
    "    #initialisation\n",
    "    list_of_list = [site_names, gRNAs, parts, assembly_limits, assembly_names]\n",
    "    not_empty =[]\n",
    "    \n",
    "    #Make sure part is dseqrecord\n",
    "    if list_of_list[2] != None:\n",
    "        for part in list_of_list[2]:\n",
    "            part = pydna.dseqrecord.Dseqrecord.from_SeqRecord(part)\n",
    "        \n",
    "\n",
    "    #Iterate through list to differentate None/list values\n",
    "    for list_no in range(0,len(list_of_list)):\n",
    "        if list_of_list[list_no] != None:\n",
    "            not_empty.append(list_of_list[list_no])\n",
    "        else:\n",
    "            list_of_list[list_no] = []\n",
    "\n",
    "    #Check list lengths match and assign variables. \n",
    "    if len({len(lists) for lists in not_empty}) != 1:\n",
    "        print(\"The provided lists's lengths do not match.\")\n",
    "        for value in not_empty:\n",
    "            print(str(len(value)))\n",
    " \n",
    "\n",
    "    if site_names is None:\n",
    "        site_names = []\n",
    "    else:\n",
    "        site_names = site_names\n",
    "    \n",
    "    if gRNAs is None:\n",
    "        gRNAs = []\n",
    "    else:\n",
    "        gRNAs = gRNAs\n",
    "          \n",
    "    if parts is None:\n",
    "        parts = []\n",
    "    else:\n",
    "        parts = parts\n",
    "            \n",
    "    if assembly_limits is None:\n",
    "        assembly_limits = []\n",
    "    else:\n",
    "        assembly_limits = assembly_limits\n",
    "        \n",
    "    if assembly_names is None:\n",
    "        assembly_names = []\n",
    "    else:\n",
    "        assembly_names = assembly_names\n",
    "        \n",
    "    assemblies = []\n",
    "    for int_no in range(0,len(gRNAs)):  \n",
    "        \n",
    "        gRNA = gRNAs[int_no]\n",
    "\n",
    "        site_name = site_names[int_no]\n",
    "        \n",
    "        site_extract = extract_sites([site_name], [bg_strain], [site_name])[0]\n",
    "        \n",
    "        UP, DW = CAS9_cutting(gRNA, site_extract)\n",
    "        \n",
    "        fragments = [UP, pydna.dseqrecord.Dseqrecord.from_SeqRecord(parts[int_no]), DW]\n",
    "        \n",
    "        assembly = pydna.assembly.Assembly(fragments).assemble_linear()[0]\n",
    "        \n",
    "        # sometimes pydna.assembly.Assembly distorts the start, end location of features to become negative which produces and error when printing. This function is created as a workaround. \n",
    "        # CPR assembly gives DW_XI_3 annotation called \"DW_XI_3\" a negative start location.\n",
    "        # A quick workaround is to remove featuere\n",
    "        remove_features_with_negative_loc(assembly)\n",
    "        \n",
    "        assembly.name = assembly_names[int_no]\n",
    "        \n",
    "        assembly_feat = Bio.SeqFeature.SeqFeature(Bio.SeqFeature.FeatureLocation(0, len(assembly), strand=1), type=\"misc_feature\")\n",
    "        assembly_feat.qualifiers['name']= site_names[int_no]\n",
    "        assembly_feat.qualifiers['label']= site_names[int_no]\n",
    "        assembly.features.append(assembly_feat)\n",
    "        \n",
    "        if verbose:\n",
    "            DNAs = [UP] + [assembly] + [DW]\n",
    "            for DNA in DNAs:\n",
    "                DNA.write(\"./\" + DNA.name + \".gb\") # \"../data/processed/\"\n",
    "        \n",
    "        if to_benchling:\n",
    "            to_benchling(assembly, \"to_benchling\")\n",
    "        \n",
    "        assemblies.append(assembly)\n",
    "        \n",
    "    return functools.reduce(lambda x, y: x+y, assemblies)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d504b08e-41b5-4e5b-9fc2-d1ca0c1d55e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def CAS9_cutting(gRNA_record, background_record):\n",
    "    \"\"\" Simulates cutting by CAS9 given a gRNA\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    gRNA_sequence: string\n",
    "    \n",
    "    background_record: pydna.dseqrecord (pydna.assembly)\n",
    "    \n",
    "    up_surrounding: int (number of bp)\n",
    "\n",
    "    dw_surrounding: int (number of bp)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    \n",
    "    Two dseqrecords\n",
    "\n",
    "    Examples\n",
    "    --------\n",
    "    \n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    gRNA_sequence = gRNA_record.seq.watson.upper()\n",
    "    \n",
    "    background_sequence = background_record.seq.upper()\n",
    "    \n",
    "    gRNA_strand = 1\n",
    "    if background_sequence.find(gRNA_sequence) == -1:\n",
    "    #    print('not on +1')\n",
    "        gRNA_strand = -1\n",
    "        gRNA_sequence = reverse_complement(gRNA_sequence)\n",
    "        if background_sequence.find(gRNA_sequence) ==-1:\n",
    "            print(\"not on -1, CAN'T FIND THE CUT SITE IN YOUR SEQUENCE\")\n",
    "    #print(gRNA_strand)\n",
    "    #print(gRNA_sequence)\n",
    "    \n",
    "    if gRNA_strand == 1:\n",
    "        cut_pos_rel_start = 17\n",
    "    else:\n",
    "        cut_pos_rel_start = 3\n",
    "    \n",
    "    gRNA_start = background_sequence.find(gRNA_sequence)\n",
    "    \n",
    "    cut_site = gRNA_start + cut_pos_rel_start\n",
    " \n",
    "    up = background_record[0 : cut_site]\n",
    "    dw = background_record[cut_site : len(background_sequence)]\n",
    "    \n",
    "    up.name = 'UP' + '_' + gRNA_record.name + '_' + background_record.name\n",
    "    dw.name = 'DW' + '_' + gRNA_record.name + '_' + background_record.name\n",
    "    \n",
    "    up_feature = Bio.SeqFeature.SeqFeature(Bio.SeqFeature.FeatureLocation(0, len(up)), type=\"misc_feature\", strand=+1)\n",
    "    up_feature.qualifiers[\"label\"] = up.name\n",
    "    up.features.append(up_feature)\n",
    "    \n",
    "    dw_feature = Bio.SeqFeature.SeqFeature(Bio.SeqFeature.FeatureLocation(0, len(dw)), type=\"misc_feature\", strand=+1)\n",
    "    dw_feature.qualifiers[\"label\"] = dw.name\n",
    "    dw.features.append(dw_feature)\n",
    "    \n",
    "    up = pydna.dseqrecord.Dseqrecord(up,)\n",
    "    dw = pydna.dseqrecord.Dseqrecord(dw)\n",
    "    \n",
    "    # UPS more than one cut site?\n",
    "    if dw.seq.find(gRNA_sequence) != -1:\n",
    "        print(\"OBS\", gRNA_sequence, \"cuts more than one location!\")\n",
    "        \n",
    "    return(up,dw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "838f2fe1-d9d9-47e9-8ecd-db6c797f4e4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pydna.dseqrecord import Dseqrecord\n",
    "import Bio\n",
    "\n",
    "template = Dseqrecord('TCTTTGAAAAGATAATGTATGATTATGCTTTCACTCATATTTATACAGAAACTTGATGTTTTCTTTCGAGTATATACAAGGTGATTACATGTACGTTTGAAGTACAACTCTAGATTTTGTAGTGCCCTCTTGGGCTAGCGGTAAAGGTGCGCATTTTTTCACACCCTACAATGTTCTGTTCAAAAGATTTTGGTCAAACGCTGTAGAAGTGAAAGTTGGTGCGCATGTTTCGGCGTTCGAAACTTCTCCGCAGTGAAAGATAAATGATCGCCGTAGTAACGTCGCTGTCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGGTGCTTTTTTTGTTTTTTATGTCTTCGAGTCATGTAATTAGTTAAGTGCAGGT')\n",
    "gRNA = Dseqrecord('TCTAGATTTTGTAGTGCCCT')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "dd57c544-684b-4767-93aa-c73711560057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TCTTTGAAAAGATAATGTATGATTATGCTTTCACTCATATTTATACAGAAACTTGATGTTTTCTTTCGAGTATATACAAGGTGATTACATGTACGTTTGAAGTACAACTCTAGATTTTGTAGTGC"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "up, dw = CAS9_cutting(gRNA, template)\n",
    "\n",
    "up.seq.watson\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "e41446f9-ca44-4431-8438-3884fbd24beb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CCTCTTGGGCTAGCGGTAAAGGTGCGCATTTTTTCACACCCTACAATGTTCTGTTCAAAAGATTTTGGTCAAACGCTGTAGAAGTGAAAGTTGGTGCGCATGTTTCGGCGTTCGAAACTTCTCCGCAGTGAAAGATAAATGATCGCCGTAGTAACGTCGCTGTCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGGTGCTTTTTTTGTTTTTTATGTCTTCGAGTCATGTAATTAGTTAAGTGCAGGT"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dw.seq.watson"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "80bd8773-f3cd-48a3-b49f-8c80b9e92816",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def CRIPSIR_knockout(gRNA_record, insertion_site, repair_DNA):\n",
    "    from pydna.assembly import Assembly\n",
    "    \"\"\"Cuts the insertion site with CAS9_cutting and assebmle knockout\"\"\"\n",
    "    #Create fragments after CAS9 cut\n",
    "    IS_UP, IS_DW = CAS9_cutting(gRNA_record, insertion_site)\n",
    "    \n",
    "    #create list of parts and assemble to knockout sequence\n",
    "    assmeble_parts = IS_UP, repair_DNA, IS_DW\n",
    "    assembled_knockout = pydna.assembly.Assembly(assmeble_parts).assemble_linear()[0]\n",
    "    \n",
    "    return assembled_knockout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "c2ad7106-e8f2-4e73-b5a5-6f9ae4e59e2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "insertion_site = Dseqrecord('TCTTTGAAAAGATAATGTATGATTATGCTTTCACTCATATTTATACAGAAACTTGATGTTTTCTTTCGAGTATATACAAGGTGATTACATGTACGTTTGAAGTACAACTCTAGATTTTGTAGTGCCCTCTTGGGCTAGCGGTAAAGGTGCGCATTTTTTCACACCCTACAATGTTCTGTTCAAAAGATTTTGGTCAAACGCTGTAGAAGTGAAAGTTGGTGCGCATGTTTCGGCGTTCGAAACTTCTCCGCAGTGAAAGATAAATGATCGCCGTAGTAACGTCGCTGTCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGGTGCTTTTTTTGTTTTTTATGTCTTCGAGTCATGTAATTAGTTAAGTGCAGGT')\n",
    "gRNA = Dseqrecord('TCTAGATTTTGTAGTGCCCT')\n",
    "repair_template = Dseqrecord('CGTTTGAAGTACAACTCTAGATTTTGTAGTGCCCTCTTGGGCTAGCGGTAAAGGTGCGCATTTTTTCACACCCTACAATGT')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "9b812df3-9954-47ef-9542-61863e66f79d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Knock_out = CRIPSIR_knockout(gRNA,insertion_site, repair_template)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "0048d372-fb83-412c-8ee6-b19a6effa61c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TCTTTGAAAAGATAATGTATGATTATGCTTTCACTCATATTTATACAGAAACTTGATGTTTTCTTTCGAGTATATACAAGGTGATTACATGTACGTTTGAAGTACAACTCTAGATTTTGTAGTGCCCTCTTGGGCTAGCGGTAAAGGTGCGCATTTTTTCACACCCTACAATGTTCTGTTCAAAAGATTTTGGTCAAACGCTGTAGAAGTGAAAGTTGGTGCGCATGTTTCGGCGTTCGAAACTTCTCCGCAGTGAAAGATAAATGATCGCCGTAGTAACGTCGCTGTCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGGTGCTTTTTTTGTTTTTTATGTCTTCGAGTCATGTAATTAGTTAAGTGCAGGT"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Knock_out.seq.watson"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "7b45013c-363d-4992-98e7-da5ce78df7bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://biosustain.benchling.com/api/v2 sk_o6KcYcFcXYJYLQQBN6xlAEOKRooID\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.append('../constrain/lims')\n",
    "from benchling_api import from_benchling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d1b438b-3134-4f4a-a404-37fe4bf6ea92",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "df2cad2d-3473-4f2d-9f89-fd8e44875961",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_template_amplification_sites(templates, names, terminator):\n",
    "    template_amplification_sites = []\n",
    "    for name, template in zip(names, templates):\n",
    "        for feature in template.features:\n",
    "            if feature.qualifiers['name'] == name:\n",
    "                \n",
    "                CDS_strand = feature.strand\n",
    "                if CDS_strand == 1:\n",
    "                    start = feature.location.start\n",
    "                    print(start)\n",
    "                else:\n",
    "                    end = feature.location.end\n",
    "                    print(end)\n",
    "\n",
    "            if feature.qualifiers['name'].endswith(terminator):\n",
    "                terminator_strand = feature.strand\n",
    "                if terminator_strand == 1:\n",
    "                    end = feature.location.end\n",
    "                    print(end)\n",
    "                    \n",
    "                else:\n",
    "                    start = feature.location.start\n",
    "                    print(start)\n",
    "        \n",
    "        template_amplification_site = template[start:end]\n",
    "        \n",
    "        if 'batches' in template_amplification_site.annotations.keys():\n",
    "            template_amplification_site.annotations[\"batches\"].append(template.annotations['batches'][0]) #{'location': template.annotations['batches'][0]['location']}\n",
    "        else: \n",
    "            template_amplification_site.annotations[\"batches\"] = []\n",
    "            template_amplification_site.annotations[\"batches\"].append(template.annotations) #{'location': template.annotations['batches'][0]['location']}\n",
    "\n",
    "        template_amplification_sites.append(template_amplification_site)\n",
    "    return template_amplification_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "58cee599-0177-4dd4-b2e2-0e82aa363e3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "G8H_template_names = [\"pESC-CEN_RS-OeuG8H\",\"PL_10_A1pMIA_JZ-40_CroG8H\",\"pMIA_086_Rsep_G8O_in_p0057_(TRP)\",\"pMIA_087_Vmin_G8O_in_p0057_(TRP)\",\"pMIA_088_Opum_G8O_in_p0057_(TRP)\",\"pMIA_089_Cacu_G8O_in_p0057_(TRP)\",\"pMIA_090_Smus_G8O_in_p0057_(TRP)\",\"pMIA_091_Ccal_G8O_in_p0057_(TRP)\"]\n",
    "\n",
    "#G8H_templates = [from_benchling(tname) for tname in G8H_template_names]\n",
    "#G8H_CDS_annotations = [\"OeuG8O\", \"CroG8O\", \"Rsep_G8O_opt\", \"Vmin_G8O_opt\", \"Opum_G8O_opt\", \"Cacu_G8O_opt\", \"Smus_G8O_opt\", \"Ccal_G8O_opt\"]\n",
    "\n",
    "\n",
    "\n",
    "#G8H_sites = extract_template_amplification_sites(G8H_templates, G8H_CDS_annotations, \"T ADH1\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28479d2d-76bc-4f79-ac38-30219393824f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "b1a29a82-66a3-423e-bac3-f5c038ede4ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "features1 = [SeqFeature(FeatureLocation(0, 200, strand=1), type='CDS')]\n",
    "features2 = [SeqFeature(FeatureLocation(200, 300, strand=1), type='terminator')]\n",
    "dict1 = {'name': 'ATF'}\n",
    "dict2 = {'name': 'ATF_tTEF'}\n",
    "features1[0].qualifiers = dict1\n",
    "features2[0].qualifiers = dict2\n",
    "\n",
    "TEMPLATES = [SeqRecord(seq = Seq('ATGATGA'*1000), id='seq_DPrPIMvy', name='ATF_plasmid', description='<unknown description>', dbxrefs=[], features = features1+ features2)]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "af453fa7-6aae-45e0-9436-31e366df7cbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7000"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(TEMPLATES[0].seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "cded4e8d-8594-4af5-bf84-e041f1057ae3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "300\n",
      "ATF_plasmid\n",
      "300\n"
     ]
    }
   ],
   "source": [
    "names_of_sites_we_want_to_amplify = ['ATF']\n",
    "name_of_terminator_site_to_be_incorporated= 'tTEF'\n",
    "\n",
    "extractions_sites = extract_template_amplification_sites(TEMPLATES, names_of_sites_we_want_to_amplify, name_of_terminator_site_to_be_incorporated)\n",
    "print(str(extractions_sites[0].name))\n",
    "print(len(extractions_sites[0].seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "810965b5-7368-4c7a-a5c7-75f20bbb1a86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{}]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "extractions_sites[0].annotations['batches']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "3d10fc8a-5059-47af-b67f-a3ec9ecf5406",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_sites(annotations, templates, names):\n",
    "    ''' This function extracts the sequences from annotated sequences based on they names\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    sites = []\n",
    "    for anno, template, name in zip(annotations, templates, names):\n",
    "        for feature in template.features:\n",
    "            site = template[feature.location.start:feature.location.end]\n",
    "            if feature.qualifiers['name'] == anno:\n",
    "                site.name = name\n",
    "\n",
    "            # If there is an batch anotation we can save it \n",
    "            if 'batches' in site.annotations.keys():\n",
    "                site.annotations[\"batches\"].append(template.annotations['batches'][0])\n",
    "            else: \n",
    "                site.annotations[\"batches\"] = []\n",
    "                site.annotations[\"batches\"].append(template.annotations)\n",
    "\n",
    "            sites.append(site) \n",
    "    return(sites)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "78da2388-fa4c-4007-848d-7d2d744057c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pCYC1\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqRecord import SeqRecord\n",
    "from pydna.dseqrecord import Dseqrecord\n",
    "\n",
    "features1 = [SeqFeature(FeatureLocation(2, 100, strand=1), type='Terminator1')]\n",
    "features2 = [SeqFeature(FeatureLocation(200, 300, strand=1), type='Terminator2')]\n",
    "features1[0].qualifiers['name'] =  'tTEF'\n",
    "features2[0].qualifiers['name'] =  'tTEF'\n",
    "\n",
    "prom_seq = Seq('CAGCATTTTCAAAGGTGTGTTCTTCGTCAGACATGTTTTAGTGTGTGAATGAAAATATAGATAGATGATGATGATA'*50)\n",
    "prom_template = [SeqRecord(seq = prom_seq, id='seq_DPrPIMvy', name='pCYC1', description='<unknown description>', dbxrefs=[], features = features1+ features2)]\n",
    "\n",
    "p_annotations = 'pCYC1'\n",
    "p_names = 'pCYC1'\n",
    "\n",
    "promoter_sites =  extract_sites(p_annotations, prom_template, p_names)\n",
    "print(promoter_sites[0].name)\n",
    "rec_vector.add_feature(61,80, label = ['terminator'], type = 'terminator', name = 'tTEF')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb908156-7d4a-47fc-84aa-588f209fd05e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e18b2ad-b047-44dc-8d3b-683991c7b7ab",
   "metadata": {},
   "source": [
    "#####  NOT WORKING RN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "5435f6ec-f194-424d-8bb3-407ee541d576",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extract_template_amplification_site(template, name):\n",
    "    \n",
    "    template_amplification_site = []\n",
    "    for feature in template.features:\n",
    "        if name in feature.qualifiers.get('name',\"\"):\n",
    "            \n",
    "            template_amp = template[feature.location.start:feature.location.end]\n",
    "            template_amp.name = feature.qualifiers.get('name',\"\")\n",
    "            template_amplification_site.append(template_amp)\n",
    "    \n",
    "    return template_amplification_site\n",
    "\n",
    "extract_template_amplification_site(rec_vector, 'ATF')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "b9905296-a21b-46a2-a372-9048f551b82d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../constrain/lims')\n",
    "from benchling_api import from_benchling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "b1af0f85-7675-447c-af13-53ce712263ae",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "991"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pG8H_template_names = ['pCYC1','pENO2','pPCK1','pRPL15B']\n",
    "pG8H_annotations = ['pCYC1','pENO2','pPCK1','pRPL15B']\n",
    "pG8H_names = ['pCYC1','pENO2','pPCK1','pRPL15B']\n",
    "\n",
    "\n",
    "pG8H_templates = [from_benchling(pname) for pname in pG8H_template_names]\n",
    "len(pG8H_templates[0].seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "260d6101-ec8d-4ca3-888b-f084c0464aee",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqFeature(FeatureLocation(ExactPosition(2), ExactPosition(989), strand=1), type='promoter')]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pG8H_templates[0].features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "f04fc8df-2ff9-4e59-bf4b-ff27f4aa8454",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "987"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pG8H_sites1 =  extract_sites(pG8H_annotations, pG8H_templates, pG8H_names)\n",
    "len(pG8H_sites1[0].seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "fa38e813-7c3f-410b-a57d-67306473c471",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pCYC1\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "features1 = [SeqFeature(FeatureLocation(2, 100, strand=1), type='promoter1')]\n",
    "features2 = [SeqFeature(FeatureLocation(200, 300, strand=1), type='promoter2')]\n",
    "features1[0].qualifiers['name'] =  'pCYC1'\n",
    "features2[0].qualifiers['name'] =  'pCYC1'\n",
    "\n",
    "prom_seq = Seq('CAGCATTTTCAAAGGTGTGTTCTTCGTCAGACATGTTTTAGTGTGTGAATGAAAATATAGATAGATGATGATGATA'*50)\n",
    "prom_template = [SeqRecord(seq = prom_seq, id='seq_DPrPIMvy', name='pCYC1', description='<unknown description>', dbxrefs=[], features = features1+ features2)]\n",
    "\n",
    "p_annotations = 'pCYC1'\n",
    "p_names = 'pCYC1'\n",
    "\n",
    "promoter_sites =  extract_sites(p_annotations, prom_template, p_names)\n",
    "print(promoter_sites[0].name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "f9976fcf-1e2a-4306-8a8d-c53bb3734ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def seq_to_annotation(seqrec_from, seqrec_onto, aType):\n",
    "    from Bio.Seq import Seq\n",
    "\n",
    "    ''' Anotate an amplicon object from another amplicon object'''\n",
    "    seq_from = seqrec_from.seq.watson.upper()\n",
    "    seq_onto = seqrec_onto.seq.watson.upper()\n",
    "    \n",
    "    # Find the match\n",
    "    strand = 1\n",
    "    match_index = seq_onto.find(seq_from)\n",
    "        \n",
    "    # if there is match\n",
    "    if match_index !=-1:\n",
    "        start = match_index\n",
    "        end =  start + len(seq_from)\n",
    "    else:\n",
    "        # If no match we look at the reverse complement\n",
    "        seq_onto = Seq(seq_onto)\n",
    "        rev_match_index = seq_onto.reverse_complement().find(seq_from)\n",
    "        \n",
    "        # if we get a match here\n",
    "        if rev_match_index != -1:\n",
    "            strand = -1\n",
    "            reclength = len(str(seqrec_onto.seq))\n",
    "            end = reclength - rev_match_index\n",
    "            start = end - len(seq_from)\n",
    "\n",
    "        else:\n",
    "            print(\"no match! seq:\" + str(seqrec_from.name) + \"\\nnot annealing to:\" + str(seqrec_onto.name))\n",
    "    \n",
    "    # add the feature to the amplicon\n",
    "    feature = Bio.SeqFeature.SeqFeature(Bio.SeqFeature.FeatureLocation(start, end), type=aType, strand=strand)\n",
    "    feature.qualifiers[\"label\"] = seqrec_from.id\n",
    "    seqrec_onto.features.append(feature)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "483d18e0-8e3e-44ac-8520-220854177461",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqFeature(FeatureLocation(ExactPosition(0), ExactPosition(22), strand=1), type='primer_bind'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(116), ExactPosition(142), strand=-1), type='primer_bind')]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template1 = 'TCTTTGAAAAGATAATGTATGATTATGCTTTCACTCATATTTATACAGAAACTTGATGTTTTCTTTCGAGTATATACAAGGTGATTACATGTACGTTTGAAGTACAACTCTAGATTTTGTAGTGCCCTCTTGGGCTAGCGGTAAAGGTGCGCATTTTTTCACACCCTACAATGTTCTGTTCAAAAGATTTTGGTCAAACGCTGTAGAAGTGAAAGTTGGTGCGCATGTTTCGGCGTTCGAAACTTCTCCGCAGTGAAAGATAAATGATCGCCGTAGTAACGTCGCTGTCGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGGTGCTTTTTTTGTTTTTTATGTCTTCGAGTCATGTAATTAGTTAAGTGCAGGT'\n",
    "primerF1 = 'CGTGCGAUTCTTTGAAAAGATAATGTATGA'\n",
    "primerR1 = 'ACCTGCACUTAACTAATTACATGACTCGA'\n",
    "rec_vector1 = pcr(primerF,primerR, template)\n",
    "\n",
    "template2 = 'ATATTTATACAGAAACTTGATGTTTTCTTTCGAGTATATACAAGGTGATTACATGTACGTTTGAAGTACAACTCTAGATTTTGTAGTGCCCTCTTGGGCTAGCGGTAAAGGTGCGCATTTTTTCACACCCTACAATGTTCTG'\n",
    "primerF2 = 'ATATTTATACAGAAACTTGATG'\n",
    "primerR2 = 'CAGAACATTGTAGGGTGTGAAAAAAT'\n",
    "rec_vector2 = pcr(primerF2,primerR2, template2)\n",
    "\n",
    "rec_vector2.features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "63dc7ad9-8dfe-4301-b960-e5eb6c2dac63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dseq(-142)\n",
       "ATAT..TCTG\n",
       "TATA..AGAC"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rec_vector2.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f4e38233-83e3-4e2b-82c6-bedb1c243c4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_to_annotation(rec_vector2, rec_vector1, 'AMPLICON2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "745cf141-3aa9-4410-ac00-300aa7edb3f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "186"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rec_vector1.features[2].location.end.position"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4177e2f7-86bc-4c12-9465-40f2c377bf95",
   "metadata": {},
   "source": [
    "## Test of CASEMBLER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "2fc50e09-a8da-4434-a42f-5136b09f10cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../constrain/lims')\n",
    "from benchling_api import from_benchling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "ad01f14d-de3f-4b82-a923-58d6a3cbeafd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sites = [\"MIA-HA-2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "16a895e3-a5ad-4da2-adaf-60fb921dbd5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "CM3 = [from_benchling(name) for name in sites]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "1484199d-6faa-494a-925d-f8a10cbaf398",
   "metadata": {},
   "outputs": [],
   "source": [
    "ATF1_gRNA = pydna.dseqrecord.Dseqrecord(\"GAATGGGCATGCTCGGCGTA\", name = \"ATF1\")\n",
    "CroCPR_gRNA = pydna.dseqrecord.Dseqrecord(\"GCCGTAGTAACGTCGCTGTC\", name = \"CroCPR\") #CroCPR gRNA-2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "b48153ff-ef71-4467-bf9f-d22f812dccd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pRS413U_P2_PGK1_mKate2 = from_benchling(\"PL_02_E2(pRS413U-P2_PGK1-mKate2)\")\n",
    "pPGK1_Seq_Rv = from_benchling(\"pPGK1_Seq_Rv\", 'Primer')\n",
    "USER_XhoI_Fw = from_benchling(\"USER_XhoI_Fw\", 'Primer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "b073ab8b-328f-4206-880e-fa7f5f0ae4dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "Amplicon(659)"
      ],
      "text/plain": [
       "Amplicon(659)"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATF1_repair_template = pydna.amplify.pcr(pPGK1_Seq_Rv, USER_XhoI_Fw, pRS413U_P2_PGK1_mKate2)\n",
    "ATF1_repair_template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "0c10721e-686b-4173-ad85-ff53d93c3c18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "665806"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create function to download PCR template\n",
    "\n",
    "\n",
    "XIChromosomeSeq      = Bio.SeqIO.read('../data/raw/CENPK113-7d_11.fasta', \"fasta\")\n",
    "XIChromosomeSeq.annotations['batches'] = []\n",
    "XIChromosomeSeq.annotations['batches'].append({'location':\"l4_A01\"})\n",
    "len(XIChromosomeSeq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "a5972b67-a99c-4158-a67a-b55d8ff25efc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seq_OvLTlQju 20-mer:5'-GTGCTTGATTTGCGTCATTC-3'"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "XI_3_up_out_sq = from_benchling(\"XI-3- up-out-sq\", 'Primer')\n",
    "XI_3_down_out_sq = from_benchling(\"XI-3- down-out-sq\", 'Primer')\n",
    "XI_3_up_out_sq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "bbb4bb82-9a3c-4b92-b42c-b216fd906241",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "Amplicon(1449)"
      ],
      "text/plain": [
       "Amplicon(1449)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pydna.amplify import pcr\n",
    "\n",
    "CPR_repair_template = pydna.amplify.pcr(XI_3_up_out_sq,XI_3_down_out_sq, XIChromosomeSeq)\n",
    "CPR_repair_template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "5b3a6fcc-9375-439f-97c3-7f6dac87bb47",
   "metadata": {},
   "outputs": [],
   "source": [
    "parameters = {\n",
    "'bg_strain': CM3,\n",
    "'site_names': [\"X-3\", \"XI-3\"],\n",
    "'gRNAs': [ATF1_gRNA, CroCPR_gRNA],\n",
    "'parts': [ATF1_repair_template,CPR_repair_template],\n",
    "'assembly_limits':[200,400],\n",
    "'assembly_names': [\"X_3_tADH1_P2_pPGK1\", \"XI_3_UP_DW\"],\n",
    "'verbose': False,\n",
    "'to_benchling': False  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "3765e8b2-ea71-4891-8fd3-ab50ee89b8e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "20\n",
      "20\n",
      "659\n",
      "1449\n"
     ]
    }
   ],
   "source": [
    "print(len(CM3))\n",
    "print(len(ATF1_gRNA))\n",
    "print(len(CroCPR_gRNA))\n",
    "print(len(ATF1_repair_template))\n",
    "print(len(CPR_repair_template))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "d0f1caad-3b0d-41a8-a564-8ce86b4fabae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def casembler(bg_strain, site_names=None, gRNAs=None, parts=None, assembly_limits=None, assembly_names=None, verbose=False, to_benchling=False):\n",
    "    \n",
    "\n",
    "    \"\"\"\n",
    "    Simulate in vivo assembly and integration\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    bg_strain      \n",
    "    site_names      list of names                      e.g. [X-3, XI-3]    \n",
    "    gRNAs           list of 20 bp seqrecords           e.g. [ATF1_gRNA, CroCPR_gRNA]\n",
    "    parts           list of list of parts              e.g. [[ATF1_repair_template],[CPR_repair_template]]\n",
    "    assembly_limits list of numbers of bp              e.g. [200,400]\n",
    "    assembly_names  list of names of DNA post assembly e.g. [\"X_3_tADH1_P2_pPGK1\", \"XI_3_UP_DW\"]\n",
    "    verbose         write DNA                          e.g. False\n",
    "    to_benchling    upload DNA                         e.g. False\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    \n",
    "    One dseqrecord\n",
    "    Examples\n",
    "    --------\n",
    "    \"\"\"\n",
    "    \n",
    "    #fix mutable default arguments\n",
    "    if site_names is None:\n",
    "        site_names = []\n",
    "    else:\n",
    "        site_names = site_names\n",
    "    \n",
    "    if gRNAs is None:\n",
    "        gRNAs = []\n",
    "    else:\n",
    "        gRNAs = gRNAs\n",
    "          \n",
    "    if parts is None:\n",
    "        parts = []\n",
    "    else:\n",
    "        parts = parts\n",
    "            \n",
    "    if assembly_limits is None:\n",
    "        assembly_limits = []\n",
    "    else:\n",
    "        assembly_limits = assembly_limits\n",
    "        \n",
    "    if assembly_names is None:\n",
    "        assembly_names = []\n",
    "    else:\n",
    "        assembly_names = assembly_names\n",
    "        \n",
    "    \n",
    "    assemblies = []\n",
    "    for int_no in range(0,len(gRNAs)):  \n",
    "        \n",
    "        gRNA = gRNAs[int_no]\n",
    "        print(gRNA)\n",
    "        site_name = site_names[int_no]\n",
    "        site = extract_sites([site_name][int_no], [bg_strain][int_no], [site_name][int_no])\n",
    "        UP, DW = CAS9_cutting(gRNA, site)\n",
    "        \n",
    "        fragments = [UP] + parts[int_no] +  [DW]\n",
    "        \n",
    "        assembly = pydna.assembly.Assembly(fragments, limit=assembly_limits[int_no]).assemble_linear()[0]\n",
    "        \n",
    "        # sometimes pydna.assembly.Assembly distorts the start, end location of features to become negative which produces and error when printing. This function is created as a workaround. \n",
    "        # CPR assembly gives DW_XI_3 annotation called \"DW_XI_3\" a negative start location.\n",
    "        # A quick workaround is to remove featuere\n",
    "        remove_features_with_negative_loc(assembly)\n",
    "        \n",
    "        assembly.name = assembly_names[int_no]\n",
    "        \n",
    "        assembly_feat = Bio.SeqFeature.SeqFeature(Bio.SeqFeature.FeatureLocation(0, len(assembly), strand=1), type=\"misc_feature\")\n",
    "        assembly_feat.qualifiers['name']= site_names[int_no]\n",
    "        assembly_feat.qualifiers['label']= site_names[int_no]\n",
    "        assembly.features.append(assembly_feat)\n",
    "        \n",
    "        if verbose:\n",
    "            DNAs = [UP] + [assembly] + [DW]\n",
    "            for DNA in DNAs:\n",
    "                DNA.write(\"./\" + DNA.name + \".gb\") # \"../data/processed/\"\n",
    "        \n",
    "        if to_benchling:\n",
    "            utils.to_benchling(assembly, \"to_benchling\")\n",
    "        \n",
    "        assemblies.append(assembly)\n",
    "        \n",
    "    return functools.reduce(lambda x, y: x+y, assemblies)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "116c4eab-4e84-455d-9c75-8bef1dfa40e8",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'int' object has no attribute 'start'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Input \u001b[0;32mIn [96]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m HA1 \u001b[38;5;241m=\u001b[39m \u001b[43mcasembler\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mparameters\u001b[49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[0;32mIn [95]\u001b[0m, in \u001b[0;36mcasembler\u001b[0;34m(bg_strain, site_names, gRNAs, parts, assembly_limits, assembly_names, verbose, to_benchling)\u001b[0m\n\u001b[1;32m     58\u001b[0m site_name \u001b[38;5;241m=\u001b[39m site_names[int_no]\n\u001b[1;32m     60\u001b[0m site \u001b[38;5;241m=\u001b[39m extract_sites([site_name][int_no], [bg_strain][int_no], [site_name][int_no])\n\u001b[0;32m---> 62\u001b[0m UP, DW \u001b[38;5;241m=\u001b[39m CAS9_cutting(\u001b[43mgRNA\u001b[49m\u001b[43m[\u001b[49m\u001b[43mint_no\u001b[49m\u001b[43m]\u001b[49m, site)\n\u001b[1;32m     64\u001b[0m fragments \u001b[38;5;241m=\u001b[39m [UP] \u001b[38;5;241m+\u001b[39m parts[int_no] \u001b[38;5;241m+\u001b[39m  [DW]\n\u001b[1;32m     66\u001b[0m assembly \u001b[38;5;241m=\u001b[39m pydna\u001b[38;5;241m.\u001b[39massembly\u001b[38;5;241m.\u001b[39mAssembly(fragments, limit\u001b[38;5;241m=\u001b[39massembly_limits[int_no])\u001b[38;5;241m.\u001b[39massemble_linear()[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/constrain/lib/python3.8/site-packages/pydna/dseqrecord.py:835\u001b[0m, in \u001b[0;36mDseqrecord.__getitem__\u001b[0;34m(self, sl)\u001b[0m\n\u001b[1;32m    832\u001b[0m answer\u001b[38;5;241m.\u001b[39mseq \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mseq\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__getitem__\u001b[39m(sl)\n\u001b[1;32m    833\u001b[0m \u001b[38;5;66;03m# answer.seq.alphabet = self.seq.alphabet\u001b[39;00m\n\u001b[0;32m--> 835\u001b[0m sl_start \u001b[38;5;241m=\u001b[39m \u001b[43msl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstart\u001b[49m \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m  \u001b[38;5;66;03m# 6\u001b[39;00m\n\u001b[1;32m    836\u001b[0m sl_stop \u001b[38;5;241m=\u001b[39m sl\u001b[38;5;241m.\u001b[39mstop \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(answer\u001b[38;5;241m.\u001b[39mseq)  \u001b[38;5;66;03m# 1\u001b[39;00m\n\u001b[1;32m    838\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlinear \u001b[38;5;129;01mor\u001b[39;00m sl_start \u001b[38;5;241m<\u001b[39m sl_stop:\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'int' object has no attribute 'start'"
     ]
    }
   ],
   "source": [
    "HA1 = casembler(**parameters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9507a54c-d1ac-4091-b8c7-7ca412ca40a0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
