---
title: "DataAnalysis"
date: "`r format(Sys.time(), '%d %B, %Y')`"
knit: (function(inputFile, encoding) {rmarkdown::render(inputFile, encoding = encoding, output_dir = "../reports/") })
output:
  pdf_document:
    latex_engine: xelatex
editor_options: 
  markdown: 
    wrap: 72
---

#Disclaimer

```{r 'setup', echo = FALSE, include=FALSE}
#A large part of this document was provided by Søren Petersen, and the learning outcomes were to understand the code and learn to use R, perform sequencing of the missing top producers, redo data analysis with the new data and redo machine learning with the new data. More data mangadgement and plots have been crated during the thesis. The original file can be found as 06-SDP-LibCharacterisation.Rmd

```

# Document setup

```{r 'setup', echo = FALSE, include=FALSE}
knitr::opts_chunk$set(dev = c('pdf','png'), fig.path='../reports/figures/', fig.align = 'center') #, fig.height = 4, fig.width = 4, fig.cap = " " cache=TRUE,
```

```{r, echo=FALSE, include=FALSE, include=FALSE}
#Import libraries
library(tidyverse)
library(here)
library(ggsci)
theme_set(theme_classic(20))
```

#Functions

```{r}
#Import sanger files from Sørens run
sanger_import_file_pattern_to_tibble <- function(pattern) {
    imported_data <- tibble(abs_path = list.files(here("data","raw","010_Sequencing files","Sanger", "Order_11107492995-1_Results"), full.names = TRUE,   pattern = pattern, recursive = TRUE))
    return(imported_data)
}

#Function to separate column
df_separate_filename_column <- function(df_with_fn_column) {
  df_fn_separated <- df_with_fn_column %>%
    mutate(plate = str_extract(filename, "yp\\d\\d"),
           well = str_extract(filename, "[A-H]\\d?\\d"),
           primer = substring(str_extract(filename, "(_[:alnum:]+){3}$"), 2)) %>%
    # str pad well column
    mutate(row = str_extract(well, "^([a-zA-Z])"),
           col = str_extract(well, "([0-9]+)$"),
           col = str_pad(col, width=2, pad = "0"),
           well = str_c(row,col,"")) %>%
  left_join(primer_trans, by = c("primer"))
  return(df_fn_separated)
}

#function to import data of a specific pattern
sanger_import_file_pattern_to_tibble_second <- function(pattern) {
    imported_data <- tibble(abs_path = list.files(here("data","raw","010_Sequencing files","Sanger", "11107560852-1_Results"), full.names = TRUE,   pattern = pattern, recursive = TRUE))
    return(imported_data)
}


### Count function
# Count the number of times each part occurs in ML
predcount <- function(predicted_genotypes) {
  pred_count_one <- predicted_genotypes %>% 
    ungroup() %>%
    count(predicted_genotypes$'pG8H')%>%
    add_column("part" = 'pG8H')%>%
    left_join(promoters, by = c( `predicted_genotypes$pG8H`="number"))%>%
    select(n, name, part) %>%
    arrange(n) 
  
  pred_count_two <- predicted_genotypes %>% 
    ungroup() %>%
    count(predicted_genotypes$'pCPR')%>%
    add_column("part" = 'pCPR')%>%
    left_join(promoters, by = c( `predicted_genotypes$pCPR`="number"))%>%
    select(n, name, part)%>%
    arrange(n)
  
  #Count CPR and G8H
  pred_count_three <- predicted_genotypes %>% 
    ungroup() %>%
    count(predicted_genotypes$'CPR')%>%
    add_column("part" = 'CPR')%>%
    left_join(cpr_trans_int , by = c(`predicted_genotypes$CPR`='3'))%>%
    select(n, name = homolog, part)%>%
    arrange(n)
  
  pred_count_zero <- predicted_genotypes %>% 
    ungroup() %>%
    count(predicted_genotypes$'G8H')%>%
    add_column("part" = 'G8H')%>%
    left_join(g8h_trans_int, by = c( `predicted_genotypes$G8H`='0'))%>%
    select(n, name = homolog, part)%>%
    arrange(n)
  
  pred_table <- rbind(pred_count_zero, pred_count_one, pred_count_two, pred_count_three)
  
  return(pred_table)
}


#plot the percental distribution
pred_percentage_plot <- function(pred_table, name, number) {
  header = paste('Part distribution of top',number,' from ML', sep = '')
  name = paste(name,'.pdf', sep ='')
  
  #Find percental
  plot <- pred_table%>%
    group_by(part) %>%
    mutate(percent_weight = round(as.numeric(n / sum(n) * 100),1))%>% ungroup() %>%
    group_by(part) %>% mutate(cs = cumsum((percent_weight))) %>% ungroup() %>%
    
    #plot
    ggplot(aes(x = factor(part, level = c('G8H','pG8H','pCPR','CPR')), y=percent_weight, fill = cs)) +
    geom_bar(stat="identity", color = "black") +
    geom_text(aes(y=cs, label=str_c(name,percent_weight,sep=" ")), vjust=1.6, color="black", size=3.5) +
    theme_minimal(20) +
    labs(x = "part", y = "%", title = header)+
    scale_fill_gradient(low = "dodgerblue4", high = "skyblue")+
    theme(legend.position="none", plot.title = element_text(hjust = 0.1))
  
  #Save plot
  #ggsave(plot = plot, here("reports","ch_presentation","figures",name), width = 7, height = 7, units = "in")
  return(plot)
}

#Plot count of part occurances 
pred_count_plot <- function(pred_table, name, number) {
  header = paste('Part distribution of top',number,' from ML', sep = '')
  name = paste(name,'.pdf', sep ='')
  
  #Plot
  plot <- pred_table%>%
    group_by(part) %>% mutate(cs = cumsum((n))) %>% ungroup() %>%
    ggplot(aes(x = factor(part, level = c('G8H','pG8H','pCPR','CPR')), y=n, fill = n)) +
    geom_bar(stat="identity", color = "black") +
    geom_text(aes(y=cs, label=str_c(name,n,sep=" ")), vjust=1.6, color="black", size=3.5) +
    theme_minimal(20) +
    labs(x = "part", y = "count", title = header) +
    scale_fill_gradient(low = "dodgerblue4", high = "skyblue")+
    theme(legend.position="none", plot.title = element_text(hjust = 0.1))

  #Save plot
  #ggsave(plot = plot, here("reports","ch_presentation","figures",name), width = 7, height = 7, units = "in")
  return(plot)
}

```

# Import promoter data

```{r echo=FALSE, fig.height=5, fig.width=7.5, message=FALSE}
#Initialise
level_key <- c(`1` = "high", `2` = "high_low", `3` = "low_high", `4` = "low")
promoters <- read_csv(here("/Users/lucaslevassor/Desktop/g8h_cpr_library 2/data/raw/promoters.csv"))

#Extract promoter data, transpose the dataframe and convert sampling hours to rows
promoters_long <- promoters %>%
  mutate(NIL_6h_FPKM_freq = NIL_6h_FPKM/sum(NIL_6h_FPKM),
          NIL_12h_FPKM_freq = NIL_12h_FPKM/sum(NIL_12h_FPKM),
          NIL_24h_FPKM_freq = NIL_24h_FPKM/sum(NIL_24h_FPKM))  %>%
  select(no_by_pos, position, name, id,NIL_6h_FPKM, NIL_12h_FPKM, NIL_24h_FPKM) %>%
  `colnames<-`(c("no_by_pos", "position", "name", "id", "6", "12", "24")) %>%
  pivot_longer(
     cols = `6`:`24`,
     names_to = "sampling_time",
     values_to = "mRNA_count",
     values_drop_na = TRUE) %>%
  mutate(sampling_hour = as.numeric(sampling_time),
         no_by_pos = as.character(no_by_pos),
         by_pos = recode(no_by_pos, !!!level_key))

#Create a plot with sampling hour against mRNA amount
p_promoters <- promoters_long %>%
  mutate(part = ifelse(position == "1", "G8H", "CPR")) %>%
  rename(Name = name) %>%
  ggplot() +
  geom_line(aes(x=sampling_hour, y = mRNA_count, color = Name)) +
  facet_grid(.~part) +
  scale_fill_viridis_d(alpha = 1, begin = 0, end = 1, direction = 1, option = "D", aesthetics = "fill")+
  #scale_color_jco()+
  scale_x_continuous(breaks = c(6,12,24))+
  xlab("sampling h") +
  ylab("mRNA count") +
  theme(axis.title.x = element_text(),axis.title.y = element_text())

p_promoters
#ggsave(p_promoters,filename=here("reports","ch_presentation","figures","promoter_expression.png"))
```

# Analytics

## Data wrangling

```{r echo=FALSE, message=FALSE, include=FALSE}
#1. In Emacs; Result_table_210330_tmet494_Soren.xlsx to .csv N.D. -> N.A.; N.A. -> NA.
#2. Change names to plates names e.g.Result_table_210330_tmet494_Soren.csv -> Result_table_210330_tmet494_yp48Feed-Nofeed.xlsx
results_1st_ana <- tibble(abs_path = list.files(here("data","raw","023_LC-MS", "all_results_lcms"), full.names = TRUE, pattern = ".csv")) %>%
  #Extract data from filenames
  mutate(filename = tools::file_path_sans_ext(basename(abs_path))) %>%
  separate(filename, into=c("a","b","date","tmetab","description"), sep="_") %>%
  arrange(date) %>%
  mutate(data = map(abs_path,read_csv)) %>%
  unnest(data) %>% 
  select(-abs_path, -a, -b) %>%
  
  #Filter standards and those without production data
  filter(!grepl("*_v13*", sample_notes)) %>%

  # Excel contains values both in ug and uM. Keep uM and remove "Amt_ul_L_".
  #select(!starts_with("Amt_ug_L_"), -...8,-...11, -data_file, -description) %>%
  
  #Fix 210330 sample_notes column
  mutate(sample_notes = gsub("*(_P[12])*","",sample_notes)) %>%
  separate(sample_notes,  sep = "_", into = c('plate','well')) %>%
  arrange(plate, well) %>%
  
  #Rname tmet494 no feed vs feed and fix plate column
  mutate(medium = ifelse(plate=="96WP988","nofeed", ifelse(plate=="96WP989","feed","NA"))) %>%
  mutate(plate = gsub("96WP","",plate),
         plate = gsub("1056","YP49",plate),
         plate = gsub("988|989","YP48",plate),
         plate = gsub("YP","yp",plate)) %>%
  filter(plate == "yp49" | plate == "yp48"| plate == "yp50"| plate == "yp51"| plate == "yp52") %>%
 
  # str pad well column
  mutate(row = str_extract(well, "^([a-zA-Z])"),
         col = str_extract(well, "([0-9]+)$"),
         col = str_pad(col, width=2, pad = "0"),
         well = str_c(row,col,"")) %>%
  select(-row,-col)

#Check those with NA
NA_results_1st_ana <- results_1st_ana %>% filter((is.na(Amt_uM_Loganin) & is.na(Amt_uM_Secologanin) & is.na(Amt_uM_Strictosidine) & is.na(Amt_uM_Tryptamine))) 
  
#Remove NA and replace the rest with 0
results_1st_ana <- results_1st_ana %>%
  filter(!(is.na(Amt_uM_Loganin) & is.na(Amt_uM_Secologanin) & is.na(Amt_uM_Strictosidine) & is.na(Amt_uM_Tryptamine))) %>%
  
  #Replace na's with 0
  mutate(across(starts_with("Amt"), ~replace_na(.x, 0)))

#filter(results, Amt_uM == 0)
```

##Unix code used by Søren to organise data
<!-- ```{bash, echo = FALSE, include=FALSE} -->
<!-- #Merge yp48-52 metadata into single csv file -->

<!-- LCMS="$HOME/projects/g8h_cpr_library/data/raw/023_LC-MS" -->

<!-- procLCMS="$HOME/projects/g8h_cpr_library/data/processed/023_LC-MS" -->

<!-- # convert from xls with tabs to single csv files -->

<!-- ssconvert "$LCMS/plateDesign_yp48-51_MIA-HA-1-17_21-100.xls" "$LCMS/plateDesign.csv" --export-file-per-sheet -->

<!-- mv "$LCMS/plateDesign.csv.0" "$LCMS/yp50.csv" -->

<!-- mv "$LCMS/plateDesign.csv.1" "$LCMS/yp51.csv" -->

<!-- mv "$LCMS/plateDesign.csv.2" "$LCMS/yp49.csv" -->

<!-- mv "$LCMS/plateDesign.csv.3" "$LCMS/yp48.csv" -->

<!-- mv "$LCMS/plateDesign.csv.4" "$LCMS/yp52.csv" -->

<!-- # Add fbname as column in csv -->

<!-- for f in "$LCMS/yp48.csv" "$LCMS/yp49.csv" "$LCMS/yp50.csv" "$LCMS/yp51.csv" "$LCMS/yp52.csv" -->

<!-- do -->

<!--   fbname=$(basename "$f" .csv) -->

<!--   gawk -v d="$fbname" -F"," 'BEGIN {OFS = ","} {$13=d; print}' $f > "$LCMS/Plate$fbname.csv" -->

<!-- done -->

<!-- # create header and append to output file -->

<!-- # replace ",yp49" from header with ",plate" -->

<!-- header=$(head -1 "$LCMS/Plateyp48.csv"| sed 's/.....$//') -->

<!-- header="$header,plate" -->

<!-- echo $header > "$LCMS/yp48-52_meta.csv" -->

<!-- # concatenate files without headers -->

<!-- for f in "$LCMS/Plateyp48.csv" "$LCMS/Plateyp49.csv" "$LCMS/Plateyp50.csv" "$LCMS/Plateyp51.csv" "$LCMS/Plateyp52.csv" -->

<!-- do -->

<!--     echo "Processing $f" -->

<!--     grep -v '^p' $f >> "$LCMS/yp48-52_meta.csv" -->

<!-- done -->

<!-- mv "$LCMS/yp48-52_meta.csv" "$procLCMS/yp48-52_meta.csv" -->

<!-- # cleanup -->

<!-- #rm "$LCMS/plateDesign.csv.0" "$LCMS/plateDesign.csv.1" "$LCMS/plateDesign.csv.2" "$LCMS/plateDesign.csv.3" -->

<!-- rm "$LCMS/yp48.csv" "$LCMS/yp49.csv" "$LCMS/yp50.csv" "$LCMS/yp51.csv" "$LCMS/yp52.csv" -->

<!-- rm "$LCMS/Plateyp48.csv" "$LCMS/Plateyp49.csv" "$LCMS/Plateyp50.csv" "$LCMS/Plateyp51.csv" "$LCMS/Plateyp52.csv" -->

<!-- ``` -->

```{r, echo = FALSE, include=FALSE}
##Plate position information
#Read well information into R
here()
meta <- read_csv(here("data/processed/023_LC-MS/yp48-52_meta_LL.csv")) %>%
  select(plate, well = position, Strain, col_no = `Replicate #`, genotype,medium = Medium) %>%
  mutate(Strain = gsub("MIA-","",Strain),
         genotype = gsub("CPR","",genotype),
         genotype = gsub("G8H","",genotype))
```

```{r,  echo = FALSE, include=FALSE}
#Join meta with strictosidine results 1st ana
results_meta_1st_ana <- meta %>%
  left_join(results_1st_ana, by=c("plate","well")) %>%
  mutate(strain = str_c(Strain,well, str_replace_na(genotype), str_replace_na(col_no), sep="_"),
         Strain = str_c(Strain, str_replace_na(genotype), sep="_")) %>%
  mutate(medium = ifelse(plate == "yp48", medium.y, medium.x)) %>%
  select(-medium.y,-medium.x)

```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
#Tidying data
Results_long_1st_ana <- results_meta_1st_ana %>%
  pivot_longer(
     cols = `Amt_uM_Loganic acid`:`Amt_uM_Strictosidine`,
     names_to = "compound",
     names_prefix = "Amt_uM_",
     values_to = "Amt_uM",
     values_drop_na = TRUE)
```

## Feeding exp: YP48 (Cro strains) - Søren

\*\*\* Experiment 1. Check if MIA-HA-2 - 17 strains produces MIA's with
/ without feeding. We suspect that the base strain MIA-HA-1 has lost GES
and or TDC (outside strictosidine module). If this is the case our
strains will only produce MIA when feed

Strains\
1. YP48

Media feed: 0.2 mM geraniol & 1 mM tryptamine 1. YP + 2 % glucose 2.
YP + 2 % glucose + feed

Cultivation 30C, 300 rpm 1. (Wednesday 24): Inoculate precultures.
Transfer 10 of YP48 g.stock to 90 media 1 (-feed). 2. (Friday 26):
Inoculate maincultures. Transfer 10 of preculture to 500 media 1 and
media 2 3. (Monday 8): For each media, one at a time: 1. Transfer 10 of
mainculture into 140 to measure OD (in incubator while measuring) 2.
Transfer 200 into new 96 well plate, 3. Add 20 uL 10 mg/L caffeine stock
4. Add X ul Std_MM_v13 standards 5. Transfer 220 onto filter in
collection plate, centrifuge (2000 g x 1 min; put collection plate in
-80C) 6. Measure MIAs (extracellular loganin) 4. (TBD): Khem run plate
5. (TBD): Data analysis

\*\*\*\* Analytics notes: Notes from Khem: N:D.: not detected or below
quantification limits; N.A.: not analyzed; Loganin, strictosidine and
tryptamine are quantifiable and others are not quantifiable. Some
samples might have been not integrated because of low abundance of the
peak

```{r,  echo = FALSE, include=FALSE}
#Extract YP48 and tidy
Results_long_norm_yp48 <- Results_long_1st_ana %>%
  filter(plate == "yp48",
         compound == "Strictosidine") %>%
  group_by(Strain, medium, compound) %>%
  summarise(meAmt = mean(Amt_uM, na.rm = TRUE),
            seAmt = sd(Amt_uM, na.rm = TRUE) / sqrt(n())) %>%
  ungroup()


```

```{r  echo=FALSE, fig.height=5, fig.width=5}
#Plot YP48 feeding study
stric_plot <- Results_long_norm_yp48 %>%
  ggplot() +
  geom_bar(aes(x=Strain, y=meAmt), stat="identity", fill="skyblue", alpha=0.7) +
  geom_errorbar( aes(x=Strain, ymin=meAmt-seAmt, ymax=meAmt+seAmt), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  facet_grid(medium ~ compound) +
  theme_classic(15) +
  theme(axis.text.x = element_text(size = 10, angle = 90, vjust = 0.5, hjust=1)) +
  labs(x = "Strain_name", y = "Mean amount in uM +- SE (n=3)") 
stric_plot  
#Save figure
#ggsave(stric_plot,filename=here("reports","presentation","figures","background_mut_test.png"))
```

\*\*\* Conclusion 1. Bg strain mutated & lost genes 2. Promoters course
large variation - continously low (pRPL15b) doesn't work well for G8H
but does for alright for CPR (pURE) 3. Has strain lost tryptamine
synthase or geraniol synthesis genes?

## Production YP49-51 (library strains) - Søren

\*\*\* Experiment: - Strains: 3 colonies from each of 80 library plates.
Each plate has specific combination of G8H and CPR but any of 16
possible promoter combinations - Media: YP + 2 % glucose + feed 0.2 mM
geraniol & 1 mM tryptamine - Sampling: day 6

\*\*\*\* Analytics notes: - More than 500 uM tryptamine in all samples -
No Loganic acid or Secologanin detected in any sample

**Wells not injected/acquired:**

-   Khem is not sure why.
    -   Volume is high enough.
    -   No bubbles as samples where filtered before handover (as well as
        spun down by Khem)
    -   Column clogging due to YPD?

Run 1 - YP49 - A3, B1

Run 2 - YP50 - :A9 A10, B3, B5, B9, C2, C4, C5, C9, D6 - YP51 - :A3, A4,
A9, A11, B7, B9, B10, C9, D3, E4

Run 3:\
Rerun of samples not injected in run 2 as well as control(H7-11) for
both plates rerun - YP50 - C5 and H7-11 - YP51 - A11 and B9

```{r,  echo = FALSE, include=FALSE}
#Normalise results of yp49-51

#Assumptions for the code: 
# Loganing is less interesting as its amount mirrors Strictosidine amount
# Desired outcome: comepare amounts (loganin and strictosidine) between plates.
# Assumption: amounts within one plate is comparable
# Each plate contains the same 4 colonies of MIA-CH-A2
# scale amounts from each plate to MIA-CH-A2 levels

#filter YP49-51
Results_long_norm_yp49_51_1st_ana <- Results_long_1st_ana %>%
  filter(plate == "yp49" | plate == "yp50" | plate == "yp51") %>%
  filter(compound %in% c("Strictosidine")) %>%
  group_by(plate, compound) %>% 
  mutate(Amt_norm = Amt_uM / mean(Amt_uM[well %in% c("H07","H08","H09", "H10")], na.rm = TRUE)*100) %>%
  ungroup()
Results_long_norm_yp49_51_1st_ana
```

```{r echo=FALSE, fig.height=10, fig.width=20}
#Find mean and standard deviations
Results_long_norm_yp49_51_plot <- Results_long_norm_yp49_51_1st_ana %>%
  group_by(strain,compound) %>%
  summarise(Amt_uM_me = mean(Amt_uM, na.rm = TRUE),
            Amt_uM_se = sd(Amt_uM, na.rm = TRUE) / sqrt(n()),
            Amt_norm_me = mean(Amt_norm, na.rm = TRUE),
            Amt_norm_se = sd(Amt_norm, na.rm = TRUE) / sqrt(n())) %>%
  ungroup() %>%
  mutate(control = ifelse(grepl("^CH-A2", strain), "yes",
                          ifelse(grepl("^CM-3", strain), "perhaps", "no")))

#plot 49-51 normalised strictosidine productions
p49_51_analytics <- Results_long_norm_yp49_51_plot %>%
  ggplot() +
  geom_bar(aes(x=strain, y=Amt_norm_me, fill=control), stat="identity", alpha=0.7) + 
  scale_fill_manual(values = c( "yes"="tomato", "perhaps"="grey", "no"="skyblue" ), guide = FALSE ) +
  geom_errorbar(aes(x=strain, ymin=Amt_norm_me-Amt_norm_se, ymax=Amt_norm_me+Amt_norm_se), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_hline(yintercept=100, linetype="dashed", color = "black", size=1) +
  facet_grid(compound~.) +
  theme(axis.text.x = element_text(size = 7, angle = 90, vjust = 0.5, hjust=1)) +
  labs(x = "Strain_name (MIA-_G8H_CPR_colony-number)", y = "Normalised Strictosidine Concentration(uM/uM CH-A2 * 100)")
p49_51_analytics  

#Save plot
#ggsave(p49_51_analytics,filename=here("reports","presentation","figures","g8h_cpr_library_analytics.png"), height = 10, width = 25)
```

```{r}
#Plot top 20 producers
p49_51_top_prod <- top_n(Results_long_norm_yp49_51_plot , n=25, Amt_norm_me) %>%
  ggplot() +
  geom_bar(aes(x=strain, y=Amt_norm_me, fill=control), stat="identity", alpha=0.7) + 
  scale_fill_manual(values = c( "yes"="tomato", "perhaps"="grey", "no"="skyblue" ), guide = FALSE ) +
  geom_errorbar(aes(x=strain, ymin=Amt_norm_me-Amt_norm_se, ymax=Amt_norm_me+Amt_norm_se), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_hline(yintercept=100, linetype="dashed", color = "black", size=1) +
  facet_grid(compound~.) +
  theme(axis.text.x = element_text(size = 14, angle = 90, vjust = 0.5, hjust=1)) +
  labs(x = "Strain_name (MIA-_G8H_CPR_colony-number)", y = "Normalised Strictosidine Concentration(uM/uM CH-A2 * 100)")
p49_51_top_prod
#ggsave(p49_51_top_prod,filename=here("reports","ch_presentation","figures","g8h_cpr_library_analytics_top20.png"), height = 10, width = 20)

```

\*\*\* Results & Discussion:

-   SmusG8H AraCPR_3 higher amounts than CM-3
-   CM-3 Strictosidine amount is drifting from \~40 to \~120 between
    runs while CH-A2 is not. Discuss CM-3 drift with Khem
-   CM-3 mutation increases geraniol uptake?
-   corner well evaporation increases concentration?

1.  Strict \> loganin

-   6 day cultivation

2.  Strains with low amount

-   inactive/incompatible enzymes.
-   Ethanol phase promoters. To surprise, cultivatons show still glucose
    after 6 days. Thus G8H/CPR may be inactivate.

\*\*\* Next steps considerations: 1. Genotyping 2. Modeling and
recommend sublibrary/ies 3. Construct and test more colonies

-   Optimise signal?
    -   sampling time?
    -   media
        -   lower glucose amount, more ethanol?
-   Test more colonies
    -   if a company we would just run 1300 colonies. Academically we
        want to check whether we can find best strain with fewer tests.
        Therefore genotyping ML route.

<!-- Notes 2021-06- 17 -->

<!-- Dear Khem -->

<!-- The data looks very good. We even see one strain that seems to produce more loganin and Strictosidine than our current best (CM-3)! :). -->

<!-- I have one question: -->

<!-- all 3 plates contain: -->

<!-- - 4xMIA-CH-A2 colonies (H07-H10) -->

<!-- - 1xMIA-CM-3 colony (H11) -->

<!-- The 4x CH-A2 consistently produce 30-40 µM across plates -->

<!-- The 1xCM-3 produces from ~40 µM, 70, and 120 between plates -->

<!-- Do you think it makes sense to double check the processing? -->

<!-- Kind regards Søren -->

<!-- Dear three -->

<!-- I have now double checked data with Khem -->

<!-- We focused on caffeine and strictosidine for strains CM-3 and YP50 D3 (highest measurements) -->

<!-- We cannot find any errors in the data processing. -->

<!-- Peaks are well defined and clear and for caffeine they do not very significantly in size from surrounding samples -->

<!-- Khem has suggested to run the CM-3 sample again for all three plates but I do not know whether that makes sense? -->

<!-- [10.49] Jie Zhang -->

<!--     Hi Søren, thanks for the update. if Khem has the time I think it’s better to rum these samples again. I am not sure whether this was an error, but having >130 uM strictosidine in deep well plate with sub-optimal growth medium/condition sounds too high to be true. But on the other hand,I also think we should focus on the library vs CH-A2, which are more comparable. -->

## Prodution YP52 (Selected library strains & controls) - Søren

```{r,  include=FALSE, message=FALSE, echo = FALSE}
#Save keys for 
level_key <- c("HA-71_Cro Cro" = "HA-71_Cro-NA_NA-Cro", "HA-72_Cro Aan" = "HA-72_Cro-NA_NA-Aan","CH-A2_Cro Cro"= "CH-A2_Cro-pTDH3_pTEF1-Cro", "CM-3_Cro Cro"= "CM-3_Cro-pTDH3_pTEF1-Cro")

#Normalise yp52
Results_long_norm_yp52 <- Results_long_1st_ana %>%
  filter(plate == "yp52") %>%
  filter(compound == "Strictosidine") %>%
  mutate(Strain = factor(Strain, levels = c("HA-71_Cro Cro", "HA-72_Cro Aan", "CH-A2_Cro Cro", "CM-3_Cro Cro"))) %>%
  mutate(Strain = recode(Strain, !!!level_key)) %>%
  group_by(Strain, medium, compound) %>%
  summarise(meAmt = mean(Amt_uM, na.rm = TRUE),
            seAmt = sd(Amt_uM, na.rm = TRUE) / sqrt(n())) %>%
  ungroup()
```

```{r  echo=FALSE, message=FALSE, fig.height=5, fig.width=5}
# Plot yp52 productions
p_52 <- Results_long_norm_yp52 %>%
  ggplot() +
  geom_bar(aes(x=Strain, y=meAmt), stat="identity", fill="skyblue", alpha=0.7) +
  geom_errorbar( aes(x=Strain, ymin=meAmt-seAmt, ymax=meAmt+seAmt), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  facet_grid(factor(medium, levels=c('-','g','t','g&t')) ~ .) +
  theme_classic(10) +
  theme(axis.text.x = element_text(size = 10, angle = 25, vjust = 0.5)) + #
  labs(x = "Strain_name", y = "Mean Strictosidine Concentration in uM +- SE (n=3)") 
p_52

# Save plot
#ggsave(p_52,filename=here("reports","figures","HA71-72_CH-A2_CM3.png"), height = 10, width = 10)

```

\*\*\* Conclusion 1. Stable measurements of CM-3 2. My background strain
can synthesize tryptamine but not geraniol. 3. MIA-HA-71 promoters may
improve production compared to MIA-CH-A2 and MIA-CM-3 promoters 4.
AanCPR seems to work better than CroCPR

# Genotyping:

Purpose: Determine promoters from library strains to enable ML based
sampling of library space

Methods: DNA sequencing of assembled clusters. Genomic DNA was extracted
from overnight cultures using the LiOAc/SDS method adapted to a 96-well
microtiter plate format. Each extract was used as a template in 2 PCR
reactions spanning the 2 integrated promoter gene pairs and amplifying
from 2500 to 3200 bp. The PCR products were validated using a LabChip GX
II (Perkin Elmer) and sequenced using PlateSeq PCR Kits (Eurofins)
according to the manufacturer's instructions. From the LabChip results,
a PCR reaction was considered as trusted if it showed a strong band of
the correct size; not trusted if it showed a strong band of the wrong
size, and as no information (NI) gained if it showed a weak or no band.
From the sequencing results, a sequencing reaction was considered as
trusted if it showed an unambiguous sequence of the expected length
(i.e., only limited by length of PCR fragment, stretches of the same
nucleotide in the promoter or of \~1 kb limit of sanger sequencing
reactions), not trusted if it showed an unambiguous sequence of the
expected length with an assembly error, and NI gained if there were no
or bad sequence results. If one or more sequencing results from the same
strain showed double peaks in the promoter region the strain was
considered as a double population. Finally, the promoter was noted as a
failed assembly if either LabChip and or sequencing results were
considered not trusted, as NI if the sequencing result was NI and else
as the promoter predicted by pairwise alignment between sequencing
results and promoter sequence.

## Labchip

[Results folder](/Users/Christine/Documents/Speciale/)

![Labchip
results](/Users/Christine/Documents/Speciale/g8h_cpr_library/data/raw/009_LabChip/YP49_G8H2_NNFCB-LABCHIP_g8h_cpr_library_2021-06-30_12-12-51_Gel%20copy.png)
Søren: Results show that not band in all wells. Take to long to get
perfect results. Cost benefit to low. How many bands before good enough?
How many strain fully sequenced before we have enough and can proceed?
Do we want to pick some wells to resequence. If we resequence where do
we put the new data and make sure to use that instead of the old? Both
in terms of .ab1 files but also in plateseq reports

Thoughts: - When all/most PCR/sanger fail per strain probably template
problem - When single/few fail per strain probably PCR failure

## Sequencing

### Submission

[Documents
folder](/Users/Christine/Documents/Speciale/g8h_cpr_library/data/raw/010_Sequencing%20files/Sanger/Submission)
Eurofins \| SDP_ID \| cassette ------------ \| ------ \| --------
PCR_00020195 \| yp49 \| g8h PCR_00020204 \| yp50 \| g8h PCR_00020205 \|
yp51 \| g8h PCR_00020206 \| yp49 \| cpr PCR_00020207 \| yp50 \| cpr
PCR_00020208 \| yp51 \| cpr

In plateseq kits well h12 is reserved for quality control. In all 6
plates yp49-51 for g8h & cpr well h12 contains medium control which do
not need to be sequenced. Thus no need to rearrange samples.

### Download and movement of files to correct folders

[Eurofins
Results](https://eurofinsgenomics.eu/pages/core/trackorder.aspx?orderRef=11107492995-1&secretRef=9e6a6a47-5a6a-4dde-99e0-f8ada3b2c411)

<!-- ```{bash} -->

<!-- #results archieve 2021-07-03 -->

<!-- # save at "/Users/sorpet/projects/g8h_cpr_library/data/raw/010_Sequencing files/Sanger/Order_11107492995-1_Results" -->

<!-- ## Plate_yp51_cpr.fasta & PlateReport_yp51_cpr.xls not included -->

<!-- https://ws.e-sequence.medigenomix.de/eseq-download/OrderResultArchiveDownload?mId=2&oId=11107492995-1&trackingCode=9e6a6a47-5a6a-4dde-99e0-f8ada3b2c411&gsd=2021-07-08T11%3a24&gsh=CE5EEBA993649A3CE510C909E685670299C20511D0A57E626DC79A1462356D57 -->

<!-- # Upon writing email the following were uploaded on 07-07-2021 -->

<!-- ## Plate_yp51_cpr.fasta  -->

<!-- https://apigateway.eurofinsgenomics.com/download/HttpDownload?get=Z2lkPWViZSZmcD1ZMkV4T1RVNU1qUXRPRFl5TUMwME5USTFMV0l4WldZdFpqSXhOV0kyT0dOak1ETmtMMUJzWVhSbFgzbHdOVEZmWTNCeUxtWmhjM1JoJmNkPTIwMjEtMDctMDhUMTElM2EyNCZleGQ9MjAyMS0wNy0wOFQxMiUzYTI0JnNpZz00NkQ4NjY2N0VGN0MxOTdGQzBENjkxQ0VDQzJDOEVGQUZDRENBMTQ5NkU1MkQzNThFQzcxRkZFMkNEODc4NjI0 -->

<!-- ## PlateReport_yp51_cpr.xls  -->

<!-- https://ws.e-sequence.medigenomix.de/eseq-download/FileDownload?fId=225180530&uuid=1554ee7e-663c-4c9c-891d-39f725fbe1a5&gsd=2021-07-08T11%3a24&gsh=6E44B529B30E38F933D358126559BD93701C3987C5D1A5DB8A9A5E06784B0679 -->

<!-- https://ws.e-sequence.medigenomix.de/eseq-download/FileDownload?fId=225180530&uuid=1554ee7e-663c-4c9c-891d-39f725fbe1a5&gsd=2021-07-08T11%3a24&gsh=6E44B529B30E38F933D358126559BD93701C3987C5D1A5DB8A9A5E06784B0679 -->

<!-- #  Generally happy with results. However, for plate PCR_00020205 almost all sequencing reactions failed (see PlateReport_yp51_g8h.xls). This surprised me since this is the only plate that fails, and I have a gel showing (also for this plate) bands in almost all wells (see attached). Eurofins agreed to resequence -->

<!-- ``` -->

### Look-up tables

```{r}
#Tables that allows transfer from code to name or name to code
primer_trans <- tibble(primer = c("pad_g8h_fw", "pad_cpr_fw"), position = c("1","2"))
primer_trans_2 <- tibble(primer = c("pad_pg8h_fw", "pad_pcpr_fw"), position = c("1","2"))
g8h_trans <- tibble(homolog = c("Cacu", "Opum","Cro","Vmin","Smus","Rsep","Oeu","Ccal"), `0` = c("1","2","3","4","5","6","7","8"))
cpr_trans <- tibble(homolog = c("Cro", "Aan","Ara","Clo","Rse","Ahu","Ani","Cac","Oeu","Cpo"), `3` = c("1","2","3","4","5","6","7","8","9","10"))
pg8h_trans <- tibble(homolog = c("CYC1", "ENO2","PCK1","RPL15B"), `1` = c("1","2","3","4"))
pcpr_trans <- tibble(homolog = c("CCW12", "TPI1","MLS1","URE2"), `2` = c("5","6","7","8"))
position_trans <- tibble(cassette = c("g8h", "cpr"), position = c("1","2"))
features = rbind(g8h_trans,cpr_trans %>% select(homolog, `0`=`3`))


#Get convertion list in integers
g8h_trans_int <- tibble(homolog = c("Cacu", "Opum","Cro","Vmin","Smus","Rsep","Oeu","Ccal"), `0` = c(1,2,3,4,5,6,7,8))
cpr_trans_int <- tibble(homolog = c("Cro", "Aan","Ara","Clo","Rse","Ahu","Ani","Cac","Oeu","Cpo"), `3` = c(1,2,3,4,5,6,7,8,9,10))
pg8h_trans_int <- tibble(homolog = c("CYC1", "ENO2","PCK1","RPL15B"), `1` = c(1,2,3,4))
pcpr_trans_int <- tibble(homolog = c("CCW12", "TPI1","MLS1","URE2"), `2` = c(5,6,7,8))
```

### Infer promoter by alignment

```{r}
#take 010_Sequencing files of promoter region and determine promoter through alignment with promoter sequences
#This function is slow thus we cache results
cached_sanger_files_path <- here("data","interim","010_Sequencing files","sanger_files_fromR.csv")
if (!file.exists(cached_sanger_files_path)) {
  sanger_files <- sanger_import_file_pattern_to_tibble(".ab1") %>%
   mutate(filename = tools::file_path_sans_ext(basename(abs_path))) %>%
   df_separate_filename_column() %>%
   mutate(sangerseq_obj = map(abs_path, ~ sangerseqR::readsangerseq(.x)),
         prim_seq = map_chr(sangerseq_obj, ~ sangerseqR::primarySeq(.x, string = TRUE)),
         nchar = map_dbl(prim_seq, nchar),
         p_align_obj = map(prim_seq, ~ Biostrings::pairwiseAlignment(promoters$sequence_5_to_3, .x, type='local')),
         p_align_scores = map(p_align_obj, ~ BiocGenerics::score(.x)),
         p_align_hscore = map_dbl(p_align_scores, ~ max(.x)),
         inf_promoter = map_dbl(p_align_scores, ~ which.max(.x))) %>%
  select(plate, well, position, nchar, p_align_hscore, inf_promoter)
  write_csv(sanger_files,cached_sanger_files_path)
} else {
sanger_files <- read_csv(cached_sanger_files_path)
}
```

```{r}
#Extract quality information and change structure
plate_reports <- sanger_import_file_pattern_to_tibble(".xls") %>%
  mutate(data = map(abs_path,readxl::read_excel, na = "n.a.", skip = 5)) %>%
  unnest(data) %>%
  select(filename = `Sample-Name`, used, avg_qual = AvgQual, length = Length) %>%
  mutate(plate = str_extract(filename, "yp\\d\\d"),
        well = str_extract(filename, "[A-H]\\d?\\d"),
        primer = substring(str_extract(filename, "(_[:alnum:]+){3}$"), 2))%>%
  mutate(row = str_extract(well, "^([a-zA-Z])"),
           col = str_extract(well, "([0-9]+)$"),
           col = str_pad(col, width=2, pad = "0"),
           well = str_c(row,col,"")) %>%
  left_join(primer_trans, by = c("primer")) %>%
  select(plate, well, position, length, used, avg_qual)
```

### Combine data and check data

```{r}
#Filter away all promoters with sequencing not above critical criteria
sanger_data_by_position <- sanger_files %>%
  mutate(position = as.character(position)) %>% # coerce to character to enable left_join
  left_join(plate_reports, by = c("plate","well","position")) %>%
  mutate(inf_promoter = ifelse(used < 50 | avg_qual < 25 | is.na(avg_qual), NA , inf_promoter)) #Criterias)

#Check if the allignment determined a promoter that should not be present at that location?
sanger_data_by_position %>%
  filter((position == "1" & (inf_promoter < 1 | inf_promoter > 4)) |
           (position == "2" & (inf_promoter <  5 | inf_promoter >  8)))

#Change the wrong promoters
sanger_data_by_position %>%
  mutate(inf_promoter = ifelse(position == "1" & (inf_promoter < 1 | inf_promoter > 4), 1,
                        ifelse(position == "2" & (inf_promoter <  5 | inf_promoter >  8), 5,inf_promoter)))

#Test if ab1 file and plate report match
sanger_data_by_position %>% mutate(test = nchar - length) %>% filter(test != 0) #No mismatch

#Can we determine which reads gives na inf_promoter?
sanger_data_by_position %>% filter(is.na(inf_promoter)) %>% arrange(desc(avg_qual)) #3 sequences have avg qual higher than 25
sanger_data_by_position %>% filter(is.na(inf_promoter)) %>% arrange(desc(used)) #No used is longer than 49
sanger_data_by_position %>% filter(is.na(inf_promoter)) %>% arrange(desc(length)) #the longest length is 1316

```

### Sanger data by well and Strain info

```{r}
#Rearrange data 
sanger_data_by_well <- sanger_data_by_position %>% 
    select(plate,well,position, inf_promoter) %>%
    spread(position, inf_promoter) %>% 
    arrange(plate,well) %>%
    left_join(strain_info, by = c("plate","well")) 

#Combine meta information with code
strain_info <- meta %>%
  filter(plate %in% c("yp49","yp50","yp51")) %>%
  select(plate,well,genotype) %>%
  separate(genotype, sep = " ", into = c("g8h","cpr")) %>%
  left_join(g8h_trans, by = c("g8h" = "homolog")) %>%
  left_join(cpr_trans, by = c("cpr" = "homolog")) %>%
  select(-g8h,-cpr) #%>%
```

### Plotting

```{r}
#plot sequencing length used vs the quality
sanger_qual_vs_used <- here("reports","ch_presentation","figures", "sanger_qual_vs_used.png")
if (!file.exists(sanger_qual_vs_used)) {
  p <- sanger_data_by_position %>%
  mutate(avg_qual = ifelse(is.na(avg_qual), 0, avg_qual), 
         used = ifelse(is.na(used), 0, used)) %>%
  ggplot(aes(x = used, y = avg_qual)) +
  geom_point() +
  theme_classic(20) +
  ggtitle("Sequencing quality")
  #ggsave(p,filename=sanger_qual_vs_used, height = 10, width = 10)
}
```

```{r}
#For how many strains have we sequenced all promoters?
seq_pl_data <- sanger_data_by_position %>% 
  filter(!is.na(inf_promoter)) %>% 
  count(plate, well) %>% 
  count(n,name = "nn") 

#Plot reads count pr strain
p <-ggplot( seq_pl_data, aes(n, nn), stat="identity") +
    geom_col() +
    geom_text(aes(label = nn, group = n), stat = 'summary', fun = sum, vjust = 2)+
    labs(title = "sequencing status", x = "no. of good sequencing reads", y = "no. of strains")
p
ggsave(p, filename = here("reports","figures","sequencing_status.pdf"), width = 5, height = 5, units = "in")
```

## Detailed checking:

**Detailed overview by seq to ref in CLC** - For each library strain
make clc assembly - Assemble sanger data to XI_2 contig of each strain
in the library. - Make folder that contains XI_2 contigs (or best guess
of) of for each library strain (identified by well in yp49-51) - 1. Make
folder with every possible contig (possible_genotypes_by_promoter). - 2.
For each well in yp49-51(/library strain), determine contig (or best
guess of; determined in this rscript, uses sanger results and promoter
list to determine promoter, homolog combination is known from
transformation reaction) - 3. copy and rename that particular contig
from /possible_genotypes_by_promoter to /sampled_genotypes_by_well.
rename that contig from by_promoter name to by_well name, e.g. 8_4\_4_10
to yp49_A01... etc. - 5. Add sanger .ab1 files to CLC - 6. Batch
assemble sequence to reference (in CLC) - 7. Go through assemblies to
double check our sequencing conclusions are correct.

### 1 possible_genotypes_by_promoter

-   Output from jupyterlab 05-SDP-AllLibraries for first transformation
    round

### 3.,4. copy and rename contigs

```{r}

reCopyAndRename = FALSE

if (reCopyAndRename) {
  #Create combination code pr well
  promotercomb_to_well_lookup <- sanger_data_by_well %>%
  mutate(`0` = ifelse(is.na(`0`),1,`0`),
         `1` = ifelse(is.na(`1`),1,`1`),
         `2` = ifelse(is.na(`2`),5,`2`),
         `3` = ifelse(is.na(`3`),1,`3`),
         promotercomb = str_c(`0`,`1`, `2`, `3`, sep="_"),
         well_id = str_c(plate,well, sep="_")) %>%
  select(promotercomb,well_id)

  # subtract 4 from every cpr promoter no. to match jupyter notebook output file formatiing
  promotercomb_to_well_lookup <- promotercomb_to_well_lookup %>%
    separate(promotercomb, sep="_", into = c("g","pg","pc","c")) %>%
    mutate(pc = as.numeric(pc) - 4) %>%
    mutate(promotercomb = str_c(g, pg, pc, c, sep="_")) %>%
    select(-g, -pg, -pc,-c)
  
  #Create clc files of expected sequences
  source_folder <- here("data", "clc_g8h_cpr_lib", "Y.Strains", "possible_genotypes_by_promoter")
  distination_folder <- here("data", "clc_g8h_cpr_lib", "Y.Strains", "sampled_genotypes_by_well")
  for (row in 1:nrow(promotercomb_to_well_lookup)) {
    name_by_genotype <- paste0(promotercomb_to_well_lookup[[row,2]],".clc")
    name_by_well <- paste0(promotercomb_to_well_lookup[[row,1]],".clc")
    file.copy(from = str_c(source_folder,name_by_genotype,sep = "/"), to = str_c(distination_folder,name_by_well,sep = "/"))
  }
}

```

### 5. Add sanger .ab1 files to CLC

```{bash}
# These clc_g8h_cpr_lib is a folder in CLC
#cd /Users/sorpet/projects/g8h_cpr_library/data/clc_g8h_cpr_lib/sequencing
#mkdir ./yp49_g8h_PCR_00020195
#mkdir ./yp50_g8h_PCR_00020204
#mkdir ./yp51_g8h_PCR_00020205
#mkdir ./yp49_cpr_PCR_00020206
#mkdir ./yp50_cpr_PCR_00020207
#mkdir ./yp51_cpr_PCR_00020208

# isolate .ab1 files for easier import into CLC
#cd /Users/sorpet/projects/g8h_cpr_library/data/raw/010_Sequencing files/Sanger/Order_11107492995-1_Results/
#mkdir ./plate_yp49_g8h/ab1
#mkdir ./plate_yp50_g8h/ab1
#mkdir ./plate_yp51_g8h/ab1
#mkdir ./plate_yp49_cpr/ab1
#mkdir ./plate_yp50_cpr/ab1
#mkdir ./plate_yp51_cpr/ab1

#mv ./plate_yp49_g8h/*.ab1  ./plate_yp49_g8h/ab1
#mv ./plate_yp50_g8h/*.ab1  ./plate_yp50_g8h/ab1
#mv ./plate_yp51_g8h/*.ab1  ./plate_yp51_g8h/ab1
#mv ./plate_yp49_cpr/*.ab1  ./plate_yp49_cpr/ab1
#mv ./plate_yp50_cpr/*.ab1  ./plate_yp50_cpr/ab1
#mv ./plate_yp51_cpr/*.ab1  ./plate_yp51_cpr/ab1

# from CLC import ab1 files
```

### 6 assemble seq to ref

#### 6.1 check

use [mailed workflow from CLC
support](/Users/sorpet/projects/g8h_cpr_library/references/assembling%20sequences%20to%20references%20CLC.pdf)
\#### 6.2. assemble_file create assemble_file (produced in this script)

```{r}
#Create lookup table for sequencing files
pre_CLC_assembly_lookup <- sanger_data_by_position %>%
    mutate(inf_promoter = ifelse(is.na(inf_promoter) & position == "1", 1,
                          ifelse(is.na(inf_promoter) & position == "2", 5,inf_promoter))) %>%
  select(plate,well,position,inf_promoter) %>%
  spread(position, inf_promoter) %>%
  mutate(ref = str_c(`1`,`2`, sep="_")) %>%
  gather(position,inf_promoter,-plate,-well,-ref) %>%
  select(plate,well,position,inf_promoter,ref) %>%
  arrange(plate,well) %>%
  
  # remove well string padding to match Eurofins .ab1 file names
  mutate(row = str_extract(well, "^([a-zA-Z])"),
         col = str_extract(well, "([0-9]+)$"),
         col = str_remove(col, "^0+"),
         well_non_padded = str_c(row,col,"")) %>%
  select(-row,-col) %>%
  
  #Create ID, sample and type column
  left_join(primer_trans, by = c("position")) %>%
  left_join(position_trans, by = c("position")) %>%
  mutate(ID = str_c(plate,cassette,well_non_padded,primer,sep = "_"),
         Sample = str_c(plate,well,sep = "_"),
         Type = "sample") %>%
  select(ID,Sample,Type)

#Create table for CLC lookup
#insert type = reference lines in 
CLC_assembly_lookup <- as.data.frame(pre_CLC_assembly_lookup)
New = c()
for (i in rev(seq(from=2, to=500, by=2))) {
  New = c(CLC_assembly_lookup[[i,"Sample"]], CLC_assembly_lookup[[i,"Sample"]], "reference")

  CLC_assembly_lookup <- DataCombine::InsertRow(CLC_assembly_lookup, NewRow = New, RowNum = i+1)
}
CLC_assembly_lookup <- CLC_assembly_lookup  %>% drop_na() 

#save table
#openxlsx::write.xlsx(CLC_assembly_lookup, file = here("data","processed","010_Sequencing files","CLC_assembly_lookup.xlsx"))

```

### 7. Detailed check of assemblies

-   Manuel filtering. trusted / not_trusted ab1.
-   For certains strains multiple ab1 sequences get ambiguous only in
    promoters -\> double population
-   ab1 sequences quality drops preliminary due to stretches of same
    nucleotides in certain promoters -\> trusted even though small seq
    alignment
-   Low quality ab1 -\> not_trusted

# Christine: Manual ab1 and promoter check

```{r}
#Create excel sheet for checking
sanger_check_path <- here("data","interim","sequence_check2.xlsx")

if (!file.exists(sanger_check_path)){
  sanger_check <- sanger_data_by_position %>% 
    select(plate, well, position, inf_promoter) %>%
    add_column(ab1 = NA) %>%
    add_column(ab1_comment = NA) %>%
    add_column(gel = NA) %>%
    add_column(gel_comment = NA)
  index <- with(sanger_check, order(plate, well))
  sanger_check <- sanger_check[index, ]
  openxlsx::write.xlsx(sanger_check, file = here("data","interim","sequence_check2.xlsx"))
} else {
  sequence_check<-readxl::read_excel(here("data","interim", "sequence_check2.xlsx"))
}
```

## Manual check vs criteria fitering

```{r}
#List those that had not been accepted
sanger_data_by_position %>%
   filter(plate == 'yp49' | plate == 'yp50' | plate == 'yp51') %>%
   left_join(sequence_check, b =c("plate", "well", "position")) %>%
   mutate(inf_promoter.y = ifelse(is.na(inf_promoter.y), "False", inf_promoter.y))%>%
   filter(inf_promoter.y != "False" & Keep == 'False')

#Combine sequence quality and manual control and keep those following acceptance crit.
seq_info <- sanger_data_by_position %>%
   filter(!well %in% c("H07","H08","H09","H10","H11","H12")) %>%
   left_join(sequence_check, b =c("plate", "well", "position")) %>%
   mutate(inf_promoter.y = ifelse(used < 50 | avg_qual < 25 | is.na(avg_qual) |Keep == FALSE | Keep == "False", NA , inf_promoter.y))


seq_info %>%
  filter(!is.na(inf_promoter.x) & is.na(inf_promoter.y)) #11 were different after manual check
  
```

```{r}
#### Check sequencing distribution
#filter data
sequences_by_well_manual_checked <- seq_info %>%
    rename(inf_promoter = inf_promoter.y) %>%
    mutate(inf_promoter = ifelse((Keep == "True"|Keep == TRUE), inf_promoter, NA)) %>%
    select(plate, well, position, inf_promoter) %>% 
    filter(plate == 'yp49' | plate == 'yp50' | plate == 'yp51') %>%
    spread(position, inf_promoter) %>% 
    arrange(plate,well) %>%
    left_join(strain_info, by = c("plate","well"))

#Find resuls of sequencing
nrow(sequences_by_well_manual_checked) #240 in total
sum(is.na(sequences_by_well_manual_checked$`1`)&is.na(sequences_by_well_manual_checked$`2`), na.rm =TRUE) #11 without any promoter seq
sum(is.na(sequences_by_well_manual_checked$`1`)& !(is.na(sequences_by_well_manual_checked$`2`)), na.rm =TRUE) # 11 without 1 promoter
sum(!(is.na(sequences_by_well_manual_checked$`1`)) & is.na(sequences_by_well_manual_checked$`2`), na.rm =TRUE) # 66 without 2 promoter
nrow(sequences_by_well_manual_checked%>% drop_na()) #152 have both

```

```{r}
##### Plot new sequencing status
#For how many strains have we sequenced all promoters?
seq_pl_data_2 <- seq_info %>% 
  filter(Keep == "True") %>% 
  count(plate, well) %>% 
  count(n,name = "nn") 


#Plot reads count pr strain
p <-ggplot( seq_pl_data_2, aes(n, nn), stat="identity", fill = seq_pl_data_2$n) +
    geom_col() +
    geom_text(aes(label = nn, group = n), stat = 'summary', fun = sum, vjust = 2)+
    labs(title = "Sequencing status", x = "No. of good sequencing reads", y = "No. of strains")+
    theme(plot.title = element_text(hjust = 0.5))
p
ggsave(p, filename = here("reports","ch_presentation","figures","sequencing_status2.pdf"), width = 5, height = 5, units = "in")

```

## Merge strict and seq

```{r}
#merge rows from same well and filter NA - end with 157 strains
strict_1st_ana <- Results_long_norm_yp49_51_1st_ana %>% select('plate', 'well', 'Amt_norm', 'Amt_uM')

nrow(strict_1st_ana %>% filter(!well %in% c("H07","H08","H09","H10","H11","H12"))) #235 of the 240 strains had analytics results

#Combine strictosidine and sequencing data
sequences_strict_1st_ana <- sequences_by_well_manual_checked %>% 
    left_join(strict_1st_ana, by = c("plate","well")) 

#Remove all data missing a seq or strictodisine
sequences_strict_NO_NA_1st_ana <- sequences_strict_1st_ana %>%
    drop_na() 
nrow(sequences_strict_NO_NA_1st_ana) #149 had strict and seq data

```

### Investigate the non included strains

```{r}
#Take strictosidine data and rearrange
res_1st_ana<- Results_long_norm_yp49_51_1st_ana %>%
  group_by(strain,compound, plate, well) %>%
  summarise(Amt_uM_me = mean(Amt_uM, na.rm = TRUE),
            Amt_uM_se = sd(Amt_uM, na.rm = TRUE) / sqrt(n()),
            Amt_norm_me = mean(Amt_norm, na.rm = TRUE),
            Amt_norm_se = sd(Amt_norm, na.rm = TRUE) / sqrt(n())) %>%
  ungroup() %>%
  mutate(control = ifelse(grepl("^CH-A2", strain), "yes",
                          ifelse(grepl("^CM-3", strain), "perhaps", "no"))) %>%
  left_join(sequences_strict , by = c('plate','well')) %>%
  left_join(g8h_trans, by= '0') %>%
  rename('g8h' = homolog ) %>%
  left_join(pg8h_trans_int, by= '1') %>%
  rename('pg8h' = homolog) %>%
  left_join(pcpr_trans_int, by= '2')%>%
  rename('pcpr' = homolog) %>%
  left_join(cpr_trans, by= '3') %>%
  rename('cpr' = homolog) %>%
  select(strain, plate, well, Amt_norm_me, Amt_norm_se, control, g8h, pg8h, pcpr, cpr,compound)

#Create column to tell if result are used or not
res_1st_ana$ID <- paste(res_1st_ana$plate, "_",res_1st_ana$well)  
sequences_strict_NO_NA_1st_ana$ID <- paste(sequences_strict_NO_NA_1st_ana$plate, "_",sequences_strict_NO_NA_1st_ana$well)  
res_1st_ana$included <- ifelse(res_1st_ana$ID %in% sequences_strict_NO_NA_1st_ana$ID, 'yes','no')

#find top 20 strictodisine producers
top_20_1st_ana  <- res_1st_ana[(res_1st_ana$control=="no"),] %>%
  top_n(n=20, Amt_norm_me) %>%
  arrange(desc(Amt_norm_me))

#Those not included
top_20_not_included_1st_ana <- top_20_1st_ana%>%
  filter(top_20_1st_ana$included == 'no')

nrow(top_20_not_included_1st_ana)##The 10 strains from top 10 all have 1 prom sequenced, and this is one of the included promoters

#Find top 50
top_50_1st_ana  <- res_1st_ana[(res_1st_ana$control=="no"),] %>%
  top_n(n=50, Amt_norm_me)%>%
  arrange(desc(Amt_norm_me))
  
#Those not included
top_50_not_included_1st_ana <- top_50_1st_ana%>%filter(top_50_1st_ana$included == 'no')
nrow(top_50_not_included_1st_ana) #18 are not included

#Find top 20 not sequenced
not_incl_top_1st_ana <- res_1st_ana[(res_1st_ana$control=="no"),]%>%
  filter(included == 'no')%>%
  top_n(n=20, Amt_norm_me)%>%
  arrange(desc(Amt_norm_me))
```

```{r}
# Plot strict data of top 20, and divde based on included or not
p_top20 <- top_20_1st_ana%>%
  ggplot() +
  geom_bar(aes(x=strain, y=Amt_norm_me, fill=included, order = Amt_norm_me), stat="identity", alpha=0.7) + #  # , fill = run
    scale_fill_manual(values = c( "no"="tomato", "yes"="skyblue" ),  labels = c("Not included", "Included") ) +
  geom_errorbar(aes(x=strain, ymin=Amt_norm_me-Amt_norm_se, ymax=Amt_norm_me+Amt_norm_se), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_hline(yintercept=100, linetype="dashed", color = "black", size=1) +
  facet_grid(compound~.) +
  theme(axis.text.x = element_text(size = 14, angle = 90, vjust = 0.5, hjust=1)) +
  labs(x = "Strain_name (MIA-_G8H_CPR_colony-number)", y = "Normalised Strictosidine Concentration(uM/uM CH-A2 * 100)", title = "Top 20 strictodisine producers")
p_top20
#Save plot
#ggsave(p_top20,filename=here("reports","ch_presentation","figures","g8h_cpr_library_top_20_includ_notinclud.png"), height = 10, width = 20)
```

\*\*\* Conclusion: We see that there are 10 out of the 20 that are not
included to ML! Therefore the top 20 producers that were not included
should be resequenced.

## Second sequencing

#### Infer promoter by alignment

```{r}
#Get seq data overview
path <- here('data', 'raw','010_Sequencing files', 'Mix2Seq_EF0000042174.xls')
tube_code <-readxl::read_excel(path) %>%
  select(tube_code, primer, plate, well, Sequencing) %>%
  filter(!is.na(Sequencing))


#take 010_Sequencing files of promoter region and determine promoter
cached_sanger_files_path_2 <- here("data","interim","010_Sequencing files","sanger_files_200821.csv")
if (!file.exists(cached_sanger_files_path_2)) {
  sanger_files_2 <- sanger_import_file_pattern_to_tibble_second(".ab1") %>%
   mutate(filename = tools::file_path_sans_ext(basename(abs_path))) %>%
   separate(filename, into = c('file1','file2'), sep ='_') %>%
   left_join(tube_code, by = c('file1'='tube_code')) %>%
   filter(Sequencing == 'good')  %>%
   left_join(primer_trans_2, by = c("primer"))%>%
   mutate(sangerseq_obj = map(abs_path, ~ sangerseqR::readsangerseq(.x)),
         prim_seq = map_chr(sangerseq_obj, ~ sangerseqR::primarySeq(.x, string = TRUE)),
         nchar = map_dbl(prim_seq, nchar),
         p_align_obj = map(prim_seq, ~ Biostrings::pairwiseAlignment(promoters$sequence_5_to_3, .x, type='local')),
         p_align_scores = map(p_align_obj, ~ BiocGenerics::score(.x)),
         p_align_hscore = map_dbl(p_align_scores, ~ max(.x)),
         inf_promoter = map_dbl(p_align_scores, ~ which.max(.x))) %>%
  select(plate, well, position, nchar, p_align_hscore, inf_promoter)
  write_csv(sanger_files_2,cached_sanger_files_path_2)
} else {
sanger_files_2 <- read_csv(cached_sanger_files_path_2)
}
```

```{r}
### Combine data
#Check if there are any mismatches  
sanger_files_2 %>%
  mutate(position = as.character(position)) %>% filter((position == "1" & (inf_promoter < 1 | inf_promoter > 4)) | (position == "2" & (inf_promoter <  5 | inf_promoter >  8))) # No problems 

#Change format of new seq, to allow substitution
sanger_files_2 <- sanger_files_2 %>%
  mutate(length = 50, used = 50, avg_qual = 50, Keep = "True") 
sanger_files_2$ID <- paste(sanger_files_2$plate, "_",sanger_files_2$well, "_",sanger_files_2$position) 
sanger_files_2$position <- as.character(sanger_files_2$position )

#Change format to match new seq
reseq <- seq_info  %>% 
  select(plate, well, position, nchar, p_align_hscore, inf_promoter = inf_promoter.y, length, used, avg_qual, Keep) 
reseq$ID <-  paste(reseq$plate, "_",reseq$well, "_",reseq$position) 

#Substitue new sequencing results with old
reseq[match(sanger_files_2$ID, reseq$ID), ] <- sanger_files_2
reseq 
```

```{r}
#### Sanity checks
# Has the allignment determined a promoter that should not be present at that location?
reseq %>%
  filter((position == "1" & (inf_promoter < 1 | inf_promoter > 4)) |
           (position == "2" & (inf_promoter <  5 | inf_promoter >  8)))

reseq %>% filter(Keep == "True" & is.na(inf_promoter))

reseq %>% filter(Keep == "False" & (!is.na(inf_promoter)))
#sanger_qual_vs_used <- here("reports","figures", "sanger_qual_vs_used_2.pdf")
```

### Check sequencing distribution

```{r}
#filter data
sequences_by_well_re <- reseq %>%
    mutate(inf_promoter = ifelse((Keep == "True"|Keep ==TRUE), inf_promoter, NA)) %>%
    select(plate, well, position, inf_promoter) %>% 
    filter(plate == 'yp49' | plate == 'yp50' | plate == 'yp51') %>%
    spread(position, inf_promoter) %>% 
    arrange(plate,well) %>%
    left_join(strain_info, by = c("plate","well"))



#Find resuls of sequencing
nrow(sequences_by_well_re) #240 in total
sum(is.na(sequences_by_well_re$`1`)&is.na(sequences_by_well_re$`2`), na.rm =TRUE) #10 without any promoter seq
sum(is.na(sequences_by_well_re$`1`)& !(is.na(sequences_by_well_re$`2`)), na.rm =TRUE) # 9 without 1 promoter
sum(!(is.na(sequences_by_well_re$`1`)) & is.na(sequences_by_well_re$`2`), na.rm =TRUE) # 52 without 2 promoter
nrow(sequences_by_well_re%>% drop_na()) #169 have both

test <- sequences_by_well_re %>% drop_na() %>% unite(MERGED, "1":"3")
length(unique(test$MERGED)) #160 unique strains

```

```{r}
##### Plot new sequencing status
#For how many strains have we sequenced all promoters?
seq_pl_data_3 <- reseq%>% 
  filter(Keep == "True") %>% 
  count(plate, well) %>% 
  count(n,name = "nn") 
  

#Plot reads count pr strain
p <-ggplot( seq_pl_data_3, aes(n, nn), stat="identity") +
    geom_col() +
    geom_text(aes(label = nn, group = n), stat = 'summary', fun = sum, vjust = 2)+
    labs(title = "sequencing status", x = "no. of good sequencing reads", y = "no. of strains")+
    labs(title = "Sequencing status", x = "No. of good sequencing reads", y = "No. of strains")
    theme(plot.title = element_text(hjust = 0.5))
p
#ggsave(p, filename = here("reports","ch_presentation","figures","sequencing_status3.pdf"), width = 5, height = 5, units = "in")


# Can we determine which reads gives na inf_promoter?
sanger_qual_vs_used <- here("reports","figures", "sanger_qual_vs_used_1st_iteration.png")
palette_th <-  c( "#2B66AC","#BDD7EE", "#29386F", "#B4C7E7", "#097394","#67A7BC","#563B79","#715B8F","#BBBBBB")

plot_data <- reseq
plot_data$avg_qual=as.numeric(plot_data$avg_qual)
if (!file.exists(sanger_qual_vs_used)) {
  p_check <- plot_data %>%
  mutate(Included = ifelse((Keep == "True" | Keep == TRUE), "Yes", "No")) %>%
  ggplot(aes(x = used, y = avg_qual, color = Included)) + 
  geom_point() +
  scale_color_jco(alpha = 0.5)+
  #scale_color_manual(values=c("#67A7BC","#2B66AC"))+
  scale_y_continuous(name = "Average quality", breaks = pretty(plot_data$avg_qual, n = 5))+
  scale_x_continuous(name = "Length used", breaks = pretty(plot_data$used, n = 5))+
  theme(plot.title = element_blank()) +
  labs(title = "Sequence quality vs used lengths")
  ggsave(p_check, filename = sanger_qual_vs_used,  units = "in",width = 10, height = 6.5)
}
```

# Investigate strains with strict and seq

```{r}
#merge rows from same well and filter NA - end with 166
strict_1st_ana <- Results_long_norm_yp49_51_1st_ana %>% select('plate', 'well', 'Amt_norm')

nrow(strict_1st_ana %>% filter(!well %in% c("H07","H08","H09","H10","H11","H12"))) #235 of the 240 strains had analytics results
nrow(strict_1st_ana %>% filter(well %in% c("H07","H08","H09","H10","H11","H12")))

#Combine strictosidine and sequencing data
sequences_strict_re_1st_ana <- sequences_by_well_re %>% 
    left_join(strict_1st_ana, by = c("plate","well")) 

#Remove all data missing a seq or strictodisine
sequences_strict_NO_NA_re_1st_ana <- sequences_strict_re_1st_ana %>%
    drop_na() %>%
  select(plate, well, "0","1","2","3", Amt_norm)
nrow(sequences_strict_NO_NA_re_1st_ana) #166 had strict and seq data



```

## Investigate the non included strains

```{r}
#Take strictosidine data and rearrange
res_1st_ana <- Results_long_norm_yp49_51_1st_ana %>%
  group_by(strain,compound, plate, well) %>%
  summarise(Amt_uM_me = mean(Amt_uM, na.rm = TRUE),
            Amt_uM_se = sd(Amt_uM, na.rm = TRUE) / sqrt(n()),
            Amt_norm_me = mean(Amt_norm, na.rm = TRUE),
            Amt_norm_se = sd(Amt_norm, na.rm = TRUE) / sqrt(n())) %>%
  ungroup() %>%
  mutate(control = ifelse(grepl("^CH-A2", strain), "yes",
                          ifelse(grepl("^CM-3", strain), "perhaps", "no"))) %>%
  left_join(sequences_strict_re , by = c('plate','well')) %>%
  left_join(g8h_trans, by= '0') %>%
  rename('g8h' = homolog ) %>%
  left_join(pg8h_trans_int, by= '1') %>%
  rename('pg8h' = homolog) %>%
  left_join(pcpr_trans_int, by= '2')%>%
  rename('pcpr' = homolog) %>%
  left_join(cpr_trans, by= '3') %>%
  rename('cpr' = homolog) %>%
  select(strain, plate, well, Amt_norm_me, Amt_norm_se, control, g8h, pg8h, pcpr, cpr,compound)

#Create column to tell if result are used or not
res_1st_ana$ID <- paste(res_1st_ana$plate, "_",res_1st_ana$well)  
sequences_strict_NO_NA_re_1st_ana$ID <- paste(sequences_strict_NO_NA_re_1st_ana$plate, "_",sequences_strict_NO_NA_re_1st_ana$well)  
res_1st_ana$included <- ifelse(res_1st_ana$ID %in% sequences_strict_NO_NA_re_1st_ana$ID, 'yes','no')

#find top 20 strictodisine producers
top_20_1st_ana  <- res_1st_ana[(res_1st_ana$control=="no"),] %>%
  top_n(n=20, Amt_norm_me) %>%
  arrange(desc(Amt_norm_me))

#Those not included
top_20_not_included_1st_ana <- top_20_1st_ana%>%
  filter(top_20_1st_ana$included == 'no')

nrow(top_20_not_included_1st_ana)##0

```

```{r}
# Plot strict data of top 20, and divde based on included or not
p_top20 <- top_20_1st_ana%>%
  ggplot() +
  geom_bar(aes(x=strain, y=Amt_norm_me, fill=included, order = Amt_norm_me), stat="identity", alpha=0.7) + #  # , fill = run
    scale_fill_manual(values = c( "no"="tomato", "yes"="skyblue" ),  labels = c("Not included", "Included") ) +
  geom_errorbar(aes(x=strain, ymin=Amt_norm_me-Amt_norm_se, ymax=Amt_norm_me+Amt_norm_se), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_hline(yintercept=100, linetype="dashed", color = "black", size=1) +
  facet_grid(compound~.) +
  theme(axis.text.x = element_text(size = 14, angle = 90, vjust = 0.5, hjust=1)) +
  labs(x = "Strain_name (MIA-_G8H_CPR_colony-number)", y = "Normalised Strictosidine Concentration(uM/uM CH-A2 * 100)", title = "Top 20 strictodisine producers")
p_top20
#Save plot
#ggsave(p_top20,filename=here("reports","ch_presentation","figures","g8h_cpr_library_top_20_includ_notinclud_re.png"), height = 10, width = 20)
```

## Investigate occurances and dublicates

```{r}
#find dublicates - in this case 8 was found
sequences_strict_NO_NA_re_1st_ana  %>%  
  unite(MERGED, "0":"3", remove = FALSE)%>%
  group_by(MERGED)%>%
  dplyr::filter(n()>1)

#Keep only one of the dublicate
occurrence <- sequences_strict_NO_NA_re_1st_ana %>% 
  #select(-Amt_uM) %>%
  unite(MERGED, "0":"3", remove = FALSE)%>%
  group_by(MERGED)%>%
  filter(well == min(well)) %>% 
  distinct

#Count the number of occurances for each part
count_one <- occurrence %>% 
  ungroup() %>%
  count(occurrence$'1')%>%
  add_column("part" = 'pG8H')%>%
  left_join(promoters, by = c( `occurrence$"1"`="number"))%>%
  select(n, name, part) %>%
  arrange(n) 

count_two <- occurrence %>% 
  ungroup() %>%
  count(occurrence$'2')%>%
  add_column("part" = 'pCPR')%>%
  left_join(promoters, by = c( `occurrence$"2"`="number"))%>%
  select(n, name, part)%>%
  arrange(n)

count_three <- occurrence %>% 
  ungroup() %>%
  count(occurrence$'3')%>%
  add_column("part" = 'CPR')%>%
  left_join(cpr_trans , by = c(`occurrence$"3"`='3'))%>%
  select(n, name = homolog, part)%>%
  arrange(n)

count_zero <- occurrence %>% 
  ungroup() %>%
  count(occurrence$'0')%>%
  add_column("part" = 'G8H')%>%
  left_join(g8h_trans, by = c( `occurrence$"0"`='0'))%>%
  select(n, name = homolog, part)%>%
  arrange(n)
```

```{r}
### Creating plots of data distribution
palette <-  c( "#d9d9d9","#bdbdbd", "#969696", "#737373", "#525252","#252525")
palette1 <- c( "#fef0d9","#fdd49e", "#fdbb84", "#fc8d59", "#e34a33","#b30000")
palette2 <- c( "#fef0d9","#fef0d9", "#fef0d9","#fdbb84", "#fc8d59", "#e34a33")
#Plot the count of each part
rbind(count_zero, count_one, count_two, count_three)%>%
  #Count occurance of each part
  group_by(part) %>% mutate(cs = cumsum((n))) %>% ungroup() %>%
  mutate(name = fct_reorder(name, n, .desc = TRUE)) %>%
  
  #Plot
  ggplot(aes(x = factor(part, level = c('G8H','pG8H','pCPR','CPR')), y=n, fill = n)) +
  geom_bar(stat="identity", color = "black") +
  geom_text(aes(y=cs, label=str_c(name,n,sep=" ")), vjust=1.6, color="black", size=5) +
  theme_minimal(20) +
  labs(x = "Part", y = "Count", title = "Count of parts for ML") +
  scale_fill_gradient(low = "#2B66AC", high = "#BDD7EE")+
  theme(legend.position="none", plot.title = element_blank(),axis.ticks.x=element_blank(), axis.ticks.y=element_blank(), axis.text.y=element_blank())
#ggsave(here("reports","ch_presentation","figures","Count_parts_for_ML_reseq_1st_ana.png"), width = 7, height = 5, units = "in")


#plot the percental distribution
rbind(count_zero, count_one, count_two, count_three)%>%
  #Find percental distribution
  group_by(part) %>%
  mutate(percent_weight = round(as.numeric(n / sum(n) * 100),1))%>% ungroup() %>%
  mutate(name = fct_reorder(name, n, .desc = TRUE)) %>%
  group_by(part) %>% mutate(cs = cumsum((percent_weight))) %>% ungroup() %>%
  
  #plot
  ggplot(aes(x = factor(part, level = c('G8H','pG8H','pCPR','CPR')), y=percent_weight, fill = cs)) +
  geom_bar(position="stack", stat="identity", color = "black") +
  geom_text(aes(y=cs, label=str_c(name,percent_weight,sep=" ")), vjust=1.6, color="black", size=3.5) +
  scale_fill_gradient(low = "#2B66AC", high = "#BDD7EE")+
  theme_minimal(20) +
  labs(x = "part", y = "%", title = "Part distribution of uniqe strains for ML")+
  theme(legend.position="none", plot.title = element_text(hjust = 0.1))

#Save plot  
#ggsave(here("reports","ch_presentation","figures","Percentage_parts_for_ML_reseq.pdf"), width = 7, height = 7, units = "in")
```

## Output for ML

```{r}
#Save sequencing data for each promoter
reseq %>%
  mutate(inf_promoter = ifelse((Keep == "True"|Keep == TRUE), inf_promoter, NA)) %>%
  select(-ID, -Keep) %>%
  write_csv(here("data","processed","010_Sequencing files","sanger_data_by_position_reseq.csv"))

#Save seq result for each well
sequences_strict_NO_NA_re_1st_ana %>%
    #select(-Amt_uM) #%>%
    write_csv(here("data","processed","010_Sequencing files","sanger_data_by_well_reseq.csv"))

test <- sequences_strict_NO_NA_re_1st_ana %>% unite(Combinations, "0":"3")
length(unique(test$Combinations))

#Format the normalised data
Res_for_ml_1st_ana <- Results_long_norm_yp49_51_1st_ana %>%
  mutate(`Line Name` = str_c(plate,well,sep = "_")) %>%
  select(`Line Name`, Amt_norm)

#Combine accepted strains with normalised data
ML_input <- sequences_strict_NO_NA_re_1st_ana %>%
   mutate(`Line Name` = str_c(plate,well,sep = "_")) %>%
   left_join(Res_for_ml_1st_ana, by =c("Line Name" = "Line Name")) %>%
  rename(Amt_norm = Amt_norm.x)

#Save data file for ML
ML_input %>%
  select("Line Name", "0", "1", "2", "3", Amt_norm) %>%
  write_csv(here("data","processed","input_for_ml_reseq.csv")) # Be aware that this is not the original one used. 

```

#### Plot production coloured based on included/not

```{r echo=FALSE, fig.height=10, fig.width=20}
#Combine those ready for ml and the strictosidine production
define_ml <- Results_long_norm_yp49_51_1st_ana %>%
  mutate("Line Name" = str_c(plate,well,sep = "_"))
used_in_ml <- read_csv(here("data","processed","input_for_ml_reseq.csv"))  %>%
  left_join(define_ml, "Line Name") 
input_for_ml <- read_csv(here("data","processed","input_for_ml_2.csv")) 

#find mean values of controls. 
production_plot <- Results_long_norm_yp49_51_1st_ana%>%
  mutate("Line Name" = str_c(plate,well,sep = "_")) %>%
  mutate("Included" = ifelse(`Line Name` %in% input_for_ml$`Line Name`, "Yes", "No")) %>%
  group_by(strain,compound) %>%
  summarise(Amt_uM_me = mean(Amt_uM, na.rm = TRUE),
            Amt_uM_se = sd(Amt_uM, na.rm = TRUE) / sqrt(n()),
            Amt_norm_me = mean(Amt_norm, na.rm = TRUE),
            Amt_norm_se = sd(Amt_norm, na.rm = TRUE) / sqrt(n())) %>%
  ungroup() %>%
  mutate(control = ifelse(grepl("^CH-A2", strain), "yes",
                          ifelse(grepl("^CM-3", strain), "perhaps", "no"))) %>%
  mutate("Included" = ifelse(strain %in% used_in_ml$strain, 
                             "Yes", 
                                    ifelse(grepl("^CH-A2", strain), 
                                    "Reference",
                                              ifelse(grepl("^CM-3", strain), 
                                              "CM-3", 
                                                      "No"))))  %>%
  
  filter(!strain %in% c("CH-A2_B07_Cro Cro_1","CH-A2_B08_Cro Cro_2","CH-A2_B09_Cro Cro_3","CH-A2_C07_Cro Cro_1","CH-A2_C08_Cro Cro_2","CH-A2_C09_Cro Cro_3","CH-A2_D07_Cro Cro_1","CH-A2_D08_Cro Cro_2","CH-A2_D09_Cro Cro_3","CM-3_A10_Cro Cro_1","CM-3_A11_Cro Cro_2","CM-3_A12_Cro Cro_3","CM-3_B10_Cro Cro_1","CM-3_B11_Cro Cro_2", 
                          "CM-3_B12_Cro Cro_3", 
                          "CM-3_C10_Cro Cro_1", 
                          "CM-3_C11_Cro Cro_2", 
                          "CM-3_C12_Cro Cro_3", 
                          "CM-3_D10_Cro Cro_1", 
                          "CM-3_D11_Cro Cro_2", 
                          "CM-3_D12_Cro Cro_3")) %>%
  filter(!(Included == "CM-3"))


library(viridis)
library(ggsci)
#wihtout strain name
p1 <-  production_plot %>%
  ggplot() +
  geom_bar(aes(x=reorder(strain, -Amt_norm_me), y=Amt_norm_me, fill=Included), stat="identity", width =0.7) + 
  geom_errorbar(aes(x=strain, ymin=Amt_norm_me-Amt_norm_se, ymax=Amt_norm_me+Amt_norm_se), width=0.4, colour="black", alpha=0.9, size=1) +
  geom_hline(yintercept=100, linetype="dashed", color = "black", size=1) +
  #facet_grid(compound~.) +
  labs(x = "Individual strains ordered by production", y = "Normalised Strictosidine")+
  theme(axis.text.x = element_blank())+
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual(values = c( "Reference"="#0073C3", "Yes"="#868686", "No" = "#EFC000")) 
  #scale_fill_jco()

p1  
ggsave(p1,filename=here("reports","ch_presentation","figures","g8h_cpr_library_normalised_1st_1st.jpg"), height = 6, width = 16)

#With strain names
p2 <-  production_plot %>%
  ggplot() +
  geom_bar(aes(x=reorder(strain, -Amt_norm_me), y=Amt_norm_me, fill=Included), stat="identity", width =0.7) + 
  geom_errorbar(aes(x=strain, ymin=Amt_norm_me-Amt_norm_se, ymax=Amt_norm_me+Amt_norm_se), width=0.4, colour="black", alpha=0.9, size=1) +
  geom_hline(yintercept=100, linetype="dashed", color = "black", size=1) +
  labs(x = "Individual strains ordered by production", y = "Normalised Strictosidine Conc.")+
  theme(axis.text.x = element_blank())+
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_jco() +
  theme(axis.text.x = element_text(size = 7, angle = 90, vjust = 0.5, hjust=1)) +
  labs(x = "Strain_name (MIA-_G8H_CPR_well_colony-number)", y = "Normalised Strictosidine Concentration(uM/uM CH-A2 * 100)")

p2  


#ggsave(p2,filename=here("reports","ch_presentation","figures","g8h_cpr_library_normalised_1st_1st_wnames.jpg"), height = 10, width = 25)

```

### Ani and Ahu productions

```{r echo=FALSE, fig.height=10, fig.width=20}
#Investigate the production data of AniCPR and AhuCPR
Ani_Ahu <-production_plot %>%
  separate(strain, sep = "_", into = c("no", 'well', "genes"), remove = FALSE) %>%
  separate(genes, sep = " ", into = c("G8H", "CPR")) %>%
  filter(CPR == 'Ahu' | CPR == 'Ani')

p12 <-  Ani_Ahu%>%
  ggplot() +
  geom_bar(aes(x=reorder(strain, -Amt_norm_me), y=Amt_norm_me, fill=Included), stat="identity", width =0.7) + 
  geom_hline(yintercept=100, linetype="dashed", color = "black", size=1) +
  #facet_grid(compound~.) +
  labs(x = "Strain_name (MIA-_G8H_CPR_colony-number)", y = "Normalised Strictosidine")+
  theme(axis.text.x = element_text(size = 13, angle = 90, vjust = 0.5, hjust=1))+
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual(values = c("Yes"="#868686", "No" = "#EFC000")) 
  #scale_fill_jco()

p12  
ggsave(p12  ,filename=here("reports","ch_presentation","figures","g8h_cpr_library_analytics_Ahu_Ani.pdf"), height = 6, width = 10)
```

# ML

```{r}
#initialise
library(h2o)
library(tidyverse)
library(here)
source(here("src","data","extract_aml_cvpreds.R"))
source(here("src","data","extract_best_non_ensemble.R"))
source(here("src","visualization","plot_obs_vs_pred.R"))

#initialise H2O
h2o.init(max_mem_size = "4G")
h2o.removeAll()
```

## Run H2O autoML

```{r}
#This takes a while!
#Run autoML with an old input
input_for_ml <- read_csv(here("data","processed","input_for_ml_2.csv")) 

input_for_ml_h2o <- input_for_ml %>%
  as.h2o()

aml_2 <-  h2o.automl(y = "Amt_norm",
                   x = c("0","1","2","3"),
                   training_frame = input_for_ml_h2o,
                   seed = 1,
                   keep_cross_validation_predictions = TRUE,
                   #sort_metric = 'MAE',
                   max_runtime_secs = 0
                   )
```

```{r}
#Run autoML with a new input
input_for_ml <- read_csv(here("data","processed","input_for_ml_reseq.csv")) 

input_for_ml_h2o <- input_for_ml %>%
  as.h2o()

aml <-  h2o.automl(y = "Amt_norm",
                   x = c("0","1","2","3"),
                   training_frame = input_for_ml_h2o,
                   seed = 1,
                   keep_cross_validation_predictions = TRUE,
                   #sort_metric = 'MAE',
                   max_runtime_secs = 0
                   )
```

### Processing model

```{r}
#Extract model
model        = aml@leader
model_bn    = extract_best_non_ensemble(aml)
aml_cvpreds  = extract_aml_cvpreds(aml)
h2o.mae(model)


model_ori        = aml_2@leader
model_bn_ori    = extract_best_non_ensemble(aml_2)
aml_cvpreds_ori  = extract_aml_cvpreds(aml_2)
h2o.mae(model_bn_ori)

```

```{r}
### Save on local storage
#Save the model to local storage
model_path <- h2o.saveModel(object = model, path = "/Users/Christine/Documents/Speciale/g8h_cpr_library/data/processed", force = TRUE)
print(model_path)

# load the model
#saved_model <- h2o.loadModel(model_path+'/DeepLearning_grid__2_AutoML_20210823_195148_model_5')

# download the model built above to your local machine
my_local_model <- h2o.download_model(model, path = "/Users/Christine/Documents/Speciale/g8h_cpr_library/data/processed/")
my_local_model_bn <- h2o.download_model(model_bn_ori, path = "/Users/Christine/Documents/Speciale/g8h_cpr_library/data/processed/First_Ana_reseq_H2oUpdate")
```

```{r}
# upload the downloaded model to the H2O cluster
#This can only be performed with version 3.32.1.4
uploaded_model <- h2o.upload_model('/Users/Christine/Documents/Speciale/g8h_cpr_library/data/processed/DeepLearning_grid__2_AutoML_20210823_195148_model_5')
```

### Investigate model

```{r}

aml_cvpreds_id <- uploaded_model@model$cross_validation_holdout_predictions_frame_id$name
aml_cvpreds <- h2o.getFrame(aml_cvpreds_id)

#Investigate the ability to predict based on cross validations
plot_obs_vs_pred(data = input_for_ml_h2o, cvpreds=aml_cvpreds_ori, titel = "aml")
#DOES THE SAME AS ABOVE

#Create plot
cvpreds <- aml_cvpreds[!is.na(aml_cvpreds)]
cvpreds = as.h2o(cvpreds)
names(cvpreds)[1] = c("predict")
  
h2o.cbind(input_for_ml_h2o, cvpreds) %>%
    as.data.frame() %>%
    ggplot() +
    geom_point(aes(x=Amt_norm, y = predict), size = 0.7) +
    scale_colour_brewer(palette = "Set3") +
    coord_fixed(ratio = 1) +
    geom_abline(slope=1, intercept=0) +
    labs(title = "Observed vs predicted", x="Measured & normalised strictosidine concentration", y="Predictied normalised strictosidine concentration") +
    theme(legend.position="none", plot.title = element_text(hjust = 0.5),
          axis.title.y = element_text( size = 12),
          axis.title.x = element_text( size = 12))

#ggsave(here("reports","ch_presentation","figures","obs_vs_pred.pdf"), width = 7, height = 7, units = "in")
```

```{r}
#### Learning curve
learning_curve <- h2o.learning_curve_plot(model_bn)
print(learning_curve)

learning_curve_o <- h2o.learning_curve_plot(model_bn_ori)
print(learning_curve_o)
```

```{r}
#### Performance metrics
perf_st <- h2o.performance(model)
perf_st
```

```{r}
perf_bn <- h2o.performance(model_bn_ori)
perf_bn
```

### Predict phenotype

```{r}
#Find predicted genotypes
found_comb <- input_for_ml %>%
  mutate(genotype = str_c(`0`,`1`,`2`,`3`,sep="_")) %>%
  distinct(genotype) %>%
  .$genotype
length(found_comb)

possible_comb <- do.call(paste, c(expand.grid(g8h_trans$`0`, pg8h_trans$`1`,pcpr_trans$`2`,cpr_trans$`3`)))
possible_comb <- str_replace_all(possible_comb, " ", "_")
length(possible_comb)

not_found_comb <- setdiff(possible_comb, found_comb)
length(not_found_comb)

not_found_comb <- not_found_comb %>%
  as_tibble() %>%
  separate(value,  sep = "_", into = c("0","1","2","3"), convert=TRUE) %>%
  as.h2o()

#Find predictions and save in csv file  
predicted <- h2o.predict(model_bn_ori, not_found_comb)

predicted_genotypes <- h2o.cbind(not_found_comb, predicted) %>%
  as_tibble() %>%
  select("G8H" = "X0","pG8H" = "X1","pCPR" = "X2","CPR" = "X3", predict) %>%
  arrange(desc(predict)) 

write.csv(predicted_genotypes,here("data", "processed", "First_ana_original_H2oUpdate", "Predicted_strict.csv"), row.names = FALSE)

#Find predictions and save in csv file  
possible_comb <- possible_comb %>%
  as_tibble() %>%
  separate(value,  sep = "_", into = c("0","1","2","3"), convert=TRUE) %>%
  as.h2o()
predicted_all <- h2o.predict(model_bn_ori, possible_comb)

predicted_genotypes_all <- h2o.cbind(possible_comb, predicted_all) %>%
  as_tibble() %>%
  select("G8H" = "X0","pG8H" = "X1","pCPR" = "X2","CPR" = "X3", predict) %>%
  arrange(desc(predict)) 

write.csv(predicted_genotypes_all,here("data", "processed", "First_ana_original_H2oUpdate", "Predicted_strict_all_1st.csv"), row.names = FALSE)
```

## Rerun ml to get approximation of r2

```{r}
#Run autoML with an old input
input_for_ml <- read_csv(here("data","processed","input_for_ml_2.csv")) 

input_for_ml_h2o <- input_for_ml %>%
  as.h2o()

aml_re <-  h2o.automl(y = "Amt_norm",
                   x = c("0","1","2","3"),
                   training_frame = input_for_ml_h2o,
                   seed = 1,
                   keep_cross_validation_predictions = TRUE,
                   include_algos = "DeepLearning",
                   max_runtime_secs = 0
                   )

model        = aml_re@leader
aml_cvpreds  = extract_aml_cvpreds(aml_re)
h2o.mae(model)

model_path <- h2o.saveModel(object = model, path = "/Users/Christine/Documents/Speciale/Hand_in/g8h_cpr_library/data/processed/1stML", force = TRUE)
print(model_path)

# download the model built above to your local machine
my_local_model <- h2o.download_model(model, path = "/Users/Christine/Documents/Speciale/Hand_in/g8h_cpr_library/data/processed/1stML/")

cached_cv_pred_mae_files_path <-  here("data","processed","1stML", "cross_validation_predictions_rerun_ml_r2.csv")
if (!file.exists(cached_cv_pred_mae_files_path )) {
  ##Find cross validation predictions and save them 
  aml_cvpreds_id <- aml_re@leader@model$cross_validation_holdout_predictions_frame_id$name
  aml_cvpreds <- h2o.getFrame(aml_cvpreds_id)
  cvpreds <- aml_cvpreds[!is.na(aml_cvpreds)]
  cvpreds = as.h2o(cvpreds)
  names(cvpreds)[1] = c("predict")
  
  
  cv_predictions <- h2o.cbind(input_for_ml_h2o, cvpreds) %>%
      as.data.frame() 
  write_csv(cv_predictions, file =  here("data","processed","1stML", "cross_validation_predictions_rerun_ml_r2.csv"))
} else {
cv_predictions  <- read_csv(cached_cv_pred_mae_files_path )
} 
  

```

```{r}
#this model is with H2O version 3.34.0.4
model_re_1 <- h2o.upload_model('/Users/Christine/Documents/Speciale/Hand_in/g8h_cpr_library/data/processed/1stML/DeepLearning_grid_1_AutoML_2_20211231_151616_model_7')
model_re_1 
cv_predictions_1  <- read_csv(here("data","processed","1stML", "cross_validation_predictions_rerun_ml_r2.csv"))


#This model is with version 3.32.1.4
#model_re_2 <- h2o.upload_model('/Users/Christine/Documents/Speciale/Hand_in/g8h_cpr_library/data/processed/1stML/DeepLearning_grid__2_AutoML_20211231_170207_model_1')
#model_re_2
cv_predictions_2  <- read_csv(here("data","processed","1stML", "cross_validation_predictions_run_to_appendix_two.csv"))

### Find r^2 
lm_eqn <- function(df, x, y){
    m <- lm(predict ~ Amt_norm, cv_predictions_1);
    eq <- substitute(~~italic(r)^2~"="~r2, 
         list(
             r2 = format(summary(m)$r.squared, digits = 3)))
    as.character(as.expression(eq));
}

p<- cv_predictions_1 %>%
    ggplot() +
    geom_point(aes(x=Amt_norm, y = predict), size = 0.7, alpha = 0.5) +
    scale_colour_brewer(palette = "Set3") +
    coord_fixed(ratio = 1) +
    geom_abline(slope=1, intercept=0) +
    labs(title = "Observed vs cv-predictions", x="Measured & normalised strictosidine", y="CV predictied value") +
    theme(legend.position="none", plot.title = element_blank())+#,
          #axis.title.y = element_text( size = 12),
          #axis.title.x = element_text( size = 12))+
    geom_text(x = 25, y = 130, label = lm_eqn(cv_predictions_1, x = Amt_norm, y = predict), parse = TRUE)
p
ggsave(p, file = here("reports","ch_presentation","figures_reanalytics","obs_vs_pred_1st_rerunML_get_R2.jpg"), width = 10, height = 10, units = "in")


lm_eqn <- function(df, x, y){
    m <- lm(predict ~ Amt_norm, cv_predictions_2);
    eq <- substitute(~~italic(r)^2~"="~r2, 
         list(
             r2 = format(summary(m)$r.squared, digits = 3)))
    as.character(as.expression(eq));
}

p_2<- cv_predictions_2 %>%
    ggplot() +
    geom_point(aes(x=Amt_norm, y = predict), size = 0.7, alpha = 0.5) +
    scale_colour_brewer(palette = "Set3") +
    coord_fixed(ratio = 1) +
    geom_abline(slope=1, intercept=0) +
    labs(title = "Observed vs cv-predictions", x="Measured & normalised strictosidine", y="CV predictied value") +
    theme(legend.position="none", plot.title = element_blank())+#,
          #axis.title.y = element_text( size = 12),
          #axis.title.x = element_text( size = 12))+
    geom_text(x = 25, y = 130, label = lm_eqn(cv_predictions_2, x = Amt_norm, y = predict), parse = TRUE)
p_2
ggsave(p_2, file = here("reports","ch_presentation","figures_reanalytics","obs_vs_pred_1st_rerunML_get_R2_2.jpg"), width = 10, height = 10, units = "in")
```

### Shut down H2O

```{r}
h2o.shutdown(prompt=FALSE)
```

## Read the predictions and plot

```{r}
predictions <- read_csv(here("data/processed/First_ana_original_H2oUpdate/Predicted_strict.csv"))

predictions['predict'] <-round(predictions['predict'],digits = 0)
named_pred <- predictions %>%
  left_join(g8h_trans_int, by = c('G8H'='0')) %>%
  select(homolog, pG8H, pCPR, CPR, predict)%>%
  rename(G8H = homolog) %>%
  left_join(pg8h_trans_int, by = c('pG8H'='1')) %>%
  select(G8H, homolog, pCPR, CPR, predict)%>%
  rename(pG8H = homolog) %>%
  left_join(pcpr_trans_int, by = c('pCPR'='2')) %>%
  select(G8H, pG8H, homolog, CPR, predict)%>%
  rename(pCPR = homolog) %>%
  left_join(cpr_trans_int, by = c('CPR'='3')) %>%
  select(G8H, pG8H, pCPR, homolog, predict)%>%
  rename(CPR = homolog)

write.csv(named_pred,here("data", "processed",  "First_ana_original_H2oUpdate","Predicted_strict_named.csv"), row.names = FALSE)
#Python was used to iterate through the data.
```

```{r}
# Plot the predictions of the combiations used for next round
predictions_named <- read_csv(here("data/processed/Predicted_strict_named.csv")) 

predictions_named <- predictions_named[order(-predictions_named$predict),]
p <-  predictions_named %>% 
  head(20) %>%
  mutate(Combinations = str_c(G8H, pG8H, pCPR, CPR,sep="_"))%>%
  ggplot() +
  geom_bar(aes(x=reorder(Combinations, predict), y=predict), stat="identity",fill = "#0073c2") +
  geom_text(aes(x=reorder(Combinations, predict), y=5, label=predict), hjust = 'left', size=7, color="white")+
  geom_hline(yintercept=100, linetype="dashed", color = "black", size=1) +
  theme(axis.text.x = element_text(size = 2, angle = 90,  hjust=1, vjust = 0.5)) +
  labs(x = "Combinations", y = "Predicted value")+
  coord_flip()
p  
ggsave(p,filename=here("reports","ch_presentation","figures","predictions_of_encountered_parts.png"), height = 8, width = 7)
#Python was used to iterate through the data.
```

```{r}
#### Plot top 200 distribution
predictions <- read_csv(here("data/processed/Predicted_strict.csv"))

#Visualise the part distribution of the top 200
predicted_genotypes_200 <- predictions %>% head(200)
pred_table_200 = predcount(predicted_genotypes_200)
pred_percentage_plot(pred_table_200, 'First_ML_200_re', '200')
pred_count_plot(pred_table_200, 'First_ML_count_re', '200') 
```

```{r}
#this funciton is slow thus we cache results
#Find predictions 
cached_predictions_path <- here("data","processed","1stML","All_combinations_pred_1stML.csv")
if (!file.exists(cached_predictions_path)) {
  library(h2o)
  library(tidyverse)
  library(here)

  h2o.init(max_mem_size = "4G")
  h2o.removeAll()
  uploaded_model <- h2o.upload_model('/Users/Christine/Documents/Speciale/g8h_cpr_library/data/processed/DeepLearning_grid__2_AutoML_20210823_195148_model_5')


  possible_comb_h2o <- str_replace_all(possible_comb, " ", "_")%>%
  as_tibble() %>%
  separate(value,  sep = "_", into = c("0","1","2","3"), convert=TRUE) %>%
  as.h2o()
  
  predicted <- h2o.predict(uploaded_model , possible_comb_h2o)

  predicted_genotypes_1st_model<- h2o.cbind(possible_comb_h2o, predicted) %>%
    as_tibble() %>%
    select("G8H" = "X0","pG8H" = "X1","pCPR" = "X2","CPR" = "X3", predict) %>%
    arrange(desc(predict))
  
  predicted_genotypes_1st_model['predict'] <-round(predicted_genotypes_1st_model['predict'],digits = 2)
  write_csv(predicted_genotypes_1st_model,cached_predictions_path)
  All_predicted <- predicted_genotypes_1st_model
} else {
All_predicted<- read_csv(cached_predictions_path)
}

#find combiantions predicted named
cached_predictions_path <- here("data","processed","1stML","All_combinations_pred_1stML_named.csv")
if (!file.exists(cached_predictions_path)) {
  
  named_pred <- All_predicted %>%
  left_join(g8h_trans_int, by = c('G8H'='0')) %>%
  select(homolog, pG8H, pCPR, CPR, predict)%>%
  rename(G8H = homolog) %>%
  left_join(pg8h_trans_int, by = c('pG8H'='1')) %>%
  select(G8H, homolog, pCPR, CPR, predict)%>%
  rename(pG8H = homolog) %>%
  left_join(pcpr_trans_int, by = c('pCPR'='2')) %>%
  select(G8H, pG8H, homolog, CPR, predict)%>%
  rename(pCPR = homolog) %>%
  left_join(cpr_trans_int, by = c('CPR'='3')) %>%
  select(G8H, pG8H, pCPR, homolog, predict)%>%
  rename(CPR = homolog)
  write_csv(named_pred,cached_predictions_path)
} else {
All_predicted_named<- read_csv(cached_predictions_path)
}
```

#### Plot the parts for 2nd cycle

```{r}
#Parts were determined in the python script
G8H<- c('Rsep', 'Smus', 'Oeu', 'Vmin')
pG8H <- c('ENO2', 'CYC1')
pCPR <- c('TPI1', 'CCW12')
CPR <- c('Clo', 'Ara', 'Rse', 'Aan', 'Cro', 'Ani')

G8H<- c('Rsep', 'Smus', 'Oeu', 'Vmin')
pG8H <- c('ENO2', 'CYC1')
pCPR <- c('TPI1', 'CCW12')
CPR <- c('Clo', 'Ara', 'Rse', 'Aan', 'Cro', 'Ani', 'Ahu')

G8H<- c('Rsep', 'Smus', 'Oeu', 'Vmin', 'Cro')
pG8H <- c('ENO2', 'CYC1')
pCPR <- c('TPI1', 'CCW12')
CPR <- c('Clo', 'Ara', 'Rse', 'Aan', 'Cro', 'Ani', 'Ahu')



G8H<- c( 'no', 'no','Rsep', 'Smus', 'Oeu', 'Vmin', 'Cro')
G8H <- data.frame(G8H)%>%
  mutate(n = ifelse(G8H == 'no', NA, 1)) %>%
  mutate(part = 'G8H')%>%
  rename(homolog = G8H)
pG8H <- c('no', 'no', 'no', 'no', 'no', 'ENO2', 'CYC1')
pG8H <- data.frame(pG8H)%>%
  mutate(n = ifelse(pG8H == 'no', NA, 1)) %>%
  mutate(part = 'pG8H')%>%
  rename(homolog = pG8H)
pCPR <- c('no', 'no', 'no', 'no','no','TPI1', 'CCW12')
pCPR <- data.frame(pCPR)%>%
  mutate(n = ifelse(pCPR == 'no', NA, 1)) %>%
  mutate(part = 'pCPR')%>%
  rename(homolog = pCPR)
CPR <- c('Clo', 'Ara', 'Rse', 'Aan', 'Cro', 'Ani', 'Ahu')
CPR <- data.frame(CPR)%>%
  mutate(n = ifelse(CPR == 'no', NA, 1)) %>%
  mutate(part = 'CPR') %>%
  rename(homolog = CPR)

pred_table <- rbind(G8H, pG8H, pCPR, CPR)%>%
  filter(!is.na(n))

#Plot the percental distribution
header = paste('Part distribution from ML', sep = '')
plot_parts_used <- pred_table%>%
  group_by(part) %>%
  mutate(percent_weight = round(as.numeric(n / sum(n) * 100),1))%>% ungroup() %>%
  group_by(part) %>% mutate(cs = cumsum((percent_weight))) %>% ungroup() %>%
  ggplot(aes(x = factor(part, level = c('G8H','pG8H','pCPR','CPR')), y=percent_weight, fill = cs)) +
  geom_bar(stat="identity", color = "black") +
  geom_text(aes(y=cs, label=str_c(homolog,sep=" ")), vjust=1.6, color="black", size=5) +
  theme_minimal(20) +
  labs(x = "part", y = "% Distribution", title = header)+
  scale_fill_gradient(low = "#0173C5", high = "#BDD7EE")+
  theme(legend.position="none", plot.title = element_blank())
plot_parts_used
#ggsave(plot_parts_used,filename=here("reports","ch_presentation","figures","g8h_cpr_library_second_sample_space.png"), height = 5, width = 7)
```

# 2nd Analytics

\*\*\* Experiment Due to large deviations in the data found, the LC-MS
was run on all sample again \## Data wrangling

```{r echo=FALSE, message=FALSE, include=FALSE}
#Due to high error rate, the plates were send for analytics again. 
#1. In Emacs; Result_table_210330_tmet494_Soren.xlsx to .csv N.D. -> N.A.; N.A. -> NA.
#2. Change names to plates names e.g.Result_table_210330_tmet494_Soren.csv -> Result_table_210330_tmet494_yp48Feed-Nofeed.xlsx
results_2nd_ana <- tibble(abs_path = list.files(here("data","raw","023_LC-MS", "Rerun"), full.names = TRUE, pattern = ".csv"))  %>%
  #Extract data from filenames
  mutate(filename = tools::file_path_sans_ext(basename(abs_path))) %>%
  separate(filename, into=c("a","b","date","tmetab","description"), sep="_") %>%
  arrange(date) %>%
  mutate(data = map(abs_path,read_csv)) %>%
  unnest(data) %>% 
  select(-abs_path, -a, -b) %>%

  # Excel contains values both in ug and uM. Keep uM and remove "Amt_ul_L_".
  select(!starts_with("Amt_ug_L_"), -...8,-...11, -data_file, -description) %>%
  
  #Fix 210330 sample_notes column
  mutate(sample_notes = gsub("*(_P[12])*","",sample_notes)) %>%
  separate(sample_notes,  sep = "_", into = c('plate','well')) %>%
  arrange(plate, well) %>%
  
  #Rename plate names
  mutate(medium = ifelse(plate=="96WP988","nofeed", ifelse(plate=="96WP989","feed","NA"))) %>%
  mutate(plate = gsub("96WP","",plate),
         plate = gsub("YP49","yp49_old",plate),
         plate = gsub("YP50","yp50_old",plate),
         plate = gsub("YP51","yp51_old",plate),
         plate = gsub("988|989","yp48",plate),
         plate = gsub("1056","yp49",plate),
         plate = gsub("1332","yp50",plate),
         plate = gsub("1333","yp51",plate),
         plate = gsub("1334","yp55",plate),
         plate = gsub("1335","yp56",plate),
         plate = gsub("YP","yp",plate)) %>%

 
  # str pad well column
  mutate(row = str_extract(well, "^([a-zA-Z])"),
         col = str_extract(well, "([0-9]+)$"),
         col = str_pad(col, width=2, pad = "0"),
         well = str_c(row,col,"")) %>%
  select(-row,-col, -Amt_uM_Tetrahydroalstonine)%>%
  filter(!
           (plate == "yp50" & well %in% c("G01", "H01", "H11", "H12")) |
           (plate == "yp51" & well %in% c("A01", "A02", "A03", "A04", "A05", "A06", "A07", "A08", "A09", "A10", "A11", "A12", "B01", "B02", "B03", "B04", "B05", "B06", "B07", "B08", "B09", "B10", "B11", "B12", "H11")) |
           (plate == "yp55" & well %in% c("G08", "H07","H08","H09", "H11") )
           #plate == "yp56" & well %in% c() #This had double peaks
           )

#Check those with NA
NA_results_2nd_ana <- results_2nd_ana %>% filter((is.na(Amt_uM_Loganin) & is.na(Amt_uM_Secologanin) & is.na(Amt_uM_Strictosidine) & is.na(Amt_uM_Tryptamine))) 
  
#Remove NA and replace the rest with 0
results_2nd_ana <- results_2nd_ana %>%
  filter(!(is.na(Amt_uM_Loganin) & is.na(Amt_uM_Secologanin) & is.na(Amt_uM_Strictosidine) & is.na(Amt_uM_Tryptamine))) %>%
  
  #Replace na's with 0
  mutate(across(starts_with("Amt"), ~replace_na(.x, 0)))
```

```{r echo=FALSE, message=FALSE, include=FALSE}
#Join meta with strictosidine results _2nd_ana
results_meta_2nd_ana <- meta %>%
  left_join(results_2nd_ana, by=c("plate","well")) %>%
  mutate(strain = str_c(Strain,well, str_replace_na(genotype), str_replace_na(col_no), sep="_"),
         Strain = str_c(Strain, str_replace_na(genotype), sep="_")) %>%
  mutate(medium = ifelse(plate == "yp48", medium.y, medium.x)) %>%
  select(-medium.y,-medium.x)

#Tidying data
Results_long_2nd_ana <- results_meta_2nd_ana %>%
  pivot_longer(
     cols = `Amt_uM_Loganic acid`:Amt_uM_Tryptophan,
     names_to = "compound",
     names_prefix = "Amt_uM_",
     values_to = "Amt_uM",
     values_drop_na = TRUE)
```

```{r,  echo = FALSE, include=FALSE}
#Normalise results of yp49-51

#Assumptions for the code: 
# Loganing is less interesting as its amount mirrors Strictosidine amount
# Desired outcome: comepare amounts (loganin and strictosidine) between plates.
# Assumption: amounts within one plate is comparable
# Each plate contains the same 4 colonies of MIA-CH-A2
# scale amounts from each plate to MIA-CH-A2 levels

#51
Results_long_norm_yp51 <- Results_long_2nd_ana %>%
  filter(plate == "yp51") %>%
  filter(!(well %in% c("A01", "A02", "A03", "A04", "A05", "A06", "A07", "A08", "A09", "A10", "A11", "A12", "B01", "B02", "B03", "B04", "B05", "B06", "B07", "B08", "B09", "B10", "B11", "B12", "H11"))) %>%
  filter(compound %in% c("Strictosidine")) %>%
  group_by(plate, compound) %>% 
  mutate(Amt_norm = Amt_uM / mean(Amt_uM[well %in% c("H07","H08","H09", "H10")], na.rm = TRUE)*100) %>%
  ungroup()

#50
Results_long_norm_yp50 <- Results_long_2nd_ana %>%
  filter(plate == "yp50") %>%
  filter(!(well %in% c("G01", "H01", "H11", "H12"))) %>%
  filter(compound %in% c("Strictosidine")) %>%
  group_by(plate, compound) %>% 
  mutate(Amt_norm = Amt_uM / mean(Amt_uM[well %in% c("H07","H08","H09", "H10")], na.rm = TRUE)*100) %>%
  ungroup()

#49
Results_long_norm_yp49 <- Results_long_2nd_ana %>%
  filter(plate == "yp49") %>%
  filter(compound %in% c("Strictosidine")) %>%
  group_by(plate, compound) %>% 
  mutate(Amt_norm = Amt_uM / mean(Amt_uM[well %in% c("H07","H08","H09", "H10")], na.rm = TRUE)*100) %>%
  ungroup()


#Append together
Results_long_norm_yp49_51_2nd_ana <- rbind(Results_long_norm_yp49, Results_long_norm_yp50, Results_long_norm_yp51)  
Results_long_norm_yp49_51_2nd_ana
```

```{r echo=FALSE, fig.height=10, fig.width=20}
#Find mean and standard deviations
Results_long_norm_yp49_51_plot <- Results_long_norm_yp49_51_2nd_ana %>%
  group_by(strain,compound) %>%
  summarise(Amt_uM_me = mean(Amt_uM, na.rm = TRUE),
            Amt_uM_se = sd(Amt_uM, na.rm = TRUE) / sqrt(n()),
            Amt_norm_me = mean(Amt_norm, na.rm = TRUE),
            Amt_norm_se = sd(Amt_norm, na.rm = TRUE) / sqrt(n())) %>%
  ungroup() %>%
  mutate(control = ifelse(grepl("^CH-A2", strain), "yes",
                          ifelse(grepl("^CM-3", strain), "perhaps", "no")))

#plot 49-51 normalised strictosidine productions
p49_51_analytics <- Results_long_norm_yp49_51_plot %>%
  ggplot() +
  geom_bar(aes(x=strain, y=Amt_norm_me, fill=control), stat="identity", alpha=0.7) + 
  scale_fill_manual(values = c( "yes"="tomato", "perhaps"="grey", "no"="skyblue" ), guide = FALSE ) +
  geom_errorbar(aes(x=strain, ymin=Amt_norm_me-Amt_norm_se, ymax=Amt_norm_me+Amt_norm_se), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_hline(yintercept=100, linetype="dashed", color = "black", size=1) +
  facet_grid(compound~.) +
  theme(axis.text.x = element_text(size = 7, angle = 90, vjust = 0.5, hjust=1)) +
  labs(x = "Strain_name (MIA-_G8H_CPR_colony-number)", y = "Normalised Strictosidine Concentration(uM/uM CH-A2 * 100)")
p49_51_analytics  

#Save plot
ggsave(p49_51_analytics,filename=here("reports","ch_presentation","figures_reanalytics","g8h_cpr_library_analytics.png"), height = 10, width = 25)
```

```{r}
#Plot top 20 producers
p49_51_top_prod <- top_n(Results_long_norm_yp49_51_plot , n=25, Amt_norm_me) %>%
  ggplot() +
  geom_bar(aes(x=strain, y=Amt_norm_me, fill=control), stat="identity", alpha=0.7) + 
  scale_fill_manual(values = c( "yes"="tomato", "perhaps"="grey", "no"="skyblue" ), guide = FALSE ) +
  geom_errorbar(aes(x=strain, ymin=Amt_norm_me-Amt_norm_se, ymax=Amt_norm_me+Amt_norm_se), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_hline(yintercept=100, linetype="dashed", color = "black", size=1) +
  facet_grid(compound~.) +
  theme(axis.text.x = element_text(size = 14, angle = 90, vjust = 0.5, hjust=1)) +
  labs(x = "Strain_name (MIA-_G8H_CPR_colony-number)", y = "Normalised Strictosidine Concentration(uM/uM CH-A2 * 100)")
ggsave(p49_51_top_prod,filename=here("reports","ch_presentation","figures_reanalytics","g8h_cpr_library_analytics_top20.png"), height = 10, width = 20)

```

##Merge strict and seq

```{r}
#merge rows from same well and filter NA - end with 157 strains
strict_2nd_ana <- Results_long_norm_yp49_51_2nd_ana %>% select('plate', 'well', 'Amt_norm', "Amt_uM")

nrow(strict_2nd_ana %>% filter(!well %in% c("H07","H08","H09","H10","H11","H12"))) #214 of the 240 strains had analytics results

#Combine strictosidine and sequencing data
sequences_strict_re_2nd_ana <- sequences_by_well_re %>% 
    left_join(strict_2nd_ana, by = c("plate","well")) 

#Remove all data missing a seq or strictodisine
sequences_strict_NO_NA_re_2nd_ana <- sequences_strict_re_2nd_ana %>%
    drop_na() %>%
  select(plate, well, "0","1","2","3", Amt_norm, Amt_uM)
nrow(sequences_strict_NO_NA_re_2nd_ana) #151 had strict and seq data
```

```{r}
#Take strictosidine data and rearrange
res_2nd_ana <- Results_long_norm_yp49_51_2nd_ana %>%
  group_by(strain,compound, plate, well) %>%
  summarise(Amt_uM_me = mean(Amt_uM, na.rm = TRUE),
            Amt_uM_se = sd(Amt_uM, na.rm = TRUE) / sqrt(n()),
            Amt_norm_me = mean(Amt_norm, na.rm = TRUE),
            Amt_norm_se = sd(Amt_norm, na.rm = TRUE) / sqrt(n())) %>%
  ungroup() %>%
  mutate(control = ifelse(grepl("^CH-A2", strain), "yes",
                          ifelse(grepl("^CM-3", strain), "perhaps", "no"))) %>%
  left_join(sequences_strict_re , by = c('plate','well')) %>%
  left_join(g8h_trans, by= '0') %>%
  rename('g8h' = homolog ) %>%
  left_join(pg8h_trans_int, by= '1') %>%
  rename('pg8h' = homolog) %>%
  left_join(pcpr_trans_int, by= '2')%>%
  rename('pcpr' = homolog) %>%
  left_join(cpr_trans, by= '3') %>%
  rename('cpr' = homolog) %>%
  select(strain, plate, well, Amt_norm_me, Amt_norm_se, control, g8h, pg8h, pcpr, cpr,compound)

#Create column to tell if result are used or not
res_2nd_ana$ID <- paste(res_2nd_ana$plate, "_",res_2nd_ana$well)  
sequences_strict_NO_NA_re_2nd_ana$ID <- paste(sequences_strict_NO_NA_re_2nd_ana$plate, "_",sequences_strict_NO_NA_re_2nd_ana$well)  
res_2nd_ana$included <- ifelse(res_2nd_ana$ID %in% sequences_strict_NO_NA_re_2nd_ana$ID, 'yes','no')

#find top 20 strictodisine producers
top_20_2nd_ana  <- res_2nd_ana[(res_2nd_ana$control=="no"),] %>%
  top_n(n=20, Amt_norm_me) %>%
  arrange(desc(Amt_norm_me))

#Those not included
top_20_not_included_2nd_ana <- top_20_2nd_ana%>%
  filter(top_20_2nd_ana$included == 'no')

nrow(top_20_not_included_2nd_ana)
```

```{r}
#Find top 50
top_50_2nd_ana  <- res_2nd_ana[(res_2nd_ana$control=="no"),] %>%
  top_n(n=50, Amt_norm_me)%>%
  arrange(desc(Amt_norm_me))
  
#Those not included
top_50_not_included_2nd_ana <- top_50_2nd_ana%>%filter(top_50_2nd_ana$included == 'no')
nrow(top_50_not_included_2nd_ana) #18 are not included

#Find top 20 not sequenced
not_incl_top_2nd_ana <- res_2nd_ana[(res_2nd_ana$control=="no"),]%>%
  filter(included == 'no')%>%
  top_n(n=20, Amt_norm_me)%>%
  arrange(desc(Amt_norm_me))
```

```{r}
# Plot strict data of top 20, and divde based on included or not
p_top20 <- top_20_2nd_ana %>%
  ggplot() +
  geom_bar(aes(x=strain, y=Amt_norm_me, fill=included, order = Amt_norm_me), stat="identity", alpha=0.7) + #  # , fill = run
    scale_fill_manual(values = c( "no"="tomato", "yes"="skyblue" ),  labels = c("Not included", "Included") ) +
  geom_errorbar(aes(x=strain, ymin=Amt_norm_me-Amt_norm_se, ymax=Amt_norm_me+Amt_norm_se), width=0.4, colour="orange", alpha=0.9, size=1.3) +
  geom_hline(yintercept=100, linetype="dashed", color = "black", size=1) +
  facet_grid(compound~.) +
  theme(axis.text.x = element_text(size = 14, angle = 90, vjust = 0.5, hjust=1)) +
  labs(x = "Strain_name (MIA-_G8H_CPR_colony-number)", y = "Normalised Strictosidine Concentration(uM/uM CH-A2 * 100)", title = "Top 20 strictodisine producers")

#Save plot
ggsave(p_top20,filename=here("reports","ch_presentation","figures_reanalytics","g8h_cpr_library_top_20_includ_notinclud.png"), height = 10, width = 20)
```

### Output for ML

```{r}
#Save sequencing data for each promoter

#Format the normalised data
Res_for_ml <- Results_long_norm_yp49_51_2nd_ana %>%
  mutate(`Line Name` = str_c(plate,well,sep = "_")) %>%
  select(`Line Name`, Amt_norm)

#Combine accepted strains with normalised data
ML_input <- sequences_strict_NO_NA_re_2nd_ana %>%
   mutate(`Line Name` = str_c(plate,well,sep = "_")) %>%
   left_join(Res_for_ml, by =c("Line Name" = "Line Name"))%>%
   select(-Amt_norm.y) %>%
   rename(Amt_norm = Amt_norm.x)

#Save data file for ML
ML_input %>%
  select("Line Name", "0", "1", "2", "3", Amt_norm, Amt_uM) %>%
  write_csv(here("data","processed","Second_Analytics","input_for_ml_reseq.csv"))
```
